{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# task-perturb-multiomics-grn\n",
    "## Creating resources\n",
    "### by Jalil Nourisa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multiomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as ad\n",
    "import pandas as pd\n",
    "\n",
    "import numpy as np\n",
    "data_dir = '../../perturb-multiomics-grn/output/'\n",
    "\n",
    "resource_dir = '../resources/grn-benchmark'\n",
    "test_resource_dir = '../resources_test/grn-benchmark'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as ad \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy import sparse\n",
    "import scanpy as sc\n",
    "import requests\n",
    "\n",
    "par = {\n",
    "    'multiomics_rna': '../resources_test/grn-benchmark/multiomics_rna.h5ad',\n",
    "    'multiomics_rna_test': 'resources_test/grn-benchmark/multiomics_rna.h5ad',\n",
    "\n",
    "    'multiomics_atac': '../resources_test/grn-benchmark/multiomics_atac.h5ad',\n",
    "    'multiomics_atac_test': 'resources_test/grn-benchmark/multiomics_atac.h5ad',\n",
    "\n",
    "    'perturbation_data': 'resources/grn-benchmark/perturbation_data.h5ad',\n",
    "    'perturbation_data_test': 'resources_test/grn-benchmark/perturbation_data.h5ad',\n",
    "}\n",
    "\n",
    "n_cells = 1000\n",
    "n_peaks = 5000\n",
    "\n",
    "adata_rna = ad.read_h5ad(par['multiomics_rna'])\n",
    "adata_atac = ad.read_h5ad(par['multiomics_atac'])\n",
    "\n",
    "# adata_rna.var.to_csv('../resources/grn-benchmark/supp/gene_location.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True])"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs = adata_rna.obs\n",
    "\n",
    "cell_topic = pd.read_csv('../resources/grn-benchmark/supp/cell_topic.csv', index_col=0)\n",
    "obs.index.isin(cell_topic.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "adata_rna = ad.read_h5ad('../../../repo/_task-grn-benchmark/resources/grn-benchmark/multiomics_rna.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>location</th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>interval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>chr1:120931-133723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL627309.5</td>\n",
       "      <td>ENSG00000241860</td>\n",
       "      <td>chr1:149706-173862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL627309.4</td>\n",
       "      <td>ENSG00000241599</td>\n",
       "      <td>chr1:160445-160446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LINC01409</td>\n",
       "      <td>ENSG00000237491</td>\n",
       "      <td>chr1:778757-803934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC01128</td>\n",
       "      <td>ENSG00000228794</td>\n",
       "      <td>chr1:827597-860227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22782</th>\n",
       "      <td>AL354822.1</td>\n",
       "      <td>ENSG00000278384</td>\n",
       "      <td>GL000218.1:54892-54893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22783</th>\n",
       "      <td>AL592183.1</td>\n",
       "      <td>ENSG00000273748</td>\n",
       "      <td>GL000219.1:83310-83311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22784</th>\n",
       "      <td>AC240274.1</td>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>KI270711.1:24649-29626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22785</th>\n",
       "      <td>AC004556.3</td>\n",
       "      <td>ENSG00000276345</td>\n",
       "      <td>KI270721.1:2584-2585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22786</th>\n",
       "      <td>AC007325.4</td>\n",
       "      <td>ENSG00000278817</td>\n",
       "      <td>KI270734.1:131493-131494</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22787 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         location         gene_ids                  interval\n",
       "0      AL627309.1  ENSG00000238009        chr1:120931-133723\n",
       "1      AL627309.5  ENSG00000241860        chr1:149706-173862\n",
       "2      AL627309.4  ENSG00000241599        chr1:160445-160446\n",
       "3       LINC01409  ENSG00000237491        chr1:778757-803934\n",
       "4       LINC01128  ENSG00000228794        chr1:827597-860227\n",
       "...           ...              ...                       ...\n",
       "22782  AL354822.1  ENSG00000278384    GL000218.1:54892-54893\n",
       "22783  AL592183.1  ENSG00000273748    GL000219.1:83310-83311\n",
       "22784  AC240274.1  ENSG00000271254    KI270711.1:24649-29626\n",
       "22785  AC004556.3  ENSG00000276345      KI270721.1:2584-2585\n",
       "22786  AC007325.4  ENSG00000278817  KI270734.1:131493-131494\n",
       "\n",
       "[22787 rows x 3 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_location = pd.read_csv('../resources/grn-benchmark/supp/gene_location.csv')\n",
    "gene_location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata_rna.var = adata_rna.var.reset_index().merge(gene_location, on='location')\n",
    "adata_rna.write('../../../repo/_task-grn-benchmark/resources/grn-benchmark/multiomics_rna.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A1BG\n",
      "A1BG-AS1\n",
      "A2M\n",
      "A2M-AS1\n",
      "A2ML1\n",
      "A2ML1-AS1\n",
      "A3GALT2\n",
      "A4GALT\n",
      "AAAS\n",
      "AACS\n",
      "AADACL2-AS1\n",
      "AAGAB\n",
      "AAK1\n",
      "AAMDC\n",
      "AAMP\n",
      "AANAT\n",
      "AAR2\n",
      "AARD\n",
      "AARS\n",
      "Bad request for gene: AARS. Check the gene name and try again.\n",
      "AARS2\n",
      "AARSD1\n",
      "AASDH\n",
      "AASDHPPT\n",
      "AASS\n",
      "AATBC\n",
      "AATF\n",
      "AATK\n",
      "ABALON\n",
      "ABAT\n",
      "ABCA1\n",
      "ABCA10\n",
      "ABCA12\n",
      "ABCA13\n",
      "ABCA2\n",
      "ABCA3\n",
      "ABCA4\n",
      "ABCA5\n",
      "ABCA6\n",
      "ABCA7\n",
      "ABCA8\n",
      "ABCA9\n",
      "ABCA9-AS1\n",
      "ABCB1\n",
      "ABCB10\n",
      "ABCB11\n",
      "ABCB4\n",
      "ABCB5\n",
      "ABCB6\n",
      "ABCB7\n",
      "ABCB8\n",
      "ABCB9\n",
      "ABCC1\n",
      "ABCC10\n",
      "ABCC11\n",
      "ABCC12\n",
      "ABCC2\n",
      "ABCC3\n",
      "ABCC4\n",
      "ABCC5\n",
      "ABCC5-AS1\n",
      "ABCC6\n",
      "ABCC9\n",
      "ABCD1\n",
      "ABCD2\n",
      "ABCD3\n",
      "ABCD4\n",
      "ABCE1\n",
      "ABCF1\n",
      "ABCF2\n",
      "ABCF3\n",
      "ABCG1\n",
      "ABCG2\n",
      "ABCG5\n",
      "ABCG8\n",
      "ABHD1\n",
      "ABHD10\n",
      "ABHD11\n",
      "ABHD12\n",
      "ABHD12B\n",
      "ABHD13\n",
      "ABHD14A\n",
      "ABHD14B\n",
      "ABHD15\n",
      "ABHD15-AS1\n",
      "ABHD16A\n",
      "ABHD17A\n",
      "ABHD17B\n",
      "ABHD17C\n",
      "ABHD18\n",
      "ABHD2\n",
      "ABHD3\n",
      "ABHD4\n",
      "ABHD5\n",
      "ABHD6\n",
      "ABHD8\n",
      "ABI1\n",
      "ABI2\n",
      "ABI3\n",
      "ABI3BP\n",
      "ABITRAM\n",
      "ABL1\n",
      "ABL2\n",
      "ABLIM1\n",
      "ABLIM2\n",
      "ABLIM3\n",
      "ABO\n",
      "ABR\n",
      "ABRA\n",
      "ABRACL\n",
      "ABRAXAS1\n",
      "ABRAXAS2\n",
      "ABT1\n",
      "ABTB1\n",
      "ABTB2\n",
      "AC000065.1\n",
      "Bad request for gene: AC000065.1. Check the gene name and try again.\n",
      "AC000065.2\n",
      "Bad request for gene: AC000065.2. Check the gene name and try again.\n",
      "AC000068.1\n",
      "Bad request for gene: AC000068.1. Check the gene name and try again.\n",
      "AC000085.1\n",
      "Bad request for gene: AC000085.1. Check the gene name and try again.\n",
      "AC000120.1\n",
      "Bad request for gene: AC000120.1. Check the gene name and try again.\n",
      "AC000403.1\n",
      "Bad request for gene: AC000403.1. Check the gene name and try again.\n",
      "AC002044.1\n",
      "Bad request for gene: AC002044.1. Check the gene name and try again.\n",
      "AC002066.1\n",
      "Bad request for gene: AC002066.1. Check the gene name and try again.\n",
      "AC002069.2\n",
      "Bad request for gene: AC002069.2. Check the gene name and try again.\n",
      "AC002070.1\n",
      "Bad request for gene: AC002070.1. Check the gene name and try again.\n",
      "AC002072.1\n",
      "Bad request for gene: AC002072.1. Check the gene name and try again.\n",
      "AC002074.1\n",
      "Bad request for gene: AC002074.1. Check the gene name and try again.\n",
      "AC002091.1\n",
      "Bad request for gene: AC002091.1. Check the gene name and try again.\n",
      "AC002091.2\n",
      "Bad request for gene: AC002091.2. Check the gene name and try again.\n",
      "AC002094.2\n",
      "Bad request for gene: AC002094.2. Check the gene name and try again.\n",
      "AC002094.4\n",
      "Bad request for gene: AC002094.4. Check the gene name and try again.\n",
      "AC002116.2\n",
      "Bad request for gene: AC002116.2. Check the gene name and try again.\n",
      "AC002128.1\n",
      "Bad request for gene: AC002128.1. Check the gene name and try again.\n",
      "AC002128.2\n",
      "Bad request for gene: AC002128.2. Check the gene name and try again.\n",
      "AC002306.1\n",
      "Bad request for gene: AC002306.1. Check the gene name and try again.\n",
      "AC002310.1\n",
      "Bad request for gene: AC002310.1. Check the gene name and try again.\n",
      "AC002310.2\n",
      "Bad request for gene: AC002310.2. Check the gene name and try again.\n",
      "AC002331.1\n",
      "Bad request for gene: AC002331.1. Check the gene name and try again.\n",
      "AC002347.1\n",
      "Bad request for gene: AC002347.1. Check the gene name and try again.\n",
      "AC002377.1\n",
      "Bad request for gene: AC002377.1. Check the gene name and try again.\n",
      "AC002383.1\n",
      "Bad request for gene: AC002383.1. Check the gene name and try again.\n",
      "AC002398.1\n",
      "Bad request for gene: AC002398.1. Check the gene name and try again.\n",
      "AC002401.1\n",
      "Bad request for gene: AC002401.1. Check the gene name and try again.\n",
      "AC002429.2\n",
      "Bad request for gene: AC002429.2. Check the gene name and try again.\n",
      "AC002451.1\n",
      "Bad request for gene: AC002451.1. Check the gene name and try again.\n",
      "AC002454.1\n",
      "Bad request for gene: AC002454.1. Check the gene name and try again.\n",
      "AC002456.1\n",
      "Bad request for gene: AC002456.1. Check the gene name and try again.\n",
      "AC002460.2\n",
      "Bad request for gene: AC002460.2. Check the gene name and try again.\n",
      "AC002463.1\n",
      "Bad request for gene: AC002463.1. Check the gene name and try again.\n",
      "AC002467.1\n",
      "Bad request for gene: AC002467.1. Check the gene name and try again.\n",
      "AC002470.1\n",
      "Bad request for gene: AC002470.1. Check the gene name and try again.\n",
      "AC002480.2\n",
      "Bad request for gene: AC002480.2. Check the gene name and try again.\n",
      "AC002546.1\n",
      "Bad request for gene: AC002546.1. Check the gene name and try again.\n",
      "AC002550.1\n",
      "Bad request for gene: AC002550.1. Check the gene name and try again.\n",
      "AC002550.2\n",
      "Bad request for gene: AC002550.2. Check the gene name and try again.\n",
      "AC002551.1\n",
      "Bad request for gene: AC002551.1. Check the gene name and try again.\n",
      "AC002553.1\n",
      "Bad request for gene: AC002553.1. Check the gene name and try again.\n",
      "AC002553.2\n",
      "Bad request for gene: AC002553.2. Check the gene name and try again.\n",
      "AC002558.2\n",
      "Bad request for gene: AC002558.2. Check the gene name and try again.\n",
      "AC002558.3\n",
      "Bad request for gene: AC002558.3. Check the gene name and try again.\n",
      "AC003005.2\n",
      "Bad request for gene: AC003005.2. Check the gene name and try again.\n",
      "AC003035.1\n",
      "Bad request for gene: AC003035.1. Check the gene name and try again.\n",
      "AC003043.2\n",
      "Bad request for gene: AC003043.2. Check the gene name and try again.\n",
      "AC003077.1\n",
      "Bad request for gene: AC003077.1. Check the gene name and try again.\n",
      "AC003092.1\n",
      "Bad request for gene: AC003092.1. Check the gene name and try again.\n",
      "AC003093.1\n",
      "Bad request for gene: AC003093.1. Check the gene name and try again.\n",
      "AC003102.1\n",
      "Bad request for gene: AC003102.1. Check the gene name and try again.\n",
      "AC003681.1\n",
      "Bad request for gene: AC003681.1. Check the gene name and try again.\n",
      "AC003682.1\n",
      "Bad request for gene: AC003682.1. Check the gene name and try again.\n",
      "AC003684.1\n",
      "Bad request for gene: AC003684.1. Check the gene name and try again.\n",
      "AC003965.2\n",
      "Bad request for gene: AC003965.2. Check the gene name and try again.\n",
      "AC003991.1\n",
      "Bad request for gene: AC003991.1. Check the gene name and try again.\n",
      "AC003991.2\n",
      "Bad request for gene: AC003991.2. Check the gene name and try again.\n",
      "AC004000.1\n",
      "Bad request for gene: AC004000.1. Check the gene name and try again.\n",
      "AC004034.1\n",
      "Bad request for gene: AC004034.1. Check the gene name and try again.\n",
      "AC004039.1\n",
      "Bad request for gene: AC004039.1. Check the gene name and try again.\n",
      "AC004053.1\n",
      "Bad request for gene: AC004053.1. Check the gene name and try again.\n",
      "AC004067.1\n",
      "Bad request for gene: AC004067.1. Check the gene name and try again.\n",
      "AC004069.1\n",
      "Bad request for gene: AC004069.1. Check the gene name and try again.\n",
      "AC004076.2\n",
      "Bad request for gene: AC004076.2. Check the gene name and try again.\n",
      "AC004083.1\n",
      "Bad request for gene: AC004083.1. Check the gene name and try again.\n",
      "AC004112.1\n",
      "Bad request for gene: AC004112.1. Check the gene name and try again.\n",
      "AC004148.1\n",
      "Bad request for gene: AC004148.1. Check the gene name and try again.\n",
      "AC004158.1\n",
      "Bad request for gene: AC004158.1. Check the gene name and try again.\n",
      "AC004160.1\n",
      "Bad request for gene: AC004160.1. Check the gene name and try again.\n",
      "AC004217.1\n",
      "Bad request for gene: AC004217.1. Check the gene name and try again.\n",
      "AC004223.2\n",
      "Bad request for gene: AC004223.2. Check the gene name and try again.\n",
      "AC004223.4\n",
      "Bad request for gene: AC004223.4. Check the gene name and try again.\n",
      "AC004231.1\n",
      "Bad request for gene: AC004231.1. Check the gene name and try again.\n",
      "AC004232.1\n",
      "Bad request for gene: AC004232.1. Check the gene name and try again.\n",
      "AC004233.3\n",
      "Bad request for gene: AC004233.3. Check the gene name and try again.\n",
      "AC004241.1\n",
      "Bad request for gene: AC004241.1. Check the gene name and try again.\n",
      "AC004241.3\n",
      "Bad request for gene: AC004241.3. Check the gene name and try again.\n",
      "AC004253.1\n",
      "Bad request for gene: AC004253.1. Check the gene name and try again.\n",
      "AC004257.1\n",
      "Bad request for gene: AC004257.1. Check the gene name and try again.\n",
      "AC004263.2\n",
      "Bad request for gene: AC004263.2. Check the gene name and try again.\n",
      "AC004381.1\n",
      "Bad request for gene: AC004381.1. Check the gene name and try again.\n",
      "AC004448.2\n",
      "Bad request for gene: AC004448.2. Check the gene name and try again.\n",
      "AC004461.2\n",
      "Bad request for gene: AC004461.2. Check the gene name and try again.\n",
      "AC004466.1\n",
      "Bad request for gene: AC004466.1. Check the gene name and try again.\n",
      "AC004466.2\n",
      "Bad request for gene: AC004466.2. Check the gene name and try again.\n",
      "AC004471.1\n",
      "Bad request for gene: AC004471.1. Check the gene name and try again.\n",
      "AC004477.1\n",
      "Bad request for gene: AC004477.1. Check the gene name and try again.\n",
      "AC004485.1\n",
      "Bad request for gene: AC004485.1. Check the gene name and try again.\n",
      "AC004490.1\n",
      "Bad request for gene: AC004490.1. Check the gene name and try again.\n",
      "AC004492.1\n",
      "Bad request for gene: AC004492.1. Check the gene name and try again.\n",
      "AC004494.1\n",
      "Bad request for gene: AC004494.1. Check the gene name and try again.\n",
      "AC004522.5\n",
      "Bad request for gene: AC004522.5. Check the gene name and try again.\n",
      "AC004528.1\n",
      "Bad request for gene: AC004528.1. Check the gene name and try again.\n",
      "AC004528.2\n",
      "Bad request for gene: AC004528.2. Check the gene name and try again.\n",
      "AC004540.1\n",
      "Bad request for gene: AC004540.1. Check the gene name and try again.\n",
      "AC004540.2\n",
      "Bad request for gene: AC004540.2. Check the gene name and try again.\n",
      "AC004542.2\n",
      "Bad request for gene: AC004542.2. Check the gene name and try again.\n",
      "AC004551.1\n",
      "Bad request for gene: AC004551.1. Check the gene name and try again.\n",
      "AC004554.2\n",
      "Bad request for gene: AC004554.2. Check the gene name and try again.\n",
      "AC004556.3\n",
      "Bad request for gene: AC004556.3. Check the gene name and try again.\n",
      "AC004584.1\n",
      "Bad request for gene: AC004584.1. Check the gene name and try again.\n",
      "AC004584.3\n",
      "Bad request for gene: AC004584.3. Check the gene name and try again.\n",
      "AC004585.1\n",
      "Bad request for gene: AC004585.1. Check the gene name and try again.\n",
      "AC004593.1\n",
      "Bad request for gene: AC004593.1. Check the gene name and try again.\n",
      "AC004593.3\n",
      "Bad request for gene: AC004593.3. Check the gene name and try again.\n",
      "AC004594.1\n",
      "Bad request for gene: AC004594.1. Check the gene name and try again.\n",
      "AC004596.1\n",
      "Bad request for gene: AC004596.1. Check the gene name and try again.\n",
      "AC004672.1\n",
      "Bad request for gene: AC004672.1. Check the gene name and try again.\n",
      "AC004687.1\n",
      "Bad request for gene: AC004687.1. Check the gene name and try again.\n",
      "AC004702.1\n",
      "Bad request for gene: AC004702.1. Check the gene name and try again.\n",
      "AC004704.1\n",
      "Bad request for gene: AC004704.1. Check the gene name and try again.\n",
      "AC004706.1\n",
      "Bad request for gene: AC004706.1. Check the gene name and try again.\n",
      "AC004765.1\n",
      "Bad request for gene: AC004765.1. Check the gene name and try again.\n",
      "AC004771.1\n",
      "Bad request for gene: AC004771.1. Check the gene name and try again.\n",
      "AC004771.3\n",
      "Bad request for gene: AC004771.3. Check the gene name and try again.\n",
      "AC004771.4\n",
      "Bad request for gene: AC004771.4. Check the gene name and try again.\n",
      "AC004784.1\n",
      "Bad request for gene: AC004784.1. Check the gene name and try again.\n",
      "AC004797.1\n",
      "Bad request for gene: AC004797.1. Check the gene name and try again.\n",
      "AC004801.4\n",
      "Bad request for gene: AC004801.4. Check the gene name and try again.\n",
      "AC004801.6\n",
      "Bad request for gene: AC004801.6. Check the gene name and try again.\n",
      "AC004803.1\n",
      "Bad request for gene: AC004803.1. Check the gene name and try again.\n",
      "AC004812.2\n",
      "Bad request for gene: AC004812.2. Check the gene name and try again.\n",
      "AC004816.1\n",
      "Bad request for gene: AC004816.1. Check the gene name and try again.\n",
      "AC004817.3\n",
      "Bad request for gene: AC004817.3. Check the gene name and try again.\n",
      "AC004817.5\n",
      "Bad request for gene: AC004817.5. Check the gene name and try again.\n",
      "AC004825.2\n",
      "Bad request for gene: AC004825.2. Check the gene name and try again.\n",
      "AC004825.3\n",
      "Bad request for gene: AC004825.3. Check the gene name and try again.\n",
      "AC004828.2\n",
      "Bad request for gene: AC004828.2. Check the gene name and try again.\n",
      "AC004830.2\n",
      "Bad request for gene: AC004830.2. Check the gene name and try again.\n",
      "AC004832.4\n",
      "Bad request for gene: AC004832.4. Check the gene name and try again.\n",
      "AC004834.1\n",
      "Bad request for gene: AC004834.1. Check the gene name and try again.\n",
      "AC004837.2\n",
      "Bad request for gene: AC004837.2. Check the gene name and try again.\n",
      "AC004837.4\n",
      "Bad request for gene: AC004837.4. Check the gene name and try again.\n",
      "AC004839.2\n",
      "Bad request for gene: AC004839.2. Check the gene name and try again.\n",
      "AC004846.1\n",
      "Bad request for gene: AC004846.1. Check the gene name and try again.\n",
      "AC004846.2\n",
      "Bad request for gene: AC004846.2. Check the gene name and try again.\n",
      "AC004847.1\n",
      "Bad request for gene: AC004847.1. Check the gene name and try again.\n",
      "AC004854.2\n",
      "Bad request for gene: AC004854.2. Check the gene name and try again.\n",
      "AC004865.2\n",
      "Bad request for gene: AC004865.2. Check the gene name and try again.\n",
      "AC004877.1\n",
      "Bad request for gene: AC004877.1. Check the gene name and try again.\n",
      "AC004884.2\n",
      "Bad request for gene: AC004884.2. Check the gene name and try again.\n",
      "AC004889.1\n",
      "Bad request for gene: AC004889.1. Check the gene name and try again.\n",
      "AC004893.3\n",
      "Bad request for gene: AC004893.3. Check the gene name and try again.\n",
      "AC004900.1\n",
      "Bad request for gene: AC004900.1. Check the gene name and try again.\n",
      "AC004906.1\n",
      "Bad request for gene: AC004906.1. Check the gene name and try again.\n",
      "AC004908.2\n",
      "Bad request for gene: AC004908.2. Check the gene name and try again.\n",
      "AC004908.3\n",
      "Bad request for gene: AC004908.3. Check the gene name and try again.\n",
      "AC004917.1\n",
      "Bad request for gene: AC004917.1. Check the gene name and try again.\n",
      "AC004918.1\n",
      "Bad request for gene: AC004918.1. Check the gene name and try again.\n",
      "AC004918.4\n",
      "Bad request for gene: AC004918.4. Check the gene name and try again.\n",
      "AC004921.1\n",
      "Bad request for gene: AC004921.1. Check the gene name and try again.\n",
      "AC004921.2\n",
      "Bad request for gene: AC004921.2. Check the gene name and try again.\n",
      "AC004923.4\n",
      "Bad request for gene: AC004923.4. Check the gene name and try again.\n",
      "AC004938.2\n",
      "Bad request for gene: AC004938.2. Check the gene name and try again.\n",
      "AC004943.2\n",
      "Bad request for gene: AC004943.2. Check the gene name and try again.\n",
      "AC004947.3\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 25\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124minterval\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m adata_rna\u001b[38;5;241m.\u001b[39mvar:\n\u001b[1;32m     24\u001b[0m     genenames \u001b[38;5;241m=\u001b[39m adata_rna_s\u001b[38;5;241m.\u001b[39mvar_names\n\u001b[0;32m---> 25\u001b[0m     location \u001b[38;5;241m=\u001b[39m [get_gene_location(gene_name) \u001b[38;5;28;01mfor\u001b[39;00m gene_name \u001b[38;5;129;01min\u001b[39;00m genenames]\n\u001b[1;32m     26\u001b[0m     adata_rna_s\u001b[38;5;241m.\u001b[39mvar[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124minterval\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m location\n\u001b[1;32m     27\u001b[0m chr_mask \u001b[38;5;241m=\u001b[39m adata_rna_s\u001b[38;5;241m.\u001b[39mvar\u001b[38;5;241m.\u001b[39minterval\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m:\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mstr[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mchr1\u001b[39m\u001b[38;5;124m'\u001b[39m\n",
      "Cell \u001b[0;32mIn[3], line 25\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124minterval\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m adata_rna\u001b[38;5;241m.\u001b[39mvar:\n\u001b[1;32m     24\u001b[0m     genenames \u001b[38;5;241m=\u001b[39m adata_rna_s\u001b[38;5;241m.\u001b[39mvar_names\n\u001b[0;32m---> 25\u001b[0m     location \u001b[38;5;241m=\u001b[39m [\u001b[43mget_gene_location\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene_name\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m gene_name \u001b[38;5;129;01min\u001b[39;00m genenames]\n\u001b[1;32m     26\u001b[0m     adata_rna_s\u001b[38;5;241m.\u001b[39mvar[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124minterval\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m location\n\u001b[1;32m     27\u001b[0m chr_mask \u001b[38;5;241m=\u001b[39m adata_rna_s\u001b[38;5;241m.\u001b[39mvar\u001b[38;5;241m.\u001b[39minterval\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m:\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mstr[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mchr1\u001b[39m\u001b[38;5;124m'\u001b[39m\n",
      "Cell \u001b[0;32mIn[3], line 6\u001b[0m, in \u001b[0;36mget_gene_location\u001b[0;34m(gene_name)\u001b[0m\n\u001b[1;32m      4\u001b[0m server \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttps://rest.ensembl.org\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      5\u001b[0m ext \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/lookup/symbol/homo_sapiens/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mgene_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m?content-type=application/json\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m----> 6\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mserver\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mext\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m r\u001b[38;5;241m.\u001b[39mstatus_code \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m400\u001b[39m:\n\u001b[1;32m      8\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBad request for gene: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mgene_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m. Check the gene name and try again.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/py10/lib/python3.10/site-packages/requests/api.py:73\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(url, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     63\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \n\u001b[1;32m     65\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 73\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/py10/lib/python3.10/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/py10/lib/python3.10/site-packages/requests/sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    584\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    585\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m: timeout,\n\u001b[1;32m    586\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    587\u001b[0m }\n\u001b[1;32m    588\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 589\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/anaconda3/envs/py10/lib/python3.10/site-packages/requests/sessions.py:703\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    700\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    702\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    705\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    706\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m~/anaconda3/envs/py10/lib/python3.10/site-packages/requests/adapters.py:486\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    483\u001b[0m     timeout \u001b[38;5;241m=\u001b[39m TimeoutSauce(connect\u001b[38;5;241m=\u001b[39mtimeout, read\u001b[38;5;241m=\u001b[39mtimeout)\n\u001b[1;32m    485\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 486\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    487\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    488\u001b[0m \u001b[43m        \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    489\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    490\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    491\u001b[0m \u001b[43m        \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    492\u001b[0m \u001b[43m        \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    493\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    495\u001b[0m \u001b[43m        \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    496\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    497\u001b[0m \u001b[43m        \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    498\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    500\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ProtocolError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    501\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(err, request\u001b[38;5;241m=\u001b[39mrequest)\n",
      "File \u001b[0;32m~/anaconda3/envs/py10/lib/python3.10/site-packages/urllib3/connectionpool.py:789\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, preload_content, decode_content, **response_kw)\u001b[0m\n\u001b[1;32m    786\u001b[0m response_conn \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m release_conn \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    788\u001b[0m \u001b[38;5;66;03m# Make the request on the HTTPConnection object\u001b[39;00m\n\u001b[0;32m--> 789\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    790\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    791\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    792\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    793\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    794\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    795\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    796\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    797\u001b[0m \u001b[43m    \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    798\u001b[0m \u001b[43m    \u001b[49m\u001b[43mresponse_conn\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mresponse_conn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    799\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpreload_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    800\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdecode_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    801\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mresponse_kw\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    802\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    804\u001b[0m \u001b[38;5;66;03m# Everything went great!\u001b[39;00m\n\u001b[1;32m    805\u001b[0m clean_exit \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/py10/lib/python3.10/site-packages/urllib3/connectionpool.py:466\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, body, headers, retries, timeout, chunked, response_conn, preload_content, decode_content, enforce_content_length)\u001b[0m\n\u001b[1;32m    463\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    464\u001b[0m     \u001b[38;5;66;03m# Trigger any extra validation we need to do.\u001b[39;00m\n\u001b[1;32m    465\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 466\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    467\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    468\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_raise_timeout(err\u001b[38;5;241m=\u001b[39me, url\u001b[38;5;241m=\u001b[39murl, timeout_value\u001b[38;5;241m=\u001b[39mconn\u001b[38;5;241m.\u001b[39mtimeout)\n",
      "File \u001b[0;32m~/anaconda3/envs/py10/lib/python3.10/site-packages/urllib3/connectionpool.py:1095\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m   1093\u001b[0m \u001b[38;5;66;03m# Force connect early to allow us to validate the connection.\u001b[39;00m\n\u001b[1;32m   1094\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m conn\u001b[38;5;241m.\u001b[39mis_closed:\n\u001b[0;32m-> 1095\u001b[0m     \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1097\u001b[0m \u001b[38;5;66;03m# TODO revise this, see https://github.com/urllib3/urllib3/issues/2791\u001b[39;00m\n\u001b[1;32m   1098\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m conn\u001b[38;5;241m.\u001b[39mis_verified \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m conn\u001b[38;5;241m.\u001b[39mproxy_is_verified:\n",
      "File \u001b[0;32m~/anaconda3/envs/py10/lib/python3.10/site-packages/urllib3/connection.py:615\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    613\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mconnect\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    614\u001b[0m     sock: socket\u001b[38;5;241m.\u001b[39msocket \u001b[38;5;241m|\u001b[39m ssl\u001b[38;5;241m.\u001b[39mSSLSocket\n\u001b[0;32m--> 615\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock \u001b[38;5;241m=\u001b[39m sock \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_new_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    616\u001b[0m     server_hostname: \u001b[38;5;28mstr\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost\n\u001b[1;32m    617\u001b[0m     tls_in_tls \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/py10/lib/python3.10/site-packages/urllib3/connection.py:196\u001b[0m, in \u001b[0;36mHTTPConnection._new_conn\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    191\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Establish a socket connection and set nodelay settings on it.\u001b[39;00m\n\u001b[1;32m    192\u001b[0m \n\u001b[1;32m    193\u001b[0m \u001b[38;5;124;03m:return: New socket connection.\u001b[39;00m\n\u001b[1;32m    194\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    195\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 196\u001b[0m     sock \u001b[38;5;241m=\u001b[39m \u001b[43mconnection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate_connection\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    197\u001b[0m \u001b[43m        \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_dns_host\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mport\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    198\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    199\u001b[0m \u001b[43m        \u001b[49m\u001b[43msource_address\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msource_address\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    200\u001b[0m \u001b[43m        \u001b[49m\u001b[43msocket_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msocket_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    201\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    202\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m socket\u001b[38;5;241m.\u001b[39mgaierror \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    203\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m NameResolutionError(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost, \u001b[38;5;28mself\u001b[39m, e) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01me\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/py10/lib/python3.10/site-packages/urllib3/util/connection.py:60\u001b[0m, in \u001b[0;36mcreate_connection\u001b[0;34m(address, timeout, source_address, socket_options)\u001b[0m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mUnicodeError\u001b[39;00m:\n\u001b[1;32m     58\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m LocationParseError(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mhost\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, label empty or too long\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m---> 60\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m \u001b[43msocket\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetaddrinfo\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhost\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mport\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfamily\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msocket\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mSOCK_STREAM\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m     61\u001b[0m     af, socktype, proto, canonname, sa \u001b[38;5;241m=\u001b[39m res\n\u001b[1;32m     62\u001b[0m     sock \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/py10/lib/python3.10/socket.py:955\u001b[0m, in \u001b[0;36mgetaddrinfo\u001b[0;34m(host, port, family, type, proto, flags)\u001b[0m\n\u001b[1;32m    952\u001b[0m \u001b[38;5;66;03m# We override this function since we want to translate the numeric family\u001b[39;00m\n\u001b[1;32m    953\u001b[0m \u001b[38;5;66;03m# and socket type values to enum constants.\u001b[39;00m\n\u001b[1;32m    954\u001b[0m addrlist \u001b[38;5;241m=\u001b[39m []\n\u001b[0;32m--> 955\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m \u001b[43m_socket\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetaddrinfo\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhost\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mport\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfamily\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mtype\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mproto\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m    956\u001b[0m     af, socktype, proto, canonname, sa \u001b[38;5;241m=\u001b[39m res\n\u001b[1;32m    957\u001b[0m     addrlist\u001b[38;5;241m.\u001b[39mappend((_intenum_converter(af, AddressFamily),\n\u001b[1;32m    958\u001b[0m                      _intenum_converter(socktype, SocketKind),\n\u001b[1;32m    959\u001b[0m                      proto, canonname, sa))\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# shorten rna \n",
    "mask = adata_rna.obs.donor_id=='donor_0'\n",
    "adata_rna_s = adata_rna[mask]\n",
    "# only one chr and n_cells \n",
    "if 'interval' not in adata_rna.var:\n",
    "    genenames = adata_rna_s.var_names\n",
    "    location = [get_gene_location(gene_name) for gene_name in genenames]\n",
    "    adata_rna_s.var['interval'] = location\n",
    "chr_mask = adata_rna_s.var.interval.str.split(':').str[0] == 'chr1'\n",
    "adata_rna_s = adata_rna_s[:n_cells, chr_mask]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prior"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## tf names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "prior_adata = ad.AnnData()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tfs list \n",
    "tf_list = np.loadtxt(\"https://resources.aertslab.org/cistarget/tf_lists/allTFs_hg38.txt\", dtype=str)\n",
    "prior_adata.uns['tf_list'] = tf_list\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "bulk_adata = ad.read_h5ad(f'{resource_dir}/perturbation_data.h5ad')\n",
    "prior_adata.uns['gene_names_pert'] = bulk_adata.var_names.to_numpy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "bulk_adata = ad.read_h5ad(f'{resource_dir}/multiomics_rna.h5ad')\n",
    "prior_adata.uns['gene_names'] = bulk_adata.var_names.to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "bulk_adata = ad.read_h5ad(f'{resource_dir}/multiomics_atac.h5ad')\n",
    "prior_adata.uns['peak'] = bulk_adata.var_names.to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "prior_adata.uns['hvgs'] = np.loadtxt(f'{resource_dir}/hvgs.txt', dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "prior_adata.write(f'{resource_dir}/prior_data.h5ad')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savetxt(f'{work_dir}/benchmark/perturb_gene_names.txt', bulk_adata.var_names.values  , fmt='%s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "url = \"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_46/gencode.v46.annotation.gtf.gz\"\n",
    "local_filename = \"gencode.v46.annotation.gtf.gz\"\n",
    "\n",
    "with requests.get(url, stream=True) as r:\n",
    "    r.raise_for_status()\n",
    "    with open(local_filename, 'wb') as f:\n",
    "        for chunk in r.iter_content(chunk_size=8192):\n",
    "            f.write(chunk)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HVGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_hvgs = 3000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/anaconda3/envs/py10/lib/python3.10/site-packages/rpy2/rinterface.py:186: FutureWarning: Index.format is deprecated and will be removed in a future version. Convert using index.astype(str) or index.map(formatter) instead.\n",
      "  openrlib.rlib.R_tryEval(\n",
      "/root/.cache/R/basilisk/1.14.3/zellkonverter/1.12.1/zellkonverterAnnDataEnv-0.10.2/lib/python3.11/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Registered S3 methods overwritten by 'zellkonverter':\n",
       "  method                                             from      \n",
       "  py_to_r.numpy.ndarray                              reticulate\n",
       "  py_to_r.pandas.core.arrays.categorical.Categorical reticulate\n",
       "Loading required package: SummarizedExperiment\n",
       "Loading required package: MatrixGenerics\n",
       "Loading required package: matrixStats\n",
       "\n",
       "Attaching package: â€˜MatrixGenericsâ€™\n",
       "\n",
       "The following objects are masked from â€˜package:matrixStatsâ€™:\n",
       "\n",
       "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
       "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
       "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
       "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
       "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
       "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
       "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
       "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
       "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
       "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
       "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
       "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
       "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
       "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
       "    rowWeightedSds, rowWeightedVars\n",
       "\n",
       "Loading required package: GenomicRanges\n",
       "Loading required package: stats4\n",
       "Loading required package: BiocGenerics\n",
       "\n",
       "Attaching package: â€˜BiocGenericsâ€™\n",
       "\n",
       "The following objects are masked from â€˜package:statsâ€™:\n",
       "\n",
       "    IQR, mad, sd, var, xtabs\n",
       "\n",
       "The following objects are masked from â€˜package:baseâ€™:\n",
       "\n",
       "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
       "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
       "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
       "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
       "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
       "    table, tapply, union, unique, unsplit, which.max, which.min\n",
       "\n",
       "Loading required package: S4Vectors\n",
       "\n",
       "Attaching package: â€˜S4Vectorsâ€™\n",
       "\n",
       "The following object is masked from â€˜package:utilsâ€™:\n",
       "\n",
       "    findMatches\n",
       "\n",
       "The following objects are masked from â€˜package:baseâ€™:\n",
       "\n",
       "    expand.grid, I, unname\n",
       "\n",
       "Loading required package: IRanges\n",
       "Loading required package: GenomeInfoDb\n",
       "Loading required package: Biobase\n",
       "Welcome to Bioconductor\n",
       "\n",
       "    Vignettes contain introductory material; view with\n",
       "    'browseVignettes()'. To cite Bioconductor, see\n",
       "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
       "\n",
       "\n",
       "Attaching package: â€˜Biobaseâ€™\n",
       "\n",
       "The following object is masked from â€˜package:MatrixGenericsâ€™:\n",
       "\n",
       "    rowMedians\n",
       "\n",
       "The following objects are masked from â€˜package:matrixStatsâ€™:\n",
       "\n",
       "    anyMissing, rowMedians\n",
       "\n",
       "In addition: Warning messages:\n",
       "1: In (function (package, help, pos = 2, lib.loc = NULL, character.only = FALSE,  :\n",
       "  library â€˜/usr/lib/R/site-libraryâ€™ contains no packages\n",
       "2: Python '/root/.cache/R/basilisk/1.14.3/zellkonverter/1.12.1/zellkonverterAnnDataEnv-0.10.2/bin/python' was requested but '/root/anaconda3/envs/py10/bin/python' was loaded instead (see reticulate::py_config() for more information) \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i work_dir\n",
    "library(scry)\n",
    "library(zellkonverter)\n",
    "library(SingleCellExperiment)\n",
    "options(digits=5, max.print=100)  # Adjust numbers as needed\n",
    "\n",
    "\n",
    "\n",
    "adata = readH5AD(paste0(work_dir, \"/preprocess/bulk_adata_f.h5ad\")) # raw counts\n",
    "sce = devianceFeatureSelection(adata, assay=\"X\", batch=colData(adata)$plate_name)\n",
    "writeH5AD(sce, paste0(work_dir, \"/preprocess/adata_sce.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/anaconda3/envs/py10/lib/python3.10/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "adata_sce = ad.read(f\"{work_dir}/preprocess/adata_sce.h5ad\")\n",
    "binomial_deviance  = adata_sce.var['binomial_deviance']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A1BG         3242.560255\n",
       "A1BG-AS1     3081.247245\n",
       "A2M         51924.570033\n",
       "A2M-AS1     47137.741205\n",
       "A2MP1        4798.622262\n",
       "                ...     \n",
       "ZXDB         3123.833965\n",
       "ZXDC         2762.431522\n",
       "ZYG11B      14322.309802\n",
       "ZYX         60193.206178\n",
       "ZZEF1        5337.865927\n",
       "Name: binomial_deviance, Length: 15215, dtype: float64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sce.var['binomial_deviance']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "indices = binomial_deviance.argsort()[-n_hvgs:]\n",
    "mask = np.zeros(adata_sce.var.shape[0], dtype=bool)\n",
    "mask[indices] = True\n",
    "hvgs_sce = adata_sce[:, mask].var.index.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savetxt( f'{work_dir}/benchmark/hvgs.txt',hvgs_sce, fmt='%s')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AC007325.4</td>\n",
       "      <td>Protein coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AC107375.1</td>\n",
       "      <td>LncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AC022306.3</td>\n",
       "      <td>LncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALDOC</td>\n",
       "      <td>Protein coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HEMK1</td>\n",
       "      <td>Protein coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene Transcript_type\n",
       "0  AC007325.4  Protein coding\n",
       "1  AC107375.1          LncRNA\n",
       "2  AC022306.3          LncRNA\n",
       "3       ALDOC  Protein coding\n",
       "4       HEMK1  Protein coding"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load the annotations and subset it for genes in multiomics data. also, remove\n",
    "from local_utils import annotation\n",
    "\n",
    "multiomics_genes =  np.loadtxt(f'{work_dir}/benchmark/multiomics_genes.txt', dtype=str)\n",
    "annot_database = annotation.ensembl_gene_annotation()\n",
    "\n",
    "transcript_types = ['protein_coding', 'lncRNA', 'miRNA']\n",
    "\n",
    "annot_database_f = annot_database[annot_database.Gene.isin(multiomics_genes)].reset_index()\n",
    "\n",
    "annot_database_f = annot_database_f[['Gene', 'Transcript_type']].drop_duplicates()\n",
    "annot_database_f = annot_database_f[annot_database_f.Transcript_type.isin(transcript_types)]\n",
    "rename_map = {'protein_coding':'Protein coding', 'lncRNA':'LncRNA',  'miRNA':'MicroRNA'}\n",
    "annot_database_f.Transcript_type = annot_database_f.Transcript_type.map(rename_map)\n",
    "annot_database_f.reset_index(drop=True, inplace=True)\n",
    "annot_database_f.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(22139, 2)"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " #only keep genes with one annotation\n",
    "annot_size = annot_database_f.groupby('Gene').size()\n",
    "kept_genes = annot_size[~(annot_size>1)].index\n",
    "annot_database_f = annot_database_f[annot_database_f.Gene.isin(kept_genes)].reset_index(drop=True)\n",
    "annot_database_f.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Transcript_type\n",
       "Protein coding    15282\n",
       "LncRNA             6856\n",
       "MicroRNA              1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_database_f.Transcript_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save \n",
    "annot_database_f.to_csv(f'{work_dir}/benchmark/gene_annotation.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## peak annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/anaconda3/envs/py10/lib/python3.10/site-packages/anndata/_core/aligned_df.py:67: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    }
   ],
   "source": [
    "adata_atac = ad.read_h5ad(f'{work_dir}/scATAC/adata_atac.h5ad')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## check if all the peaks in grns given in atac data\n",
    "import re\n",
    "\n",
    "def format_peak(peaks):\n",
    "    formatted_peaks = []\n",
    "    for peak in peaks:\n",
    "        chr_, start, end = re.split(r'[:\\-_]', peak)\n",
    "        peak = f\"{chr_}:{start}-{end}\"\n",
    "\n",
    "        formatted_peaks.append(peak)\n",
    "    return formatted_peaks\n",
    "\n",
    "\n",
    "atac_peaks = format_peak(adata_atac.var_names)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100001032-100001800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100006075-100006963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100009475-100010367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100013993-100014884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100020278-100021136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135353</th>\n",
       "      <td>chrY</td>\n",
       "      <td>7765105-7765991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135354</th>\n",
       "      <td>chrY</td>\n",
       "      <td>7814158-7815060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135355</th>\n",
       "      <td>chrY</td>\n",
       "      <td>7818681-7819599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135356</th>\n",
       "      <td>chrY</td>\n",
       "      <td>8535565-8536421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135357</th>\n",
       "      <td>chrY</td>\n",
       "      <td>8537529-8538370</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>135358 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr                range\n",
       "0       chr10  100001032-100001800\n",
       "1       chr10  100006075-100006963\n",
       "2       chr10  100009475-100010367\n",
       "3       chr10  100013993-100014884\n",
       "4       chr10  100020278-100021136\n",
       "...       ...                  ...\n",
       "135353   chrY      7765105-7765991\n",
       "135354   chrY      7814158-7815060\n",
       "135355   chrY      7818681-7819599\n",
       "135356   chrY      8535565-8536421\n",
       "135357   chrY      8537529-8538370\n",
       "\n",
       "[135358 rows x 2 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks = pd.DataFrame({'chr':[peak.split(':')[0] for peak in atac_peaks],\n",
    "                                     'range':[peak.split(':')[1] for peak in atac_peaks]})\n",
    "peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">> preparing features information...\t\t 2024-06-18 14:17:45 \n",
      ">> identifying nearest features...\t\t 2024-06-18 14:17:46 \n",
      ">> calculating distance from peak to TSS...\t 2024-06-18 14:17:48 \n",
      ">> assigning genomic annotation...\t\t 2024-06-18 14:17:48 \n",
      ">> assigning chromosome lengths\t\t\t 2024-06-18 14:18:10 \n",
      ">> done...\t\t\t\t\t 2024-06-18 14:18:10 \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Loading required package: GenomicFeatures\n",
       "Loading required package: AnnotationDbi\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i peaks -o peaks_annotated_df\n",
    "options(digits=5, max.print=100)  # Adjust numbers as needed\n",
    "set.seed(123)\n",
    "\n",
    "# install.packages(\"IRanges\")\n",
    "# install.packages(\"GenomicRanges\")\n",
    "# install.packages(\"ggplot2\")\n",
    "# install.packages(\"TxDb.Hsapiens.UCSC.hg38.knownGene\")\n",
    "\n",
    "library(IRanges)\n",
    "library(GenomicRanges)\n",
    "library(ggplot2)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene\n",
    "\n",
    "\n",
    "peaks = GRanges(peaks$chr, IRanges(peaks$range))\n",
    "peaks_annotated = suppressMessages(ChIPseeker::annotatePeak(\n",
    "    peaks,\n",
    "    tssRegion = c(-1000, 1000), # extended from -5kb to 5\n",
    "    TxDb = txdb,\n",
    "    level = \"transcript\", \n",
    "    assignGenomicAnnotation = TRUE,  # the default\n",
    "    genomicAnnotationPriority = c(\"Promoter\", \"5UTR\", \"3UTR\", \"Exon\", \"Intron\",\n",
    "                                \"Downstream\", \"Intergenic\"),  # the default\n",
    "    annoDb = NULL,\n",
    "    sameStrand = FALSE, # the default\n",
    "    ignoreOverlap = FALSE, # the default\n",
    "    ignoreUpstream = FALSE, # the default\n",
    "    ignoreDownstream = FALSE, # the default\n",
    "    overlap = \"TSS\", # the default\n",
    "    verbose = TRUE # the default\n",
    "))\n",
    "peaks_annotated_df = as.data.frame(peaks_annotated)\n",
    "# write.table(peaks_annotated_df, paste0(temp_dir, name, '_annot.txt'), sep=',', row.names = FALSE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>geneChr</th>\n",
       "      <th>geneStart</th>\n",
       "      <th>geneEnd</th>\n",
       "      <th>geneLength</th>\n",
       "      <th>geneStrand</th>\n",
       "      <th>geneId</th>\n",
       "      <th>transcriptId</th>\n",
       "      <th>distanceToTSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100001032</td>\n",
       "      <td>100001800</td>\n",
       "      <td>769</td>\n",
       "      <td>*</td>\n",
       "      <td>Intron (ENST00000324109.9/23268, intron 1 of 16)</td>\n",
       "      <td>10</td>\n",
       "      <td>99875577</td>\n",
       "      <td>100009947</td>\n",
       "      <td>134371</td>\n",
       "      <td>2</td>\n",
       "      <td>23268</td>\n",
       "      <td>ENST00000324109.9</td>\n",
       "      <td>8147.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100006075</td>\n",
       "      <td>100006963</td>\n",
       "      <td>889</td>\n",
       "      <td>*</td>\n",
       "      <td>Intron (ENST00000324109.9/23268, intron 1 of 16)</td>\n",
       "      <td>10</td>\n",
       "      <td>99875577</td>\n",
       "      <td>100009947</td>\n",
       "      <td>134371</td>\n",
       "      <td>2</td>\n",
       "      <td>23268</td>\n",
       "      <td>ENST00000324109.9</td>\n",
       "      <td>2984.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100009475</td>\n",
       "      <td>100010367</td>\n",
       "      <td>893</td>\n",
       "      <td>*</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>10</td>\n",
       "      <td>99875577</td>\n",
       "      <td>100009947</td>\n",
       "      <td>134371</td>\n",
       "      <td>2</td>\n",
       "      <td>23268</td>\n",
       "      <td>ENST00000324109.9</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100013993</td>\n",
       "      <td>100014884</td>\n",
       "      <td>892</td>\n",
       "      <td>*</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>10</td>\n",
       "      <td>99875577</td>\n",
       "      <td>100009947</td>\n",
       "      <td>134371</td>\n",
       "      <td>2</td>\n",
       "      <td>23268</td>\n",
       "      <td>ENST00000324109.9</td>\n",
       "      <td>-4046.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr10</td>\n",
       "      <td>100020278</td>\n",
       "      <td>100021136</td>\n",
       "      <td>859</td>\n",
       "      <td>*</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>10</td>\n",
       "      <td>99875577</td>\n",
       "      <td>100009947</td>\n",
       "      <td>134371</td>\n",
       "      <td>2</td>\n",
       "      <td>23268</td>\n",
       "      <td>ENST00000324109.9</td>\n",
       "      <td>-10331.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqnames      start        end  width strand  \\\n",
       "1    chr10  100001032  100001800    769      *   \n",
       "2    chr10  100006075  100006963    889      *   \n",
       "3    chr10  100009475  100010367    893      *   \n",
       "4    chr10  100013993  100014884    892      *   \n",
       "5    chr10  100020278  100021136    859      *   \n",
       "\n",
       "                                         annotation  geneChr  geneStart  \\\n",
       "1  Intron (ENST00000324109.9/23268, intron 1 of 16)       10   99875577   \n",
       "2  Intron (ENST00000324109.9/23268, intron 1 of 16)       10   99875577   \n",
       "3                                          Promoter       10   99875577   \n",
       "4                                 Distal Intergenic       10   99875577   \n",
       "5                                 Distal Intergenic       10   99875577   \n",
       "\n",
       "     geneEnd  geneLength  geneStrand geneId       transcriptId  distanceToTSS  \n",
       "1  100009947      134371           2  23268  ENST00000324109.9         8147.0  \n",
       "2  100009947      134371           2  23268  ENST00000324109.9         2984.0  \n",
       "3  100009947      134371           2  23268  ENST00000324109.9            0.0  \n",
       "4  100009947      134371           2  23268  ENST00000324109.9        -4046.0  \n",
       "5  100009947      134371           2  23268  ENST00000324109.9       -10331.0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_annotated_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "map_={'Intron':'Intron', 'Exon':'Exon', 'Promoter':'Promoter', 'Distal':'Distal Intergenic', \"3'\":\"3' UTR\", 'Downstream':'Downstream (<=300)', \"5'\":\"5' UTR\"}\n",
    "\n",
    "ann = peaks_annotated_df.annotation.str.split(' ', expand=True)[0]\n",
    "ann = ann.map(map_)\n",
    "peaks = peaks_annotated_df['seqnames'].astype(str)+':'+peaks_annotated_df['start'].astype(str) +'-' + peaks_annotated_df['end'].astype(str)\n",
    "df = pd.DataFrame({'annotation':ann, 'peak':peaks})\n",
    "df.to_csv(f'{resource_dir}/benchmark/peak_annotation.csv')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
