{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Context specific (static) gene regulatory network (GRN) inference\n",
    "This notebooks runs the context specific GRN inference pipeline. It uses the input files prepared in `data` folder and saves context specific GRNs into a single file `output/static.h5`.\n",
    "\n",
    "The pipeline consists of several parts for platform (CPU of GPU) dependent optimization of computation speed. Each part is organized into two sections in this notebook:\n",
    "1. The execution section runs this part for network inference. You can see the commands and the output of each command.\n",
    "2. The command description section explains each command involved in this part.\n",
    "\n",
    "Depending on your mode of computation for pytorch (defined by `DEVICE` in `makefiles/config.mk`), each part may include more steps or should be totally skipped. **Read the instructions of each part carefully.** If you cannot find the description of a command in one part, search other parts.\n",
    "\n",
    "**You should change** `-j 32` option for `make` in every part to the number of parallel processes suitable for your machine. **It should be different** for CPU and GPU parts. Here the maximum number of cores used is `32*NTH=128`. `NTH` is set in `makefiles/config.mk`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.1 CPU part 1, execution\n",
    "Here GPU is used to speed up computation. Therefore this part only infers the TF binding network that serves as a constraint for GRN inference.\n",
    "\n",
    "If you encounter errors from `make` like `No rule to make target...` or `Target 'cpu' not remade because of errors`, they can be safely ignored because these targets will be produced in CPU part 2. If you see other errors especially in peak or footprint detection steps due to low cell count for select cell subsets (typically several hundreds or lower), these errors can also be ignored and these cell subsets will be removed in the reconstructed networks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Removes CPU usage limit by some jupyter versions\n",
    "import os\n",
    "os.environ['KMP_AFFINITY'] = ''\n"
   ] 
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir -p tmp_static/Subset1/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "make: *** No rule to make target `tmp_static/Subset1/net_weight.tsv.gz', needed by `tmp_static/Subset1/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset1/net_meanvar.tsv.gz', needed by `tmp_static/Subset1/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset1/net_covfactor.tsv.gz', needed by `tmp_static/Subset1/net_nweight.tsv.gz'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cp data/subsets/Subset1/names_rna.txt tmp_static/Subset1/names_rna.txt\n",
      "mkdir -p tmp_static/Subset10/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "make: *** No rule to make target `tmp_static/Subset10/net_weight.tsv.gz', needed by `tmp_static/Subset10/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset10/net_meanvar.tsv.gz', needed by `tmp_static/Subset10/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset10/net_covfactor.tsv.gz', needed by `tmp_static/Subset10/net_nweight.tsv.gz'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir -p tmp_static/Subset11/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "make: *** No rule to make target `tmp_static/Subset11/net_weight.tsv.gz', needed by `tmp_static/Subset11/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset11/net_meanvar.tsv.gz', needed by `tmp_static/Subset11/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset11/net_covfactor.tsv.gz', needed by `tmp_static/Subset11/net_nweight.tsv.gz'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cp data/subsets/Subset10/names_rna.txt tmp_static/Subset10/names_rna.txt\n",
      "cp data/subsets/Subset11/names_rna.txt tmp_static/Subset11/names_rna.txt\n",
      "mkdir -p tmp_static/Subset12/\n",
      "cp data/subsets/Subset12/names_rna.txt tmp_static/Subset12/names_rna.txt\n",
      "mkdir -p tmp_static/Subset13/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "make: *** No rule to make target `tmp_static/Subset12/net_weight.tsv.gz', needed by `tmp_static/Subset12/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset12/net_meanvar.tsv.gz', needed by `tmp_static/Subset12/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset12/net_covfactor.tsv.gz', needed by `tmp_static/Subset12/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset13/net_weight.tsv.gz', needed by `tmp_static/Subset13/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset13/net_meanvar.tsv.gz', needed by `tmp_static/Subset13/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset13/net_covfactor.tsv.gz', needed by `tmp_static/Subset13/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset14/net_weight.tsv.gz', needed by `tmp_static/Subset14/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset14/net_meanvar.tsv.gz', needed by `tmp_static/Subset14/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset14/net_covfactor.tsv.gz', needed by `tmp_static/Subset14/net_nweight.tsv.gz'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cp data/subsets/Subset13/names_rna.txt tmp_static/Subset13/names_rna.txt\n",
      "mkdir -p tmp_static/Subset14/\n",
      "cp data/subsets/Subset14/names_rna.txt tmp_static/Subset14/names_rna.txt\n",
      "mkdir -p tmp_static/Subset2/\n",
      "cp data/subsets/Subset2/names_rna.txt tmp_static/Subset2/names_rna.txt\n",
      "mkdir -p tmp_static/Subset3/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "make: *** No rule to make target `tmp_static/Subset2/net_weight.tsv.gz', needed by `tmp_static/Subset2/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset2/net_meanvar.tsv.gz', needed by `tmp_static/Subset2/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset2/net_covfactor.tsv.gz', needed by `tmp_static/Subset2/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset3/net_weight.tsv.gz', needed by `tmp_static/Subset3/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset3/net_meanvar.tsv.gz', needed by `tmp_static/Subset3/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset3/net_covfactor.tsv.gz', needed by `tmp_static/Subset3/net_nweight.tsv.gz'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cp data/subsets/Subset3/names_rna.txt tmp_static/Subset3/names_rna.txt\n",
      "mkdir -p tmp_static/Subset4/\n",
      "cp data/subsets/Subset4/names_rna.txt tmp_static/Subset4/names_rna.txt\n",
      "mkdir -p tmp_static/Subset5/\n",
      "mkdir -p tmp_static/Subset6/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "make: *** No rule to make target `tmp_static/Subset4/net_weight.tsv.gz', needed by `tmp_static/Subset4/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset4/net_meanvar.tsv.gz', needed by `tmp_static/Subset4/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset4/net_covfactor.tsv.gz', needed by `tmp_static/Subset4/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset5/net_weight.tsv.gz', needed by `tmp_static/Subset5/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset5/net_meanvar.tsv.gz', needed by `tmp_static/Subset5/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset5/net_covfactor.tsv.gz', needed by `tmp_static/Subset5/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset6/net_weight.tsv.gz', needed by `tmp_static/Subset6/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset6/net_meanvar.tsv.gz', needed by `tmp_static/Subset6/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset6/net_covfactor.tsv.gz', needed by `tmp_static/Subset6/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset7/net_weight.tsv.gz', needed by `tmp_static/Subset7/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset7/net_meanvar.tsv.gz', needed by `tmp_static/Subset7/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset7/net_covfactor.tsv.gz', needed by `tmp_static/Subset7/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset8/net_weight.tsv.gz', needed by `tmp_static/Subset8/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset8/net_meanvar.tsv.gz', needed by `tmp_static/Subset8/net_nweight.tsv.gz'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cp data/subsets/Subset5/names_rna.txt tmp_static/Subset5/names_rna.txt\n",
      "cp data/subsets/Subset6/names_rna.txt tmp_static/Subset6/names_rna.txt\n",
      "mkdir -p tmp_static/Subset7/\n",
      "cp data/subsets/Subset7/names_rna.txt tmp_static/Subset7/names_rna.txt\n",
      "mkdir -p tmp_static/Subset8/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "make: *** No rule to make target `tmp_static/Subset8/net_covfactor.tsv.gz', needed by `tmp_static/Subset8/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset9/net_weight.tsv.gz', needed by `tmp_static/Subset9/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset9/net_meanvar.tsv.gz', needed by `tmp_static/Subset9/net_nweight.tsv.gz'.\n",
      "make: *** No rule to make target `tmp_static/Subset9/net_covfactor.tsv.gz', needed by `tmp_static/Subset9/net_nweight.tsv.gz'.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cp data/subsets/Subset8/names_rna.txt tmp_static/Subset8/names_rna.txt\n",
      "mkdir -p tmp_static/Subset9/\n",
      "cp data/subsets/Subset9/names_rna.txt tmp_static/Subset9/names_rna.txt\n",
      "cp tmp_static/Subset1/names_rna.txt tmp_static/Subset1/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset1/names_rna.txt tmp_static/Subset1/expression0.tsv.gz\n",
      "cp tmp_static/Subset10/names_rna.txt tmp_static/Subset10/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset10/names_rna.txt tmp_static/Subset10/expression0.tsv.gz\n",
      "cp tmp_static/Subset11/names_rna.txt tmp_static/Subset11/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset11/names_rna.txt tmp_static/Subset11/expression0.tsv.gz\n",
      "cp tmp_static/Subset12/names_rna.txt tmp_static/Subset12/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset12/names_rna.txt tmp_static/Subset12/expression0.tsv.gz\n",
      "cp tmp_static/Subset13/names_rna.txt tmp_static/Subset13/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset13/names_rna.txt tmp_static/Subset13/expression0.tsv.gz\n",
      "cp tmp_static/Subset14/names_rna.txt tmp_static/Subset14/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset14/names_rna.txt tmp_static/Subset14/expression0.tsv.gz\n",
      "cp tmp_static/Subset2/names_rna.txt tmp_static/Subset2/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset2/names_rna.txt tmp_static/Subset2/expression0.tsv.gz\n",
      "cp tmp_static/Subset3/names_rna.txt tmp_static/Subset3/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset3/names_rna.txt tmp_static/Subset3/expression0.tsv.gz\n",
      "cp tmp_static/Subset4/names_rna.txt tmp_static/Subset4/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset4/names_rna.txt tmp_static/Subset4/expression0.tsv.gz\n",
      "cp tmp_static/Subset5/names_rna.txt tmp_static/Subset5/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset5/names_rna.txt tmp_static/Subset5/expression0.tsv.gz\n",
      "cp tmp_static/Subset6/names_rna.txt tmp_static/Subset6/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset6/names_rna.txt tmp_static/Subset6/expression0.tsv.gz\n",
      "cp tmp_static/Subset7/names_rna.txt tmp_static/Subset7/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset7/names_rna.txt tmp_static/Subset7/expression0.tsv.gz\n",
      "cp tmp_static/Subset8/names_rna.txt tmp_static/Subset8/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset8/names_rna.txt tmp_static/Subset8/expression0.tsv.gz\n",
      "cp tmp_static/Subset9/names_rna.txt tmp_static/Subset9/names_atac0.txt\n",
      "python3 -m dictys  preproc selects_rna  data/expression.tsv.gz tmp_static/Subset9/names_rna.txt tmp_static/Subset9/expression0.tsv.gz\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset13/expression0.tsv.gz tmp_static/Subset13/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset14/expression0.tsv.gz tmp_static/Subset14/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset13/expression.tsv.gz tmp_static/Subset13/names_atac0.txt tmp_static/Subset13/names_atac.txt\n",
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset13/names_atac.txt data/bams tmp_static/Subset13/reads.bam tmp_static/Subset13/reads.bai tmp_static/Subset13/peaks.bed hs\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset12/expression0.tsv.gz tmp_static/Subset12/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset11/expression0.tsv.gz tmp_static/Subset11/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset14/expression.tsv.gz tmp_static/Subset14/names_atac0.txt tmp_static/Subset14/names_atac.txt\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset10/expression0.tsv.gz tmp_static/Subset10/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset9/expression0.tsv.gz tmp_static/Subset9/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset14/names_atac.txt data/bams tmp_static/Subset14/reads.bam tmp_static/Subset14/reads.bai tmp_static/Subset14/peaks.bed hs\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset6/expression0.tsv.gz tmp_static/Subset6/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset5/expression0.tsv.gz tmp_static/Subset5/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset7/expression0.tsv.gz tmp_static/Subset7/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset4/expression0.tsv.gz tmp_static/Subset4/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset8/expression0.tsv.gz tmp_static/Subset8/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset11/expression.tsv.gz tmp_static/Subset11/names_atac0.txt tmp_static/Subset11/names_atac.txt\n",
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset11/names_atac.txt data/bams tmp_static/Subset11/reads.bam tmp_static/Subset11/reads.bai tmp_static/Subset11/peaks.bed hs\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset12/expression.tsv.gz tmp_static/Subset12/names_atac0.txt tmp_static/Subset12/names_atac.txt\n",
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset12/names_atac.txt data/bams tmp_static/Subset12/reads.bam tmp_static/Subset12/reads.bai tmp_static/Subset12/peaks.bed hs\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset10/expression.tsv.gz tmp_static/Subset10/names_atac0.txt tmp_static/Subset10/names_atac.txt\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset9/expression.tsv.gz tmp_static/Subset9/names_atac0.txt tmp_static/Subset9/names_atac.txt\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset3/expression0.tsv.gz tmp_static/Subset3/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset10/names_atac.txt data/bams tmp_static/Subset10/reads.bam tmp_static/Subset10/reads.bai tmp_static/Subset10/peaks.bed hs\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset5/expression.tsv.gz tmp_static/Subset5/names_atac0.txt tmp_static/Subset5/names_atac.txt\n",
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset9/names_atac.txt data/bams tmp_static/Subset9/reads.bam tmp_static/Subset9/reads.bai tmp_static/Subset9/peaks.bed hs\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset2/expression0.tsv.gz tmp_static/Subset2/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset5/names_atac.txt data/bams tmp_static/Subset5/reads.bam tmp_static/Subset5/reads.bai tmp_static/Subset5/peaks.bed hs\n",
      "python3 -m dictys  preproc qc_reads  tmp_static/Subset1/expression0.tsv.gz tmp_static/Subset1/expression.tsv.gz 50 10 0 200 100 0\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset6/expression.tsv.gz tmp_static/Subset6/names_atac0.txt tmp_static/Subset6/names_atac.txt\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset8/expression.tsv.gz tmp_static/Subset8/names_atac0.txt tmp_static/Subset8/names_atac.txt\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset7/expression.tsv.gz tmp_static/Subset7/names_atac0.txt tmp_static/Subset7/names_atac.txt\n",
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset6/names_atac.txt data/bams tmp_static/Subset6/reads.bam tmp_static/Subset6/reads.bai tmp_static/Subset6/peaks.bed hs\n",
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset8/names_atac.txt data/bams tmp_static/Subset8/reads.bam tmp_static/Subset8/reads.bai tmp_static/Subset8/peaks.bed hs\n",
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset7/names_atac.txt data/bams tmp_static/Subset7/reads.bam tmp_static/Subset7/reads.bai tmp_static/Subset7/peaks.bed hs\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset4/expression.tsv.gz tmp_static/Subset4/names_atac0.txt tmp_static/Subset4/names_atac.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset4/names_atac.txt data/bams tmp_static/Subset4/reads.bam tmp_static/Subset4/reads.bai tmp_static/Subset4/peaks.bed hs\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset3/expression.tsv.gz tmp_static/Subset3/names_atac0.txt tmp_static/Subset3/names_atac.txt\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset2/expression.tsv.gz tmp_static/Subset2/names_atac0.txt tmp_static/Subset2/names_atac.txt\n",
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset3/names_atac.txt data/bams tmp_static/Subset3/reads.bam tmp_static/Subset3/reads.bai tmp_static/Subset3/peaks.bed hs\n",
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset2/names_atac.txt data/bams tmp_static/Subset2/reads.bam tmp_static/Subset2/reads.bai tmp_static/Subset2/peaks.bed hs\n",
      "python3 -m dictys  preproc selects_atac  tmp_static/Subset1/expression.tsv.gz tmp_static/Subset1/names_atac0.txt tmp_static/Subset1/names_atac.txt\n",
      "python3 -m dictys  chromatin macs2 --nth 4 tmp_static/Subset1/names_atac.txt data/bams tmp_static/Subset1/reads.bam tmp_static/Subset1/reads.bai tmp_static/Subset1/peaks.bed hs\n",
      "[bam_sort_core] merging from 4 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:17:58: \n",
      "# Command line: callpeak -t tmp_static/Subset13/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset13/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:17:58: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:17:58: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:17:58: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:17:58: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:17:58: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:02:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:05:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:08:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:11:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:15:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:15: 5160329 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:15: #1 tag size is determined as 49 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:15: #1 tag size = 49.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:15: #1  total tags in treatment: 5160329 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:15: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:15: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:15: #2 Skipped... \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:15: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:15: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:15: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:15: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:15: #3 Pre-compute pvalue-qvalue table... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:18:15: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:18:24: access pq hash for 11071977 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:24: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:27: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:28: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:28: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:28: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset13/reads.bam tmp_static/Subset13/reads.bai tmp_static/Subset13/peaks.bed tmp_static/Subset13/footprints.bed\n",
      "[bam_sort_core] merging from 4 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:18:13: \n",
      "# Command line: callpeak -t tmp_static/Subset14/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset14/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:13: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:13: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:18:13: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:13: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:13: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:16:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:19:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:22:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:26:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:29:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:32:  6000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:35:  7000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:39:  8000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:40: 8359031 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:40: #1 tag size is determined as 49 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:40: #1 tag size = 49.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:40: #1  total tags in treatment: 8359031 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:40: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:40: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:40: #2 Skipped... \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:40: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:40: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:40: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:40: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:40: #3 Pre-compute pvalue-qvalue table... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:18:40: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:18:54: access pq hash for 15630375 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:54: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:31: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:31: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:32: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:32: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset14/reads.bam tmp_static/Subset14/reads.bai tmp_static/Subset14/peaks.bed tmp_static/Subset14/footprints.bed\n",
      "[bam_sort_core] merging from 8 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:18:34: \n",
      "# Command line: callpeak -t tmp_static/Subset11/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset11/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:34: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:34: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:18:34: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:34: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:34: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:37:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:40:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:43:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:46:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:49:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:53:  6000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:56:  7000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:18:59:  8000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:02:  9000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:05:  10000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:08: 10765395 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:08: #1 tag size is determined as 49 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:08: #1 tag size = 49.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:08: #1  total tags in treatment: 10765395 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:08: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:08: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:08: #2 Skipped... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:08: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:08: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:08: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:08: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:08: #3 Pre-compute pvalue-qvalue table... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:19:08: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:19:28: access pq hash for 18856970 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:29: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:07: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:07: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:08: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:08: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset11/reads.bam tmp_static/Subset11/reads.bai tmp_static/Subset11/peaks.bed tmp_static/Subset11/footprints.bed\n",
      "[bam_sort_core] merging from 12 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:19:00: \n",
      "# Command line: callpeak -t tmp_static/Subset9/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset9/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:00: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:00: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:19:00: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:00: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:00: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:03:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:06:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:09:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:12:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:15:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:18:  6000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:22:  7000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:25:  8000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:29:  9000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:32:  10000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:35:  11000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:38:  12000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:41:  13000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:44:  14000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:45: 14332692 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:46: #1 tag size is determined as 49 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:46: #1 tag size = 49.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:46: #1  total tags in treatment: 14332692 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:46: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:46: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:46: #2 Skipped... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:46: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:46: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:46: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:46: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:46: #3 Pre-compute pvalue-qvalue table... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:19:46: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:20:11: access pq hash for 25670792 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:11: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:59: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:00: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:00: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:00: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset9/reads.bam tmp_static/Subset9/reads.bai tmp_static/Subset9/peaks.bed tmp_static/Subset9/footprints.bed\n",
      "[bam_sort_core] merging from 16 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:19:14: \n",
      "# Command line: callpeak -t tmp_static/Subset12/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset12/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:14: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:14: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:19:14: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:14: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:14: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:17:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:20:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:23:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:27:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:30:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:33:  6000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:36:  7000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:39:  8000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:42:  9000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:45:  10000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:48:  11000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:52:  12000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:55:  13000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:58:  14000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:01:  15000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:04:  16000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:07:  17000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:10:  18000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:11: 18145111 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:11: #1 tag size is determined as 49 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:11: #1 tag size = 49.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:11: #1  total tags in treatment: 18145111 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:11: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:11: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:11: #2 Skipped... \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ Wed, 24 Aug 2022 13:20:11: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:11: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:11: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:11: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:11: #3 Pre-compute pvalue-qvalue table... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:20:11: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:20:44: access pq hash for 32888100 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:44: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:46: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:46: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset12/reads.bam tmp_static/Subset12/reads.bai tmp_static/Subset12/peaks.bed tmp_static/Subset12/footprints.bed\n",
      "[bam_sort_core] merging from 20 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:19:58: \n",
      "# Command line: callpeak -t tmp_static/Subset5/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset5/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:58: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:58: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:19:58: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:58: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:58: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:01:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:05:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:08:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:11:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:14:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:17:  6000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:20:  7000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:23:  8000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:27:  9000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:30:  10000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:33:  11000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:36:  12000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:39:  13000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:42:  14000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:45:  15000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:48:  16000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:52:  17000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:55:  18000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:58:  19000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:01:  20000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:04:  21000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:07:  22000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:10:  23000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:13:  24000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17:  25000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17: 25033750 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17: #1 tag size is determined as 49 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17: #1 tag size = 49.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17: #1  total tags in treatment: 25033750 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17: #2 Skipped... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17: #3 Pre-compute pvalue-qvalue table... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:21:17: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:22:18: access pq hash for 37142429 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:19: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:04: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:05: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:05: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:05: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset5/reads.bam tmp_static/Subset5/reads.bai tmp_static/Subset5/peaks.bed tmp_static/Subset5/footprints.bed\n",
      "[bam_sort_core] merging from 24 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:19:57: \n",
      "# Command line: callpeak -t tmp_static/Subset10/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset10/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:57: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:57: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:19:57: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:57: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:19:57: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:00:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:04:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:07:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:10:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:13:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:17:  6000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:20:  7000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:23:  8000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:26:  9000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:29:  10000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:32:  11000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:35:  12000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:38:  13000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:42:  14000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:45:  15000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:48:  16000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:51:  17000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:54:  18000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:57:  19000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:00:  20000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:04:  21000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:07:  22000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:10:  23000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:13:  24000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:16:  25000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:19: 25843096 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:19: #1 tag size is determined as 49 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:19: #1 tag size = 49.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:19: #1  total tags in treatment: 25843096 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:19: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:19: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:19: #2 Skipped... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:19: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:19: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:19: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:19: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:19: #3 Pre-compute pvalue-qvalue table... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:21:19: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:22:14: access pq hash for 43694530 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:14: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:11: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:12: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:12: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:12: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset10/reads.bam tmp_static/Subset10/reads.bai tmp_static/Subset10/peaks.bed tmp_static/Subset10/footprints.bed\n",
      "[bam_sort_core] merging from 24 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:20:16: \n",
      "# Command line: callpeak -t tmp_static/Subset6/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset6/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:16: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:16: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:20:16: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:16: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:16: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:19:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:23:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:26:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:29:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:32:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:35:  6000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:38:  7000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:41:  8000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:44:  9000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:48:  10000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:51:  11000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:54:  12000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:57:  13000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:00:  14000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:03:  15000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:06:  16000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:09:  17000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:12:  18000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:15:  19000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:19:  20000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:22:  21000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:25:  22000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:28:  23000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:31:  24000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:34:  25000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:37:  26000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:40:  27000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:44:  28000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:44: 28274891 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45: #1 tag size is determined as 49 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45: #1 tag size = 49.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45: #1  total tags in treatment: 28274891 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45: #2 Skipped... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45: #3 Pre-compute pvalue-qvalue table... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:21:45: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:22:55: access pq hash for 40706409 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:55: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:44: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:44: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:44: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:44: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset6/reads.bam tmp_static/Subset6/reads.bai tmp_static/Subset6/peaks.bed tmp_static/Subset6/footprints.bed\n",
      "[bam_sort_core] merging from 28 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:20:39: \n",
      "# Command line: callpeak -t tmp_static/Subset7/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset7/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:39: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:39: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:20:39: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:39: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:39: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:43:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:46:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:49:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:52:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:55:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:58:  6000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:01:  7000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:05:  8000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:08:  9000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:11:  10000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:14:  11000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:17:  12000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:20:  13000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:23:  14000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:27:  15000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:30:  16000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:33:  17000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:36:  18000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:39:  19000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:42:  20000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:45:  21000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:48:  22000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:52:  23000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:55:  24000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:58:  25000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:01:  26000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:04:  27000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:07:  28000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:10:  29000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:13:  30000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:17:  31000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:20:  32000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:23:  33000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:25: 33709717 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:26: #1 tag size is determined as 48 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:26: #1 tag size = 48.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:26: #1  total tags in treatment: 33709717 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:26: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:26: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:26: #2 Skipped... \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:26: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:26: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:26: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:26: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:26: #3 Pre-compute pvalue-qvalue table... \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG @ Wed, 24 Aug 2022 13:22:26: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:23:49: access pq hash for 48797001 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:49: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:44: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:45: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:45: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:45: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset7/reads.bam tmp_static/Subset7/reads.bai tmp_static/Subset7/peaks.bed tmp_static/Subset7/footprints.bed\n",
      "[bam_sort_core] merging from 28 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:20:43: \n",
      "# Command line: callpeak -t tmp_static/Subset4/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset4/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:43: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:43: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:20:43: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:43: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:43: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:47:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:50:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:53:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:56:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:59:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:02:  6000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:05:  7000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:09:  8000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:12:  9000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:15:  10000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:18:  11000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:21:  12000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:24:  13000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:27:  14000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:31:  15000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:34:  16000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:37:  17000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:40:  18000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:43:  19000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:46:  20000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:49:  21000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:52:  22000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:56:  23000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:59:  24000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:02:  25000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:05:  26000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:08:  27000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:11:  28000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:14:  29000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:17:  30000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:21:  31000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:24:  32000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:27:  33000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:29: 33521362 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:29: #1 tag size is determined as 49 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:29: #1 tag size = 49.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:29: #1  total tags in treatment: 33521362 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:29: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:29: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:29: #2 Skipped... \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:29: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:29: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:29: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:29: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:29: #3 Pre-compute pvalue-qvalue table... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:22:29: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:23:53: access pq hash for 46633337 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:53: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:48: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:48: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:48: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:49: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset4/reads.bam tmp_static/Subset4/reads.bai tmp_static/Subset4/peaks.bed tmp_static/Subset4/footprints.bed\n",
      "[bam_sort_core] merging from 32 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:20:50: \n",
      "# Command line: callpeak -t tmp_static/Subset8/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset8/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:50: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:50: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:20:50: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:50: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:50: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:53:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:56:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:20:59:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:02:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:05:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:08:  6000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:12:  7000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:15:  8000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:18:  9000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:21:  10000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:24:  11000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:27:  12000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:30:  13000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:33:  14000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:37:  15000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:40:  16000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:43:  17000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:46:  18000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:49:  19000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:52:  20000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:55:  21000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:58:  22000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:01:  23000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:05:  24000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:08:  25000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:11:  26000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:14:  27000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:17:  28000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:20:  29000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:23:  30000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:26:  31000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:29:  32000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:33:  33000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:36:  34000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:37: 34427505 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:38: #1 tag size is determined as 49 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:38: #1 tag size = 49.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:38: #1  total tags in treatment: 34427505 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:38: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:38: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:38: #2 Skipped... \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:38: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:38: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:38: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:38: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:38: #3 Pre-compute pvalue-qvalue table... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:22:38: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:23:59: access pq hash for 50519667 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:00: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:01: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:01: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:01: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:01: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset8/reads.bam tmp_static/Subset8/reads.bai tmp_static/Subset8/peaks.bed tmp_static/Subset8/footprints.bed\n",
      "[bam_sort_core] merging from 40 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:21:58: \n",
      "# Command line: callpeak -t tmp_static/Subset3/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset3/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:58: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:58: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:21:58: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:58: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:21:58: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:02:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:05:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:08:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:11:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:14:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:17:  6000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:21:  7000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:24:  8000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:27:  9000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:30:  10000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:33:  11000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:36:  12000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:39:  13000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:42:  14000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:46:  15000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:49:  16000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:52:  17000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:55:  18000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:58:  19000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:01:  20000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:04:  21000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:08:  22000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:11:  23000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:14:  24000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:17:  25000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:20:  26000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:23:  27000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:26:  28000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:30:  29000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:33:  30000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:36:  31000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:39:  32000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:42:  33000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:45:  34000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:48:  35000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:51:  36000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:55:  37000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:58:  38000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:01:  39000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:04:  40000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:07:  41000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:10:  42000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:13:  43000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:17:  44000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:20:  45000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:23: 45974402 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:24: #1 tag size is determined as 49 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:24: #1 tag size = 49.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:24: #1  total tags in treatment: 45974402 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:24: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:24: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:24: #2 Skipped... \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:24: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:24: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:24: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:24: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:24: #3 Pre-compute pvalue-qvalue table... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:24:24: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:26:09: access pq hash for 68733098 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:09: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:27:39: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:27:40: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:27:40: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:27:40: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset3/reads.bam tmp_static/Subset3/reads.bai tmp_static/Subset3/peaks.bed tmp_static/Subset3/footprints.bed\n",
      "[bam_sort_core] merging from 48 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:22:41: \n",
      "# Command line: callpeak -t tmp_static/Subset1/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset1/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:41: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:41: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:22:41: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:41: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:41: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:44:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:47:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:50:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:53:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:22:56:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:00:  6000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:03:  7000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:06:  8000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:09:  9000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:12:  10000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:15:  11000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:18:  12000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:22:  13000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:25:  14000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:28:  15000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:31:  16000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:34:  17000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:37:  18000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:40:  19000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:43:  20000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:47:  21000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:50:  22000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:53:  23000000 \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ Wed, 24 Aug 2022 13:23:56:  24000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:59:  25000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:02:  26000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:05:  27000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:08:  28000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:12:  29000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:15:  30000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:18:  31000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:21:  32000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:24:  33000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:27:  34000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:31:  35000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:34:  36000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:37:  37000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:40:  38000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:43:  39000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:46:  40000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:49:  41000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:52:  42000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:56:  43000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:59:  44000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:02:  45000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:05:  46000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:08:  47000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:11:  48000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:14:  49000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:18:  50000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:21:  51000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:24:  52000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:27:  53000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:29: 53439963 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:29: #1 tag size is determined as 49 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:29: #1 tag size = 49.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:29: #1  total tags in treatment: 53439963 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:29: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:29: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:29: #2 Skipped... \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:29: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:29: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:29: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:29: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:29: #3 Pre-compute pvalue-qvalue table... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:25:29: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:28:00: access pq hash for 67590098 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:28:00: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:29:09: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:29:09: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:29:10: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:29:10: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset1/reads.bam tmp_static/Subset1/reads.bai tmp_static/Subset1/peaks.bed tmp_static/Subset1/footprints.bed\n",
      "[bam_sort_core] merging from 52 files and 4 in-memory blocks...\n",
      "INFO  @ Wed, 24 Aug 2022 13:23:03: \n",
      "# Command line: callpeak -t tmp_static/Subset2/reads.bam -n 04 -g hs --nomodel --shift -75 --extsize 150 --keep-dup all --verbose 4 --call-summits -q 0.05\n",
      "# ARGUMENTS LIST:\n",
      "# name = 04\n",
      "# format = AUTO\n",
      "# ChIP-seq file = ['tmp_static/Subset2/reads.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:03: #1 read tag files... \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:03: #1 read treatment tags... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:23:03: Testing format BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:03: Detected format is: BAM \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:03: * Input file is gzipped. \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:06:  1000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:09:  2000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:13:  3000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:16:  4000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:19:  5000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:22:  6000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:25:  7000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:28:  8000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:31:  9000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:34:  10000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:38:  11000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:41:  12000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:44:  13000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:47:  14000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:50:  15000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:53:  16000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:56:  17000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:23:59:  18000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:02:  19000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:06:  20000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:09:  21000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:12:  22000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:15:  23000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:18:  24000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:21:  25000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:24:  26000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:28:  27000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:31:  28000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:34:  29000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:37:  30000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:40:  31000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:43:  32000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:46:  33000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:49:  34000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:52:  35000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:56:  36000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:24:59:  37000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:02:  38000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:05:  39000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:08:  40000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:11:  41000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:14:  42000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:17:  43000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:21:  44000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:24:  45000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:27:  46000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:30:  47000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:33:  48000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:36:  49000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:39:  50000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:42:  51000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:46:  52000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:49:  53000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:52:  54000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:55:  55000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:25:58:  56000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:01:  57000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:04:  58000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:07:  59000000 \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:08: 59097541 reads have been read. \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:09: #1 tag size is determined as 49 bps \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:09: #1 tag size = 49.0 \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:09: #1  total tags in treatment: 59097541 \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:09: #1 finished! \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:09: #2 Build Peak Model... \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:09: #2 Skipped... \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:09: #2 Use 150 as fragment length \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:09: #2 Sequencing ends will be shifted towards 5' by 75 bp(s) \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:09: #3 Call peaks... \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:09: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Wed, 24 Aug 2022 13:26:09: #3 Pre-compute pvalue-qvalue table... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:26:09: Start to calculate pvalue stat... \n",
      "DEBUG @ Wed, 24 Aug 2022 13:28:29: access pq hash for 90668689 times \n",
      "INFO  @ Wed, 24 Aug 2022 13:28:29: #3 Call peaks for each chromosome... \n",
      "INFO  @ Wed, 24 Aug 2022 13:30:08: #4 Write output xls file... 04_peaks.xls \n",
      "INFO  @ Wed, 24 Aug 2022 13:30:08: #4 Write peak in narrowPeak format file... 04_peaks.narrowPeak \n",
      "INFO  @ Wed, 24 Aug 2022 13:30:08: #4 Write summits bed file... 04_summits.bed \n",
      "INFO  @ Wed, 24 Aug 2022 13:30:09: Done! \n",
      "\n",
      "python3 -m dictys  chromatin wellington --nth 4 tmp_static/Subset2/reads.bam tmp_static/Subset2/reads.bai tmp_static/Subset2/peaks.bed tmp_static/Subset2/footprints.bed\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset13/footprints.bed data/motifs.motif data/genome tmp_static/Subset13/expression.tsv.gz tmp_static/Subset13/motifs.bed tmp_static/Subset13/wellington.tsv.gz tmp_static/Subset13/homer.tsv.gz\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 8210\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 8210\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 20 (avg size of targets)\n",
      "\tBackground files for 20 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 867 sequences from chr1\n",
      "\tExtracting 328 sequences from chr10\n",
      "\tExtracting 443 sequences from chr11\n",
      "\tExtracting 406 sequences from chr12\n",
      "\tExtracting 151 sequences from chr13\n",
      "\tExtracting 290 sequences from chr14\n",
      "\tExtracting 228 sequences from chr15\n",
      "\tExtracting 362 sequences from chr16\n",
      "\tExtracting 562 sequences from chr17\n",
      "\tExtracting 113 sequences from chr18\n",
      "\tExtracting 791 sequences from chr19\n",
      "\tExtracting 527 sequences from chr2\n",
      "\tExtracting 211 sequences from chr20\n",
      "\tExtracting 110 sequences from chr21\n",
      "\tExtracting 224 sequences from chr22\n",
      "\tExtracting 427 sequences from chr3\n",
      "\tExtracting 264 sequences from chr4\n",
      "\tExtracting 351 sequences from chr5\n",
      "\tExtracting 414 sequences from chr6\n",
      "\tExtracting 347 sequences from chr7\n",
      "\tExtracting 270 sequences from chr8\n",
      "\tExtracting 325 sequences from chr9\n",
      "\tExtracting 168 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t8180 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 8209\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 8209\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 20 (avg size of targets)\n",
      "\tBackground files for 20 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 799 sequences from chr1\n",
      "\tExtracting 288 sequences from chr10\n",
      "\tExtracting 447 sequences from chr11\n",
      "\tExtracting 386 sequences from chr12\n",
      "\tExtracting 125 sequences from chr13\n",
      "\tExtracting 279 sequences from chr14\n",
      "\tExtracting 243 sequences from chr15\n",
      "\tExtracting 435 sequences from chr16\n",
      "\tExtracting 539 sequences from chr17\n",
      "\tExtracting 106 sequences from chr18\n",
      "\tExtracting 709 sequences from chr19\n",
      "\tExtracting 565 sequences from chr2\n",
      "\tExtracting 233 sequences from chr20\n",
      "\tExtracting 101 sequences from chr21\n",
      "\tExtracting 222 sequences from chr22\n",
      "\tExtracting 428 sequences from chr3\n",
      "\tExtracting 274 sequences from chr4\n",
      "\tExtracting 342 sequences from chr5\n",
      "\tExtracting 441 sequences from chr6\n",
      "\tExtracting 391 sequences from chr7\n",
      "\tExtracting 273 sequences from chr8\n",
      "\tExtracting 357 sequences from chr9\n",
      "\tExtracting 192 sequences from chrX\n",
      "\tExtracting 2 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t8177 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 8210\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 8210\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 20 (avg size of targets)\n",
      "\tBackground files for 20 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 840 sequences from chr1\n",
      "\tExtracting 289 sequences from chr10\n",
      "\tExtracting 484 sequences from chr11\n",
      "\tExtracting 414 sequences from chr12\n",
      "\tExtracting 130 sequences from chr13\n",
      "\tExtracting 277 sequences from chr14\n",
      "\tExtracting 229 sequences from chr15\n",
      "\tExtracting 404 sequences from chr16\n",
      "\tExtracting 580 sequences from chr17\n",
      "\tExtracting 98 sequences from chr18\n",
      "\tExtracting 824 sequences from chr19\n",
      "\tExtracting 566 sequences from chr2\n",
      "\tExtracting 177 sequences from chr20\n",
      "\tExtracting 120 sequences from chr21\n",
      "\tExtracting 212 sequences from chr22\n",
      "\tExtracting 440 sequences from chr3\n",
      "\tExtracting 226 sequences from chr4\n",
      "\tExtracting 305 sequences from chr5\n",
      "\tExtracting 400 sequences from chr6\n",
      "\tExtracting 397 sequences from chr7\n",
      "\tExtracting 241 sequences from chr8\n",
      "\tExtracting 374 sequences from chr9\n",
      "\tExtracting 157 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t8184 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 8210\n",
      "\t\tpeakfile formatted lines: 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 8210\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 20 (avg size of targets)\n",
      "\tBackground files for 20 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 874 sequences from chr1\n",
      "\tExtracting 277 sequences from chr10\n",
      "\tExtracting 420 sequences from chr11\n",
      "\tExtracting 377 sequences from chr12\n",
      "\tExtracting 115 sequences from chr13\n",
      "\tExtracting 258 sequences from chr14\n",
      "\tExtracting 188 sequences from chr15\n",
      "\tExtracting 416 sequences from chr16\n",
      "\tExtracting 584 sequences from chr17\n",
      "\tExtracting 94 sequences from chr18\n",
      "\tExtracting 959 sequences from chr19\n",
      "\tExtracting 463 sequences from chr2\n",
      "\tExtracting 182 sequences from chr20\n",
      "\tExtracting 393 sequences from chr21\n",
      "\tExtracting 195 sequences from chr22\n",
      "\tExtracting 384 sequences from chr3\n",
      "\tExtracting 250 sequences from chr4\n",
      "\tExtracting 287 sequences from chr5\n",
      "\tExtracting 359 sequences from chr6\n",
      "\tExtracting 395 sequences from chr7\n",
      "\tExtracting 218 sequences from chr8\n",
      "\tExtracting 316 sequences from chr9\n",
      "\tExtracting 127 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t8132 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset13/wellington.tsv.gz tmp_static/Subset13/homer.tsv.gz tmp_static/Subset13/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset13/expression.tsv.gz tmp_static/Subset13/wellington.tsv.gz data/gene.bed tmp_static/Subset13/tssdist.tsv.gz\n",
      "python3 -m dictys  chromatin linking  tmp_static/Subset13/binding.tsv.gz tmp_static/Subset13/tssdist.tsv.gz tmp_static/Subset13/linking.tsv.gz\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset13/linking.tsv.gz tmp_static/Subset13/binlinking.tsv.gz 20\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset14/footprints.bed data/motifs.motif data/genome tmp_static/Subset14/expression.tsv.gz tmp_static/Subset14/motifs.bed tmp_static/Subset14/wellington.tsv.gz tmp_static/Subset14/homer.tsv.gz\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 22590\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 22590\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2409 sequences from chr1\n",
      "\tExtracting 843 sequences from chr10\n",
      "\tExtracting 1260 sequences from chr11\n",
      "\tExtracting 1159 sequences from chr12\n",
      "\tExtracting 408 sequences from chr13\n",
      "\tExtracting 804 sequences from chr14\n",
      "\tExtracting 662 sequences from chr15\n",
      "\tExtracting 1105 sequences from chr16\n",
      "\tExtracting 1514 sequences from chr17\n",
      "\tExtracting 313 sequences from chr18\n",
      "\tExtracting 1879 sequences from chr19\n",
      "\tExtracting 1418 sequences from chr2\n",
      "\tExtracting 572 sequences from chr20\n",
      "\tExtracting 334 sequences from chr21\n",
      "\tExtracting 578 sequences from chr22\n",
      "\tExtracting 1198 sequences from chr3\n",
      "\tExtracting 746 sequences from chr4\n",
      "\tExtracting 971 sequences from chr5\n",
      "\tExtracting 1240 sequences from chr6\n",
      "\tExtracting 1054 sequences from chr7\n",
      "\tExtracting 763 sequences from chr8\n",
      "\tExtracting 856 sequences from chr9\n",
      "\tExtracting 484 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t22570 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 22590\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 22590\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2326 sequences from chr1\n",
      "\tExtracting 899 sequences from chr10\n",
      "\tExtracting 1175 sequences from chr11\n",
      "\tExtracting 1109 sequences from chr12\n",
      "\tExtracting 424 sequences from chr13\n",
      "\tExtracting 800 sequences from chr14\n",
      "\tExtracting 729 sequences from chr15\n",
      "\tExtracting 1084 sequences from chr16\n",
      "\tExtracting 1567 sequences from chr17\n",
      "\tExtracting 326 sequences from chr18\n",
      "\tExtracting 1714 sequences from chr19\n",
      "\tExtracting 1540 sequences from chr2\n",
      "\tExtracting 588 sequences from chr20\n",
      "\tExtracting 324 sequences from chr21\n",
      "\tExtracting 561 sequences from chr22\n",
      "\tExtracting 1157 sequences from chr3\n",
      "\tExtracting 699 sequences from chr4\n",
      "\tExtracting 980 sequences from chr5\n",
      "\tExtracting 1251 sequences from chr6\n",
      "\tExtracting 1056 sequences from chr7\n",
      "\tExtracting 803 sequences from chr8\n",
      "\tExtracting 912 sequences from chr9\n",
      "\tExtracting 534 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t22558 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 22587\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 22587\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2312 sequences from chr1\n",
      "\tExtracting 969 sequences from chr10\n",
      "\tExtracting 1158 sequences from chr11\n",
      "\tExtracting 1138 sequences from chr12\n",
      "\tExtracting 450 sequences from chr13\n",
      "\tExtracting 812 sequences from chr14\n",
      "\tExtracting 699 sequences from chr15\n",
      "\tExtracting 967 sequences from chr16\n",
      "\tExtracting 1563 sequences from chr17\n",
      "\tExtracting 277 sequences from chr18\n",
      "\tExtracting 1561 sequences from chr19\n",
      "\tExtracting 1555 sequences from chr2\n",
      "\tExtracting 589 sequences from chr20\n",
      "\tExtracting 309 sequences from chr21\n",
      "\tExtracting 639 sequences from chr22\n",
      "\tExtracting 1248 sequences from chr3\n",
      "\tExtracting 754 sequences from chr4\n",
      "\tExtracting 1035 sequences from chr5\n",
      "\tExtracting 1236 sequences from chr6\n",
      "\tExtracting 1016 sequences from chr7\n",
      "\tExtracting 800 sequences from chr8\n",
      "\tExtracting 847 sequences from chr9\n",
      "\tExtracting 627 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t22561 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 22590\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 22590\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2375 sequences from chr1\n",
      "\tExtracting 803 sequences from chr10\n",
      "\tExtracting 1200 sequences from chr11\n",
      "\tExtracting 1152 sequences from chr12\n",
      "\tExtracting 374 sequences from chr13\n",
      "\tExtracting 775 sequences from chr14\n",
      "\tExtracting 657 sequences from chr15\n",
      "\tExtracting 1128 sequences from chr16\n",
      "\tExtracting 1530 sequences from chr17\n",
      "\tExtracting 234 sequences from chr18\n",
      "\tExtracting 2187 sequences from chr19\n",
      "\tExtracting 1401 sequences from chr2\n",
      "\tExtracting 569 sequences from chr20\n",
      "\tExtracting 580 sequences from chr21\n",
      "\tExtracting 566 sequences from chr22\n",
      "\tExtracting 1123 sequences from chr3\n",
      "\tExtracting 701 sequences from chr4\n",
      "\tExtracting 856 sequences from chr5\n",
      "\tExtracting 1137 sequences from chr6\n",
      "\tExtracting 1125 sequences from chr7\n",
      "\tExtracting 715 sequences from chr8\n",
      "\tExtracting 930 sequences from chr9\n",
      "\tExtracting 395 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t22514 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset14/wellington.tsv.gz tmp_static/Subset14/homer.tsv.gz tmp_static/Subset14/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset14/expression.tsv.gz tmp_static/Subset14/wellington.tsv.gz data/gene.bed tmp_static/Subset14/tssdist.tsv.gz\n",
      "python3 -m dictys  chromatin linking  tmp_static/Subset14/binding.tsv.gz tmp_static/Subset14/tssdist.tsv.gz tmp_static/Subset14/linking.tsv.gz\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset14/linking.tsv.gz tmp_static/Subset14/binlinking.tsv.gz 20\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset11/footprints.bed data/motifs.motif data/genome tmp_static/Subset11/expression.tsv.gz tmp_static/Subset11/motifs.bed tmp_static/Subset11/wellington.tsv.gz tmp_static/Subset11/homer.tsv.gz\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2511 sequences from chr1\n",
      "\tExtracting 976 sequences from chr10\n",
      "\tExtracting 1284 sequences from chr11\n",
      "\tExtracting 1237 sequences from chr12\n",
      "\tExtracting 431 sequences from chr13\n",
      "\tExtracting 917 sequences from chr14\n",
      "\tExtracting 799 sequences from chr15\n",
      "\tExtracting 1186 sequences from chr16\n",
      "\tExtracting 1706 sequences from chr17\n",
      "\tExtracting 378 sequences from chr18\n",
      "\tExtracting 2117 sequences from chr19\n",
      "\tExtracting 1680 sequences from chr2\n",
      "\tExtracting 653 sequences from chr20\n",
      "\tExtracting 370 sequences from chr21\n",
      "\tExtracting 655 sequences from chr22\n",
      "\tExtracting 1288 sequences from chr3\n",
      "\tExtracting 801 sequences from chr4\n",
      "\tExtracting 1047 sequences from chr5\n",
      "\tExtracting 1352 sequences from chr6\n",
      "\tExtracting 1174 sequences from chr7\n",
      "\tExtracting 842 sequences from chr8\n",
      "\tExtracting 1000 sequences from chr9\n",
      "\tExtracting 556 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24960 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2622 sequences from chr1\n",
      "\tExtracting 1005 sequences from chr10\n",
      "\tExtracting 1201 sequences from chr11\n",
      "\tExtracting 1271 sequences from chr12\n",
      "\tExtracting 474 sequences from chr13\n",
      "\tExtracting 870 sequences from chr14\n",
      "\tExtracting 822 sequences from chr15\n",
      "\tExtracting 1148 sequences from chr16\n",
      "\tExtracting 1677 sequences from chr17\n",
      "\tExtracting 397 sequences from chr18\n",
      "\tExtracting 1943 sequences from chr19\n",
      "\tExtracting 1708 sequences from chr2\n",
      "\tExtracting 657 sequences from chr20\n",
      "\tExtracting 312 sequences from chr21\n",
      "\tExtracting 698 sequences from chr22\n",
      "\tExtracting 1317 sequences from chr3\n",
      "\tExtracting 844 sequences from chr4\n",
      "\tExtracting 1086 sequences from chr5\n",
      "\tExtracting 1296 sequences from chr6\n",
      "\tExtracting 1177 sequences from chr7\n",
      "\tExtracting 885 sequences from chr8\n",
      "\tExtracting 946 sequences from chr9\n",
      "\tExtracting 617 sequences from chrX\n",
      "\tExtracting 2 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24975 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 24997\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 24997\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2623 sequences from chr1\n",
      "\tExtracting 1041 sequences from chr10\n",
      "\tExtracting 1312 sequences from chr11\n",
      "\tExtracting 1206 sequences from chr12\n",
      "\tExtracting 490 sequences from chr13\n",
      "\tExtracting 911 sequences from chr14\n",
      "\tExtracting 856 sequences from chr15\n",
      "\tExtracting 1153 sequences from chr16\n",
      "\tExtracting 1637 sequences from chr17\n",
      "\tExtracting 386 sequences from chr18\n",
      "\tExtracting 1807 sequences from chr19\n",
      "\tExtracting 1686 sequences from chr2\n",
      "\tExtracting 671 sequences from chr20\n",
      "\tExtracting 303 sequences from chr21\n",
      "\tExtracting 591 sequences from chr22\n",
      "\tExtracting 1303 sequences from chr3\n",
      "\tExtracting 781 sequences from chr4\n",
      "\tExtracting 1114 sequences from chr5\n",
      "\tExtracting 1390 sequences from chr6\n",
      "\tExtracting 1253 sequences from chr7\n",
      "\tExtracting 855 sequences from chr8\n",
      "\tExtracting 982 sequences from chr9\n",
      "\tExtracting 629 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24980 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2542 sequences from chr1\n",
      "\tExtracting 899 sequences from chr10\n",
      "\tExtracting 1268 sequences from chr11\n",
      "\tExtracting 1281 sequences from chr12\n",
      "\tExtracting 427 sequences from chr13\n",
      "\tExtracting 889 sequences from chr14\n",
      "\tExtracting 705 sequences from chr15\n",
      "\tExtracting 1244 sequences from chr16\n",
      "\tExtracting 1711 sequences from chr17\n",
      "\tExtracting 333 sequences from chr18\n",
      "\tExtracting 2409 sequences from chr19\n",
      "\tExtracting 1599 sequences from chr2\n",
      "\tExtracting 632 sequences from chr20\n",
      "\tExtracting 643 sequences from chr21\n",
      "\tExtracting 597 sequences from chr22\n",
      "\tExtracting 1246 sequences from chr3\n",
      "\tExtracting 752 sequences from chr4\n",
      "\tExtracting 1008 sequences from chr5\n",
      "\tExtracting 1266 sequences from chr6\n",
      "\tExtracting 1205 sequences from chr7\n",
      "\tExtracting 756 sequences from chr8\n",
      "\tExtracting 1012 sequences from chr9\n",
      "\tExtracting 473 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24897 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset11/wellington.tsv.gz tmp_static/Subset11/homer.tsv.gz tmp_static/Subset11/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset11/expression.tsv.gz tmp_static/Subset11/wellington.tsv.gz data/gene.bed tmp_static/Subset11/tssdist.tsv.gz\n",
      "python3 -m dictys  chromatin linking  tmp_static/Subset11/binding.tsv.gz tmp_static/Subset11/tssdist.tsv.gz tmp_static/Subset11/linking.tsv.gz\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset11/linking.tsv.gz tmp_static/Subset11/binlinking.tsv.gz 20\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset9/footprints.bed data/motifs.motif data/genome tmp_static/Subset9/expression.tsv.gz tmp_static/Subset9/motifs.bed tmp_static/Subset9/wellington.tsv.gz tmp_static/Subset9/homer.tsv.gz\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2561 sequences from chr1\n",
      "\tExtracting 1096 sequences from chr10\n",
      "\tExtracting 1320 sequences from chr11\n",
      "\tExtracting 1312 sequences from chr12\n",
      "\tExtracting 428 sequences from chr13\n",
      "\tExtracting 861 sequences from chr14\n",
      "\tExtracting 775 sequences from chr15\n",
      "\tExtracting 1155 sequences from chr16\n",
      "\tExtracting 1655 sequences from chr17\n",
      "\tExtracting 357 sequences from chr18\n",
      "\tExtracting 1848 sequences from chr19\n",
      "\tExtracting 1728 sequences from chr2\n",
      "\tExtracting 664 sequences from chr20\n",
      "\tExtracting 363 sequences from chr21\n",
      "\tExtracting 589 sequences from chr22\n",
      "\tExtracting 1381 sequences from chr3\n",
      "\tExtracting 857 sequences from chr4\n",
      "\tExtracting 1083 sequences from chr5\n",
      "\tExtracting 1300 sequences from chr6\n",
      "\tExtracting 1199 sequences from chr7\n",
      "\tExtracting 878 sequences from chr8\n",
      "\tExtracting 1009 sequences from chr9\n",
      "\tExtracting 552 sequences from chrX\n",
      "\tExtracting 2 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24973 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 24997\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 24997\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2558 sequences from chr1\n",
      "\tExtracting 1093 sequences from chr10\n",
      "\tExtracting 1327 sequences from chr11\n",
      "\tExtracting 1316 sequences from chr12\n",
      "\tExtracting 470 sequences from chr13\n",
      "\tExtracting 817 sequences from chr14\n",
      "\tExtracting 778 sequences from chr15\n",
      "\tExtracting 1050 sequences from chr16\n",
      "\tExtracting 1469 sequences from chr17\n",
      "\tExtracting 408 sequences from chr18\n",
      "\tExtracting 1547 sequences from chr19\n",
      "\tExtracting 1808 sequences from chr2\n",
      "\tExtracting 687 sequences from chr20\n",
      "\tExtracting 384 sequences from chr21\n",
      "\tExtracting 592 sequences from chr22\n",
      "\tExtracting 1391 sequences from chr3\n",
      "\tExtracting 868 sequences from chr4\n",
      "\tExtracting 1165 sequences from chr5\n",
      "\tExtracting 1475 sequences from chr6\n",
      "\tExtracting 1150 sequences from chr7\n",
      "\tExtracting 934 sequences from chr8\n",
      "\tExtracting 1083 sequences from chr9\n",
      "\tExtracting 608 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24978 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2533 sequences from chr1\n",
      "\tExtracting 1137 sequences from chr10\n",
      "\tExtracting 1358 sequences from chr11\n",
      "\tExtracting 1301 sequences from chr12\n",
      "\tExtracting 492 sequences from chr13\n",
      "\tExtracting 900 sequences from chr14\n",
      "\tExtracting 792 sequences from chr15\n",
      "\tExtracting 1075 sequences from chr16\n",
      "\tExtracting 1547 sequences from chr17\n",
      "\tExtracting 396 sequences from chr18\n",
      "\tExtracting 1723 sequences from chr19\n",
      "\tExtracting 1738 sequences from chr2\n",
      "\tExtracting 614 sequences from chr20\n",
      "\tExtracting 323 sequences from chr21\n",
      "\tExtracting 608 sequences from chr22\n",
      "\tExtracting 1379 sequences from chr3\n",
      "\tExtracting 842 sequences from chr4\n",
      "\tExtracting 1079 sequences from chr5\n",
      "\tExtracting 1416 sequences from chr6\n",
      "\tExtracting 1210 sequences from chr7\n",
      "\tExtracting 900 sequences from chr8\n",
      "\tExtracting 1022 sequences from chr9\n",
      "\tExtracting 595 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24980 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 20 (avg size of targets)\n",
      "\tBackground files for 20 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tExtracting 2552 sequences from chr1\n",
      "\tExtracting 1003 sequences from chr10\n",
      "\tExtracting 1283 sequences from chr11\n",
      "\tExtracting 1236 sequences from chr12\n",
      "\tExtracting 420 sequences from chr13\n",
      "\tExtracting 818 sequences from chr14\n",
      "\tExtracting 694 sequences from chr15\n",
      "\tExtracting 1217 sequences from chr16\n",
      "\tExtracting 1714 sequences from chr17\n",
      "\tExtracting 369 sequences from chr18\n",
      "\tExtracting 2258 sequences from chr19\n",
      "\tExtracting 1640 sequences from chr2\n",
      "\tExtracting 663 sequences from chr20\n",
      "\tExtracting 498 sequences from chr21\n",
      "\tExtracting 584 sequences from chr22\n",
      "\tExtracting 1359 sequences from chr3\n",
      "\tExtracting 831 sequences from chr4\n",
      "\tExtracting 1018 sequences from chr5\n",
      "\tExtracting 1294 sequences from chr6\n",
      "\tExtracting 1209 sequences from chr7\n",
      "\tExtracting 832 sequences from chr8\n",
      "\tExtracting 1065 sequences from chr9\n",
      "\tExtracting 400 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24957 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset5/footprints.bed data/motifs.motif data/genome tmp_static/Subset5/expression.tsv.gz tmp_static/Subset5/motifs.bed tmp_static/Subset5/wellington.tsv.gz tmp_static/Subset5/homer.tsv.gz\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset9/wellington.tsv.gz tmp_static/Subset9/homer.tsv.gz tmp_static/Subset9/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset9/expression.tsv.gz tmp_static/Subset9/wellington.tsv.gz data/gene.bed tmp_static/Subset9/tssdist.tsv.gz\n",
      "python3 -m dictys  chromatin linking  tmp_static/Subset9/binding.tsv.gz tmp_static/Subset9/tssdist.tsv.gz tmp_static/Subset9/linking.tsv.gz\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset9/linking.tsv.gz tmp_static/Subset9/binlinking.tsv.gz 20\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2588 sequences from chr1\n",
      "\tExtracting 994 sequences from chr10\n",
      "\tExtracting 1367 sequences from chr11\n",
      "\tExtracting 1303 sequences from chr12\n",
      "\tExtracting 459 sequences from chr13\n",
      "\tExtracting 878 sequences from chr14\n",
      "\tExtracting 743 sequences from chr15\n",
      "\tExtracting 1234 sequences from chr16\n",
      "\tExtracting 1656 sequences from chr17\n",
      "\tExtracting 333 sequences from chr18\n",
      "\tExtracting 2058 sequences from chr19\n",
      "\tExtracting 1607 sequences from chr2\n",
      "\tExtracting 644 sequences from chr20\n",
      "\tExtracting 349 sequences from chr21\n",
      "\tExtracting 639 sequences from chr22\n",
      "\tExtracting 1257 sequences from chr3\n",
      "\tExtracting 798 sequences from chr4\n",
      "\tExtracting 1097 sequences from chr5\n",
      "\tExtracting 1342 sequences from chr6\n",
      "\tExtracting 1209 sequences from chr7\n",
      "\tExtracting 855 sequences from chr8\n",
      "\tExtracting 980 sequences from chr9\n",
      "\tExtracting 568 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24959 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2529 sequences from chr1\n",
      "\tExtracting 964 sequences from chr10\n",
      "\tExtracting 1304 sequences from chr11\n",
      "\tExtracting 1307 sequences from chr12\n",
      "\tExtracting 473 sequences from chr13\n",
      "\tExtracting 846 sequences from chr14\n",
      "\tExtracting 752 sequences from chr15\n",
      "\tExtracting 1107 sequences from chr16\n",
      "\tExtracting 1640 sequences from chr17\n",
      "\tExtracting 362 sequences from chr18\n",
      "\tExtracting 1974 sequences from chr19\n",
      "\tExtracting 1648 sequences from chr2\n",
      "\tExtracting 652 sequences from chr20\n",
      "\tExtracting 319 sequences from chr21\n",
      "\tExtracting 641 sequences from chr22\n",
      "\tExtracting 1325 sequences from chr3\n",
      "\tExtracting 854 sequences from chr4\n",
      "\tExtracting 1061 sequences from chr5\n",
      "\tExtracting 1482 sequences from chr6\n",
      "\tExtracting 1203 sequences from chr7\n",
      "\tExtracting 924 sequences from chr8\n",
      "\tExtracting 968 sequences from chr9\n",
      "\tExtracting 642 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24977 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 24997\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 24997\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2615 sequences from chr1\n",
      "\tExtracting 1040 sequences from chr10\n",
      "\tExtracting 1310 sequences from chr11\n",
      "\tExtracting 1299 sequences from chr12\n",
      "\tExtracting 513 sequences from chr13\n",
      "\tExtracting 897 sequences from chr14\n",
      "\tExtracting 767 sequences from chr15\n",
      "\tExtracting 1078 sequences from chr16\n",
      "\tExtracting 1648 sequences from chr17\n",
      "\tExtracting 348 sequences from chr18\n",
      "\tExtracting 1818 sequences from chr19\n",
      "\tExtracting 1764 sequences from chr2\n",
      "\tExtracting 704 sequences from chr20\n",
      "\tExtracting 347 sequences from chr21\n",
      "\tExtracting 631 sequences from chr22\n",
      "\tExtracting 1272 sequences from chr3\n",
      "\tExtracting 841 sequences from chr4\n",
      "\tExtracting 1105 sequences from chr5\n",
      "\tExtracting 1357 sequences from chr6\n",
      "\tExtracting 1102 sequences from chr7\n",
      "\tExtracting 890 sequences from chr8\n",
      "\tExtracting 974 sequences from chr9\n",
      "\tExtracting 650 sequences from chrX\n",
      "\tExtracting 3 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24973 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2660 sequences from chr1\n",
      "\tExtracting 857 sequences from chr10\n",
      "\tExtracting 1289 sequences from chr11\n",
      "\tExtracting 1252 sequences from chr12\n",
      "\tExtracting 430 sequences from chr13\n",
      "\tExtracting 877 sequences from chr14\n",
      "\tExtracting 657 sequences from chr15\n",
      "\tExtracting 1311 sequences from chr16\n",
      "\tExtracting 1708 sequences from chr17\n",
      "\tExtracting 278 sequences from chr18\n",
      "\tExtracting 2454 sequences from chr19\n",
      "\tExtracting 1505 sequences from chr2\n",
      "\tExtracting 628 sequences from chr20\n",
      "\tExtracting 800 sequences from chr21\n",
      "\tExtracting 577 sequences from chr22\n",
      "\tExtracting 1234 sequences from chr3\n",
      "\tExtracting 769 sequences from chr4\n",
      "\tExtracting 961 sequences from chr5\n",
      "\tExtracting 1283 sequences from chr6\n",
      "\tExtracting 1160 sequences from chr7\n",
      "\tExtracting 755 sequences from chr8\n",
      "\tExtracting 1014 sequences from chr9\n",
      "\tExtracting 390 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24850 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset5/wellington.tsv.gz tmp_static/Subset5/homer.tsv.gz tmp_static/Subset5/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset5/expression.tsv.gz tmp_static/Subset5/wellington.tsv.gz data/gene.bed tmp_static/Subset5/tssdist.tsv.gz\n",
      "python3 -m dictys  chromatin linking  tmp_static/Subset5/binding.tsv.gz tmp_static/Subset5/tssdist.tsv.gz tmp_static/Subset5/linking.tsv.gz\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset5/linking.tsv.gz tmp_static/Subset5/binlinking.tsv.gz 20\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset6/footprints.bed data/motifs.motif data/genome tmp_static/Subset6/expression.tsv.gz tmp_static/Subset6/motifs.bed tmp_static/Subset6/wellington.tsv.gz tmp_static/Subset6/homer.tsv.gz\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2657 sequences from chr1\n",
      "\tExtracting 1030 sequences from chr10\n",
      "\tExtracting 1287 sequences from chr11\n",
      "\tExtracting 1208 sequences from chr12\n",
      "\tExtracting 476 sequences from chr13\n",
      "\tExtracting 911 sequences from chr14\n",
      "\tExtracting 826 sequences from chr15\n",
      "\tExtracting 1169 sequences from chr16\n",
      "\tExtracting 1659 sequences from chr17\n",
      "\tExtracting 355 sequences from chr18\n",
      "\tExtracting 1916 sequences from chr19\n",
      "\tExtracting 1778 sequences from chr2\n",
      "\tExtracting 659 sequences from chr20\n",
      "\tExtracting 331 sequences from chr21\n",
      "\tExtracting 639 sequences from chr22\n",
      "\tExtracting 1292 sequences from chr3\n",
      "\tExtracting 772 sequences from chr4\n",
      "\tExtracting 1085 sequences from chr5\n",
      "\tExtracting 1393 sequences from chr6\n",
      "\tExtracting 1119 sequences from chr7\n",
      "\tExtracting 804 sequences from chr8\n",
      "\tExtracting 996 sequences from chr9\n",
      "\tExtracting 609 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24971 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tExtracting 2612 sequences from chr1\n",
      "\tExtracting 941 sequences from chr10\n",
      "\tExtracting 1276 sequences from chr11\n",
      "\tExtracting 1247 sequences from chr12\n",
      "\tExtracting 442 sequences from chr13\n",
      "\tExtracting 846 sequences from chr14\n",
      "\tExtracting 742 sequences from chr15\n",
      "\tExtracting 1205 sequences from chr16\n",
      "\tExtracting 1711 sequences from chr17\n",
      "\tExtracting 371 sequences from chr18\n",
      "\tExtracting 2075 sequences from chr19\n",
      "\tExtracting 1685 sequences from chr2\n",
      "\tExtracting 703 sequences from chr20\n",
      "\tExtracting 326 sequences from chr21\n",
      "\tExtracting 645 sequences from chr22\n",
      "\tExtracting 1274 sequences from chr3\n",
      "\tExtracting 780 sequences from chr4\n",
      "\tExtracting 1078 sequences from chr5\n",
      "\tExtracting 1407 sequences from chr6\n",
      "\tExtracting 1164 sequences from chr7\n",
      "\tExtracting 809 sequences from chr8\n",
      "\tExtracting 1040 sequences from chr9\n",
      "\tExtracting 577 sequences from chrX\n",
      "\tExtracting 2 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24958 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 24997\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 24997\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2512 sequences from chr1\n",
      "\tExtracting 1062 sequences from chr10\n",
      "\tExtracting 1317 sequences from chr11\n",
      "\tExtracting 1279 sequences from chr12\n",
      "\tExtracting 489 sequences from chr13\n",
      "\tExtracting 872 sequences from chr14\n",
      "\tExtracting 771 sequences from chr15\n",
      "\tExtracting 1161 sequences from chr16\n",
      "\tExtracting 1729 sequences from chr17\n",
      "\tExtracting 360 sequences from chr18\n",
      "\tExtracting 1803 sequences from chr19\n",
      "\tExtracting 1721 sequences from chr2\n",
      "\tExtracting 638 sequences from chr20\n",
      "\tExtracting 283 sequences from chr21\n",
      "\tExtracting 636 sequences from chr22\n",
      "\tExtracting 1337 sequences from chr3\n",
      "\tExtracting 872 sequences from chr4\n",
      "\tExtracting 1071 sequences from chr5\n",
      "\tExtracting 1443 sequences from chr6\n",
      "\tExtracting 1154 sequences from chr7\n",
      "\tExtracting 886 sequences from chr8\n",
      "\tExtracting 962 sequences from chr9\n",
      "\tExtracting 613 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24971 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2588 sequences from chr1\n",
      "\tExtracting 929 sequences from chr10\n",
      "\tExtracting 1279 sequences from chr11\n",
      "\tExtracting 1307 sequences from chr12\n",
      "\tExtracting 411 sequences from chr13\n",
      "\tExtracting 851 sequences from chr14\n",
      "\tExtracting 679 sequences from chr15\n",
      "\tExtracting 1297 sequences from chr16\n",
      "\tExtracting 1689 sequences from chr17\n",
      "\tExtracting 319 sequences from chr18\n",
      "\tExtracting 2386 sequences from chr19\n",
      "\tExtracting 1553 sequences from chr2\n",
      "\tExtracting 595 sequences from chr20\n",
      "\tExtracting 762 sequences from chr21\n",
      "\tExtracting 585 sequences from chr22\n",
      "\tExtracting 1245 sequences from chr3\n",
      "\tExtracting 785 sequences from chr4\n",
      "\tExtracting 974 sequences from chr5\n",
      "\tExtracting 1214 sequences from chr6\n",
      "\tExtracting 1192 sequences from chr7\n",
      "\tExtracting 769 sequences from chr8\n",
      "\tExtracting 1009 sequences from chr9\n",
      "\tExtracting 426 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24845 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset6/wellington.tsv.gz tmp_static/Subset6/homer.tsv.gz tmp_static/Subset6/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset6/expression.tsv.gz tmp_static/Subset6/wellington.tsv.gz data/gene.bed tmp_static/Subset6/tssdist.tsv.gz\n",
      "python3 -m dictys  chromatin linking  tmp_static/Subset6/binding.tsv.gz tmp_static/Subset6/tssdist.tsv.gz tmp_static/Subset6/linking.tsv.gz\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset6/linking.tsv.gz tmp_static/Subset6/binlinking.tsv.gz 20\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset7/footprints.bed data/motifs.motif data/genome tmp_static/Subset7/expression.tsv.gz tmp_static/Subset7/motifs.bed tmp_static/Subset7/wellington.tsv.gz tmp_static/Subset7/homer.tsv.gz\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2587 sequences from chr1\n",
      "\tExtracting 963 sequences from chr10\n",
      "\tExtracting 1346 sequences from chr11\n",
      "\tExtracting 1264 sequences from chr12\n",
      "\tExtracting 472 sequences from chr13\n",
      "\tExtracting 867 sequences from chr14\n",
      "\tExtracting 750 sequences from chr15\n",
      "\tExtracting 1221 sequences from chr16\n",
      "\tExtracting 1676 sequences from chr17\n",
      "\tExtracting 319 sequences from chr18\n",
      "\tExtracting 2078 sequences from chr19\n",
      "\tExtracting 1642 sequences from chr2\n",
      "\tExtracting 662 sequences from chr20\n",
      "\tExtracting 353 sequences from chr21\n",
      "\tExtracting 610 sequences from chr22\n",
      "\tExtracting 1262 sequences from chr3\n",
      "\tExtracting 812 sequences from chr4\n",
      "\tExtracting 1034 sequences from chr5\n",
      "\tExtracting 1381 sequences from chr6\n",
      "\tExtracting 1189 sequences from chr7\n",
      "\tExtracting 881 sequences from chr8\n",
      "\tExtracting 1017 sequences from chr9\n",
      "\tExtracting 578 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24965 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2629 sequences from chr1\n",
      "\tExtracting 1019 sequences from chr10\n",
      "\tExtracting 1289 sequences from chr11\n",
      "\tExtracting 1317 sequences from chr12\n",
      "\tExtracting 472 sequences from chr13\n",
      "\tExtracting 824 sequences from chr14\n",
      "\tExtracting 799 sequences from chr15\n",
      "\tExtracting 1139 sequences from chr16\n",
      "\tExtracting 1682 sequences from chr17\n",
      "\tExtracting 334 sequences from chr18\n",
      "\tExtracting 1897 sequences from chr19\n",
      "\tExtracting 1721 sequences from chr2\n",
      "\tExtracting 647 sequences from chr20\n",
      "\tExtracting 375 sequences from chr21\n",
      "\tExtracting 660 sequences from chr22\n",
      "\tExtracting 1258 sequences from chr3\n",
      "\tExtracting 809 sequences from chr4\n",
      "\tExtracting 1094 sequences from chr5\n",
      "\tExtracting 1441 sequences from chr6\n",
      "\tExtracting 1147 sequences from chr7\n",
      "\tExtracting 831 sequences from chr8\n",
      "\tExtracting 990 sequences from chr9\n",
      "\tExtracting 595 sequences from chrX\n",
      "\tExtracting 2 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24971 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 24997\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 24997\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2561 sequences from chr1\n",
      "\tExtracting 995 sequences from chr10\n",
      "\tExtracting 1288 sequences from chr11\n",
      "\tExtracting 1306 sequences from chr12\n",
      "\tExtracting 482 sequences from chr13\n",
      "\tExtracting 889 sequences from chr14\n",
      "\tExtracting 780 sequences from chr15\n",
      "\tExtracting 1116 sequences from chr16\n",
      "\tExtracting 1691 sequences from chr17\n",
      "\tExtracting 328 sequences from chr18\n",
      "\tExtracting 1834 sequences from chr19\n",
      "\tExtracting 1761 sequences from chr2\n",
      "\tExtracting 693 sequences from chr20\n",
      "\tExtracting 346 sequences from chr21\n",
      "\tExtracting 594 sequences from chr22\n",
      "\tExtracting 1333 sequences from chr3\n",
      "\tExtracting 795 sequences from chr4\n",
      "\tExtracting 1061 sequences from chr5\n",
      "\tExtracting 1443 sequences from chr6\n",
      "\tExtracting 1216 sequences from chr7\n",
      "\tExtracting 856 sequences from chr8\n",
      "\tExtracting 976 sequences from chr9\n",
      "\tExtracting 629 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24973 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2682 sequences from chr1\n",
      "\tExtracting 901 sequences from chr10\n",
      "\tExtracting 1299 sequences from chr11\n",
      "\tExtracting 1286 sequences from chr12\n",
      "\tExtracting 436 sequences from chr13\n",
      "\tExtracting 852 sequences from chr14\n",
      "\tExtracting 660 sequences from chr15\n",
      "\tExtracting 1257 sequences from chr16\n",
      "\tExtracting 1680 sequences from chr17\n",
      "\tExtracting 308 sequences from chr18\n",
      "\tExtracting 2457 sequences from chr19\n",
      "\tExtracting 1523 sequences from chr2\n",
      "\tExtracting 617 sequences from chr20\n",
      "\tExtracting 745 sequences from chr21\n",
      "\tExtracting 583 sequences from chr22\n",
      "\tExtracting 1214 sequences from chr3\n",
      "\tExtracting 766 sequences from chr4\n",
      "\tExtracting 1013 sequences from chr5\n",
      "\tExtracting 1241 sequences from chr6\n",
      "\tExtracting 1162 sequences from chr7\n",
      "\tExtracting 804 sequences from chr8\n",
      "\tExtracting 977 sequences from chr9\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tExtracting 386 sequences from chrX\n",
      "\tExtracting 2 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24851 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset4/footprints.bed data/motifs.motif data/genome tmp_static/Subset4/expression.tsv.gz tmp_static/Subset4/motifs.bed tmp_static/Subset4/wellington.tsv.gz tmp_static/Subset4/homer.tsv.gz\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset12/footprints.bed data/motifs.motif data/genome tmp_static/Subset12/expression.tsv.gz tmp_static/Subset12/motifs.bed tmp_static/Subset12/wellington.tsv.gz tmp_static/Subset12/homer.tsv.gz\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset7/wellington.tsv.gz tmp_static/Subset7/homer.tsv.gz tmp_static/Subset7/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset7/expression.tsv.gz tmp_static/Subset7/wellington.tsv.gz data/gene.bed tmp_static/Subset7/tssdist.tsv.gz\n",
      "python3 -m dictys  chromatin linking  tmp_static/Subset7/binding.tsv.gz tmp_static/Subset7/tssdist.tsv.gz tmp_static/Subset7/linking.tsv.gz\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset7/linking.tsv.gz tmp_static/Subset7/binlinking.tsv.gz 20\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2622 sequences from chr1\n",
      "\tExtracting 985 sequences from chr10\n",
      "\tExtracting 1278 sequences from chr11\n",
      "\tExtracting 1275 sequences from chr12\n",
      "\tExtracting 462 sequences from chr13\n",
      "\tExtracting 872 sequences from chr14\n",
      "\tExtracting 755 sequences from chr15\n",
      "\tExtracting 1155 sequences from chr16\n",
      "\tExtracting 1700 sequences from chr17\n",
      "\tExtracting 319 sequences from chr18\n",
      "\tExtracting 2134 sequences from chr19\n",
      "\tExtracting 1635 sequences from chr2\n",
      "\tExtracting 649 sequences from chr20\n",
      "\tExtracting 372 sequences from chr21\n",
      "\tExtracting 615 sequences from chr22\n",
      "\tExtracting 1226 sequences from chr3\n",
      "\tExtracting 778 sequences from chr4\n",
      "\tExtracting 1085 sequences from chr5\n",
      "\tExtracting 1422 sequences from chr6\n",
      "\tExtracting 1184 sequences from chr7\n",
      "\tExtracting 886 sequences from chr8\n",
      "\tExtracting 948 sequences from chr9\n",
      "\tExtracting 604 sequences from chrX\n",
      "\tExtracting 2 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24963 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2530 sequences from chr1\n",
      "\tExtracting 1021 sequences from chr10\n",
      "\tExtracting 1310 sequences from chr11\n",
      "\tExtracting 1317 sequences from chr12\n",
      "\tExtracting 492 sequences from chr13\n",
      "\tExtracting 864 sequences from chr14\n",
      "\tExtracting 726 sequences from chr15\n",
      "\tExtracting 1199 sequences from chr16\n",
      "\tExtracting 1617 sequences from chr17\n",
      "\tExtracting 361 sequences from chr18\n",
      "\tExtracting 1942 sequences from chr19\n",
      "\tExtracting 1660 sequences from chr2\n",
      "\tExtracting 683 sequences from chr20\n",
      "\tExtracting 359 sequences from chr21\n",
      "\tExtracting 608 sequences from chr22\n",
      "\tExtracting 1339 sequences from chr3\n",
      "\tExtracting 797 sequences from chr4\n",
      "\tExtracting 1069 sequences from chr5\n",
      "\tExtracting 1456 sequences from chr6\n",
      "\tExtracting 1199 sequences from chr7\n",
      "\tExtracting 869 sequences from chr8\n",
      "\tExtracting 937 sequences from chr9\n",
      "\tExtracting 622 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24977 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 24997\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 24997\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2585 sequences from chr1\n",
      "\tExtracting 1019 sequences from chr10\n",
      "\tExtracting 1339 sequences from chr11\n",
      "\tExtracting 1260 sequences from chr12\n",
      "\tExtracting 549 sequences from chr13\n",
      "\tExtracting 826 sequences from chr14\n",
      "\tExtracting 802 sequences from chr15\n",
      "\tExtracting 1149 sequences from chr16\n",
      "\tExtracting 1699 sequences from chr17\n",
      "\tExtracting 326 sequences from chr18\n",
      "\tExtracting 1732 sequences from chr19\n",
      "\tExtracting 1757 sequences from chr2\n",
      "\tExtracting 688 sequences from chr20\n",
      "\tExtracting 333 sequences from chr21\n",
      "\tExtracting 658 sequences from chr22\n",
      "\tExtracting 1299 sequences from chr3\n",
      "\tExtracting 852 sequences from chr4\n",
      "\tExtracting 1072 sequences from chr5\n",
      "\tExtracting 1426 sequences from chr6\n",
      "\tExtracting 1143 sequences from chr7\n",
      "\tExtracting 887 sequences from chr8\n",
      "\tExtracting 999 sequences from chr9\n",
      "\tExtracting 573 sequences from chrX\n",
      "\tExtracting 3 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24976 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2663 sequences from chr1\n",
      "\tExtracting 886 sequences from chr10\n",
      "\tExtracting 1311 sequences from chr11\n",
      "\tExtracting 1300 sequences from chr12\n",
      "\tExtracting 425 sequences from chr13\n",
      "\tExtracting 840 sequences from chr14\n",
      "\tExtracting 655 sequences from chr15\n",
      "\tExtracting 1250 sequences from chr16\n",
      "\tExtracting 1637 sequences from chr17\n",
      "\tExtracting 294 sequences from chr18\n",
      "\tExtracting 2371 sequences from chr19\n",
      "\tExtracting 1527 sequences from chr2\n",
      "\tExtracting 625 sequences from chr20\n",
      "\tExtracting 760 sequences from chr21\n",
      "\tExtracting 583 sequences from chr22\n",
      "\tExtracting 1232 sequences from chr3\n",
      "\tExtracting 790 sequences from chr4\n",
      "\tExtracting 994 sequences from chr5\n",
      "\tExtracting 1247 sequences from chr6\n",
      "\tExtracting 1209 sequences from chr7\n",
      "\tExtracting 816 sequences from chr8\n",
      "\tExtracting 1041 sequences from chr9\n",
      "\tExtracting 398 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24855 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2580 sequences from chr1\n",
      "\tExtracting 1032 sequences from chr10\n",
      "\tExtracting 1285 sequences from chr11\n",
      "\tExtracting 1288 sequences from chr12\n",
      "\tExtracting 470 sequences from chr13\n",
      "\tExtracting 859 sequences from chr14\n",
      "\tExtracting 750 sequences from chr15\n",
      "\tExtracting 1182 sequences from chr16\n",
      "\tExtracting 1637 sequences from chr17\n",
      "\tExtracting 387 sequences from chr18\n",
      "\tExtracting 1947 sequences from chr19\n",
      "\tExtracting 1748 sequences from chr2\n",
      "\tExtracting 699 sequences from chr20\n",
      "\tExtracting 353 sequences from chr21\n",
      "\tExtracting 614 sequences from chr22\n",
      "\tExtracting 1391 sequences from chr3\n",
      "\tExtracting 804 sequences from chr4\n",
      "\tExtracting 1011 sequences from chr5\n",
      "\tExtracting 1339 sequences from chr6\n",
      "\tExtracting 1189 sequences from chr7\n",
      "\tExtracting 820 sequences from chr8\n",
      "\tExtracting 1087 sequences from chr9\n",
      "\tExtracting 499 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24971 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 24997\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 24997\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2484 sequences from chr1\n",
      "\tExtracting 1132 sequences from chr10\n",
      "\tExtracting 1279 sequences from chr11\n",
      "\tExtracting 1241 sequences from chr12\n",
      "\tExtracting 480 sequences from chr13\n",
      "\tExtracting 809 sequences from chr14\n",
      "\tExtracting 767 sequences from chr15\n",
      "\tExtracting 1121 sequences from chr16\n",
      "\tExtracting 1639 sequences from chr17\n",
      "\tExtracting 408 sequences from chr18\n",
      "\tExtracting 1716 sequences from chr19\n",
      "\tExtracting 1717 sequences from chr2\n",
      "\tExtracting 737 sequences from chr20\n",
      "\tExtracting 326 sequences from chr21\n",
      "\tExtracting 652 sequences from chr22\n",
      "\tExtracting 1411 sequences from chr3\n",
      "\tExtracting 819 sequences from chr4\n",
      "\tExtracting 1053 sequences from chr5\n",
      "\tExtracting 1428 sequences from chr6\n",
      "\tExtracting 1221 sequences from chr7\n",
      "\tExtracting 912 sequences from chr8\n",
      "\tExtracting 1077 sequences from chr9\n",
      "\tExtracting 548 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24978 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2506 sequences from chr1\n",
      "\tExtracting 1091 sequences from chr10\n",
      "\tExtracting 1294 sequences from chr11\n",
      "\tExtracting 1286 sequences from chr12\n",
      "\tExtracting 483 sequences from chr13\n",
      "\tExtracting 847 sequences from chr14\n",
      "\tExtracting 771 sequences from chr15\n",
      "\tExtracting 1155 sequences from chr16\n",
      "\tExtracting 1619 sequences from chr17\n",
      "\tExtracting 385 sequences from chr18\n",
      "\tExtracting 1766 sequences from chr19\n",
      "\tExtracting 1720 sequences from chr2\n",
      "\tExtracting 688 sequences from chr20\n",
      "\tExtracting 314 sequences from chr21\n",
      "\tExtracting 627 sequences from chr22\n",
      "\tExtracting 1451 sequences from chr3\n",
      "\tExtracting 814 sequences from chr4\n",
      "\tExtracting 1056 sequences from chr5\n",
      "\tExtracting 1387 sequences from chr6\n",
      "\tExtracting 1174 sequences from chr7\n",
      "\tExtracting 909 sequences from chr8\n",
      "\tExtracting 1061 sequences from chr9\n",
      "\tExtracting 569 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24973 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2654 sequences from chr1\n",
      "\tExtracting 909 sequences from chr10\n",
      "\tExtracting 1304 sequences from chr11\n",
      "\tExtracting 1254 sequences from chr12\n",
      "\tExtracting 401 sequences from chr13\n",
      "\tExtracting 795 sequences from chr14\n",
      "\tExtracting 703 sequences from chr15\n",
      "\tExtracting 1259 sequences from chr16\n",
      "\tExtracting 1772 sequences from chr17\n",
      "\tExtracting 319 sequences from chr18\n",
      "\tExtracting 2275 sequences from chr19\n",
      "\tExtracting 1582 sequences from chr2\n",
      "\tExtracting 615 sequences from chr20\n",
      "\tExtracting 709 sequences from chr21\n",
      "\tExtracting 609 sequences from chr22\n",
      "\tExtracting 1258 sequences from chr3\n",
      "\tExtracting 801 sequences from chr4\n",
      "\tExtracting 1001 sequences from chr5\n",
      "\tExtracting 1245 sequences from chr6\n",
      "\tExtracting 1189 sequences from chr7\n",
      "\tExtracting 776 sequences from chr8\n",
      "\tExtracting 1013 sequences from chr9\n",
      "\tExtracting 427 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24871 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset10/footprints.bed data/motifs.motif data/genome tmp_static/Subset10/expression.tsv.gz tmp_static/Subset10/motifs.bed tmp_static/Subset10/wellington.tsv.gz tmp_static/Subset10/homer.tsv.gz\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset4/wellington.tsv.gz tmp_static/Subset4/homer.tsv.gz tmp_static/Subset4/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset4/expression.tsv.gz tmp_static/Subset4/wellington.tsv.gz data/gene.bed tmp_static/Subset4/tssdist.tsv.gz\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset12/wellington.tsv.gz tmp_static/Subset12/homer.tsv.gz tmp_static/Subset12/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset12/expression.tsv.gz tmp_static/Subset12/wellington.tsv.gz data/gene.bed tmp_static/Subset12/tssdist.tsv.gz\n",
      "python3 -m dictys  chromatin linking  tmp_static/Subset4/binding.tsv.gz tmp_static/Subset4/tssdist.tsv.gz tmp_static/Subset4/linking.tsv.gz\n",
      "python3 -m dictys  chromatin linking  tmp_static/Subset12/binding.tsv.gz tmp_static/Subset12/tssdist.tsv.gz tmp_static/Subset12/linking.tsv.gz\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset4/linking.tsv.gz tmp_static/Subset4/binlinking.tsv.gz 20\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset12/linking.tsv.gz tmp_static/Subset12/binlinking.tsv.gz 20\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2646 sequences from chr1\n",
      "\tExtracting 1019 sequences from chr10\n",
      "\tExtracting 1314 sequences from chr11\n",
      "\tExtracting 1258 sequences from chr12\n",
      "\tExtracting 439 sequences from chr13\n",
      "\tExtracting 840 sequences from chr14\n",
      "\tExtracting 714 sequences from chr15\n",
      "\tExtracting 1199 sequences from chr16\n",
      "\tExtracting 1689 sequences from chr17\n",
      "\tExtracting 317 sequences from chr18\n",
      "\tExtracting 2099 sequences from chr19\n",
      "\tExtracting 1608 sequences from chr2\n",
      "\tExtracting 689 sequences from chr20\n",
      "\tExtracting 378 sequences from chr21\n",
      "\tExtracting 616 sequences from chr22\n",
      "\tExtracting 1311 sequences from chr3\n",
      "\tExtracting 820 sequences from chr4\n",
      "\tExtracting 1066 sequences from chr5\n",
      "\tExtracting 1390 sequences from chr6\n",
      "\tExtracting 1152 sequences from chr7\n",
      "\tExtracting 853 sequences from chr8\n",
      "\tExtracting 999 sequences from chr9\n",
      "\tExtracting 536 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24953 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2602 sequences from chr1\n",
      "\tExtracting 978 sequences from chr10\n",
      "\tExtracting 1316 sequences from chr11\n",
      "\tExtracting 1326 sequences from chr12\n",
      "\tExtracting 445 sequences from chr13\n",
      "\tExtracting 872 sequences from chr14\n",
      "\tExtracting 746 sequences from chr15\n",
      "\tExtracting 1182 sequences from chr16\n",
      "\tExtracting 1649 sequences from chr17\n",
      "\tExtracting 379 sequences from chr18\n",
      "\tExtracting 1858 sequences from chr19\n",
      "\tExtracting 1700 sequences from chr2\n",
      "\tExtracting 666 sequences from chr20\n",
      "\tExtracting 357 sequences from chr21\n",
      "\tExtracting 665 sequences from chr22\n",
      "\tExtracting 1341 sequences from chr3\n",
      "\tExtracting 767 sequences from chr4\n",
      "\tExtracting 1092 sequences from chr5\n",
      "\tExtracting 1381 sequences from chr6\n",
      "\tExtracting 1126 sequences from chr7\n",
      "\tExtracting 926 sequences from chr8\n",
      "\tExtracting 1022 sequences from chr9\n",
      "\tExtracting 560 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24956 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 24997\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 24997\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2502 sequences from chr1\n",
      "\tExtracting 1067 sequences from chr10\n",
      "\tExtracting 1343 sequences from chr11\n",
      "\tExtracting 1254 sequences from chr12\n",
      "\tExtracting 477 sequences from chr13\n",
      "\tExtracting 860 sequences from chr14\n",
      "\tExtracting 864 sequences from chr15\n",
      "\tExtracting 1115 sequences from chr16\n",
      "\tExtracting 1637 sequences from chr17\n",
      "\tExtracting 346 sequences from chr18\n",
      "\tExtracting 1765 sequences from chr19\n",
      "\tExtracting 1725 sequences from chr2\n",
      "\tExtracting 701 sequences from chr20\n",
      "\tExtracting 367 sequences from chr21\n",
      "\tExtracting 656 sequences from chr22\n",
      "\tExtracting 1302 sequences from chr3\n",
      "\tExtracting 848 sequences from chr4\n",
      "\tExtracting 1076 sequences from chr5\n",
      "\tExtracting 1370 sequences from chr6\n",
      "\tExtracting 1152 sequences from chr7\n",
      "\tExtracting 913 sequences from chr8\n",
      "\tExtracting 1031 sequences from chr9\n",
      "\tExtracting 581 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24952 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2698 sequences from chr1\n",
      "\tExtracting 900 sequences from chr10\n",
      "\tExtracting 1309 sequences from chr11\n",
      "\tExtracting 1231 sequences from chr12\n",
      "\tExtracting 406 sequences from chr13\n",
      "\tExtracting 839 sequences from chr14\n",
      "\tExtracting 670 sequences from chr15\n",
      "\tExtracting 1234 sequences from chr16\n",
      "\tExtracting 1788 sequences from chr17\n",
      "\tExtracting 300 sequences from chr18\n",
      "\tExtracting 2479 sequences from chr19\n",
      "\tExtracting 1541 sequences from chr2\n",
      "\tExtracting 623 sequences from chr20\n",
      "\tExtracting 649 sequences from chr21\n",
      "\tExtracting 580 sequences from chr22\n",
      "\tExtracting 1286 sequences from chr3\n",
      "\tExtracting 767 sequences from chr4\n",
      "\tExtracting 949 sequences from chr5\n",
      "\tExtracting 1218 sequences from chr6\n",
      "\tExtracting 1204 sequences from chr7\n",
      "\tExtracting 777 sequences from chr8\n",
      "\tExtracting 1004 sequences from chr9\n",
      "\tExtracting 421 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24874 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset10/wellington.tsv.gz tmp_static/Subset10/homer.tsv.gz tmp_static/Subset10/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset10/expression.tsv.gz tmp_static/Subset10/wellington.tsv.gz data/gene.bed tmp_static/Subset10/tssdist.tsv.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python3 -m dictys  chromatin linking  tmp_static/Subset10/binding.tsv.gz tmp_static/Subset10/tssdist.tsv.gz tmp_static/Subset10/linking.tsv.gz\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset10/linking.tsv.gz tmp_static/Subset10/binlinking.tsv.gz 20\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset8/footprints.bed data/motifs.motif data/genome tmp_static/Subset8/expression.tsv.gz tmp_static/Subset8/motifs.bed tmp_static/Subset8/wellington.tsv.gz tmp_static/Subset8/homer.tsv.gz\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2617 sequences from chr1\n",
      "\tExtracting 1016 sequences from chr10\n",
      "\tExtracting 1324 sequences from chr11\n",
      "\tExtracting 1281 sequences from chr12\n",
      "\tExtracting 425 sequences from chr13\n",
      "\tExtracting 897 sequences from chr14\n",
      "\tExtracting 778 sequences from chr15\n",
      "\tExtracting 1242 sequences from chr16\n",
      "\tExtracting 1668 sequences from chr17\n",
      "\tExtracting 331 sequences from chr18\n",
      "\tExtracting 2031 sequences from chr19\n",
      "\tExtracting 1608 sequences from chr2\n",
      "\tExtracting 622 sequences from chr20\n",
      "\tExtracting 371 sequences from chr21\n",
      "\tExtracting 650 sequences from chr22\n",
      "\tExtracting 1319 sequences from chr3\n",
      "\tExtracting 816 sequences from chr4\n",
      "\tExtracting 1038 sequences from chr5\n",
      "\tExtracting 1437 sequences from chr6\n",
      "\tExtracting 1100 sequences from chr7\n",
      "\tExtracting 875 sequences from chr8\n",
      "\tExtracting 1002 sequences from chr9\n",
      "\tExtracting 511 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24959 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 24997\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 24997\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2668 sequences from chr1\n",
      "\tExtracting 976 sequences from chr10\n",
      "\tExtracting 1272 sequences from chr11\n",
      "\tExtracting 1257 sequences from chr12\n",
      "\tExtracting 468 sequences from chr13\n",
      "\tExtracting 885 sequences from chr14\n",
      "\tExtracting 781 sequences from chr15\n",
      "\tExtracting 1109 sequences from chr16\n",
      "\tExtracting 1636 sequences from chr17\n",
      "\tExtracting 359 sequences from chr18\n",
      "\tExtracting 1771 sequences from chr19\n",
      "\tExtracting 1763 sequences from chr2\n",
      "\tExtracting 648 sequences from chr20\n",
      "\tExtracting 316 sequences from chr21\n",
      "\tExtracting 655 sequences from chr22\n",
      "\tExtracting 1316 sequences from chr3\n",
      "\tExtracting 867 sequences from chr4\n",
      "\tExtracting 1060 sequences from chr5\n",
      "\tExtracting 1511 sequences from chr6\n",
      "\tExtracting 1185 sequences from chr7\n",
      "\tExtracting 911 sequences from chr8\n",
      "\tExtracting 984 sequences from chr9\n",
      "\tExtracting 572 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24970 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2556 sequences from chr1\n",
      "\tExtracting 960 sequences from chr10\n",
      "\tExtracting 1269 sequences from chr11\n",
      "\tExtracting 1252 sequences from chr12\n",
      "\tExtracting 422 sequences from chr13\n",
      "\tExtracting 813 sequences from chr14\n",
      "\tExtracting 789 sequences from chr15\n",
      "\tExtracting 1156 sequences from chr16\n",
      "\tExtracting 1645 sequences from chr17\n",
      "\tExtracting 358 sequences from chr18\n",
      "\tExtracting 1948 sequences from chr19\n",
      "\tExtracting 1715 sequences from chr2\n",
      "\tExtracting 685 sequences from chr20\n",
      "\tExtracting 375 sequences from chr21\n",
      "\tExtracting 682 sequences from chr22\n",
      "\tExtracting 1385 sequences from chr3\n",
      "\tExtracting 782 sequences from chr4\n",
      "\tExtracting 1053 sequences from chr5\n",
      "\tExtracting 1495 sequences from chr6\n",
      "\tExtracting 1184 sequences from chr7\n",
      "\tExtracting 902 sequences from chr8\n",
      "\tExtracting 976 sequences from chr9\n",
      "\tExtracting 568 sequences from chrX\n",
      "\tExtracting 2 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24972 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2652 sequences from chr1\n",
      "\tExtracting 843 sequences from chr10\n",
      "\tExtracting 1312 sequences from chr11\n",
      "\tExtracting 1261 sequences from chr12\n",
      "\tExtracting 390 sequences from chr13\n",
      "\tExtracting 842 sequences from chr14\n",
      "\tExtracting 677 sequences from chr15\n",
      "\tExtracting 1228 sequences from chr16\n",
      "\tExtracting 1695 sequences from chr17\n",
      "\tExtracting 301 sequences from chr18\n",
      "\tExtracting 2405 sequences from chr19\n",
      "\tExtracting 1581 sequences from chr2\n",
      "\tExtracting 627 sequences from chr20\n",
      "\tExtracting 704 sequences from chr21\n",
      "\tExtracting 615 sequences from chr22\n",
      "\tExtracting 1234 sequences from chr3\n",
      "\tExtracting 769 sequences from chr4\n",
      "\tExtracting 988 sequences from chr5\n",
      "\tExtracting 1314 sequences from chr6\n",
      "\tExtracting 1207 sequences from chr7\n",
      "\tExtracting 822 sequences from chr8\n",
      "\tExtracting 991 sequences from chr9\n",
      "\tExtracting 390 sequences from chrX\n",
      "\tExtracting 2 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24850 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset8/wellington.tsv.gz tmp_static/Subset8/homer.tsv.gz tmp_static/Subset8/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset8/expression.tsv.gz tmp_static/Subset8/wellington.tsv.gz data/gene.bed tmp_static/Subset8/tssdist.tsv.gz\n",
      "python3 -m dictys  chromatin linking  tmp_static/Subset8/binding.tsv.gz tmp_static/Subset8/tssdist.tsv.gz tmp_static/Subset8/linking.tsv.gz\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset8/linking.tsv.gz tmp_static/Subset8/binlinking.tsv.gz 20\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset1/footprints.bed data/motifs.motif data/genome tmp_static/Subset1/expression.tsv.gz tmp_static/Subset1/motifs.bed tmp_static/Subset1/wellington.tsv.gz tmp_static/Subset1/homer.tsv.gz\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2644 sequences from chr1\n",
      "\tExtracting 991 sequences from chr10\n",
      "\tExtracting 1311 sequences from chr11\n",
      "\tExtracting 1364 sequences from chr12\n",
      "\tExtracting 454 sequences from chr13\n",
      "\tExtracting 846 sequences from chr14\n",
      "\tExtracting 758 sequences from chr15\n",
      "\tExtracting 1183 sequences from chr16\n",
      "\tExtracting 1644 sequences from chr17\n",
      "\tExtracting 349 sequences from chr18\n",
      "\tExtracting 2033 sequences from chr19\n",
      "\tExtracting 1666 sequences from chr2\n",
      "\tExtracting 581 sequences from chr20\n",
      "\tExtracting 378 sequences from chr21\n",
      "\tExtracting 616 sequences from chr22\n",
      "\tExtracting 1380 sequences from chr3\n",
      "\tExtracting 775 sequences from chr4\n",
      "\tExtracting 1050 sequences from chr5\n",
      "\tExtracting 1355 sequences from chr6\n",
      "\tExtracting 1120 sequences from chr7\n",
      "\tExtracting 864 sequences from chr8\n",
      "\tExtracting 1000 sequences from chr9\n",
      "\tExtracting 591 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24953 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 24997\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 24997\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2571 sequences from chr1\n",
      "\tExtracting 1056 sequences from chr10\n",
      "\tExtracting 1297 sequences from chr11\n",
      "\tExtracting 1246 sequences from chr12\n",
      "\tExtracting 495 sequences from chr13\n",
      "\tExtracting 900 sequences from chr14\n",
      "\tExtracting 827 sequences from chr15\n",
      "\tExtracting 1090 sequences from chr16\n",
      "\tExtracting 1653 sequences from chr17\n",
      "\tExtracting 363 sequences from chr18\n",
      "\tExtracting 1810 sequences from chr19\n",
      "\tExtracting 1695 sequences from chr2\n",
      "\tExtracting 687 sequences from chr20\n",
      "\tExtracting 340 sequences from chr21\n",
      "\tExtracting 607 sequences from chr22\n",
      "\tExtracting 1285 sequences from chr3\n",
      "\tExtracting 845 sequences from chr4\n",
      "\tExtracting 1096 sequences from chr5\n",
      "\tExtracting 1463 sequences from chr6\n",
      "\tExtracting 1137 sequences from chr7\n",
      "\tExtracting 903 sequences from chr8\n",
      "\tExtracting 962 sequences from chr9\n",
      "\tExtracting 642 sequences from chrX\n",
      "\tExtracting 3 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24973 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 18 (avg size of targets)\n",
      "\tBackground files for 18 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2543 sequences from chr1\n",
      "\tExtracting 1050 sequences from chr10\n",
      "\tExtracting 1348 sequences from chr11\n",
      "\tExtracting 1231 sequences from chr12\n",
      "\tExtracting 497 sequences from chr13\n",
      "\tExtracting 897 sequences from chr14\n",
      "\tExtracting 731 sequences from chr15\n",
      "\tExtracting 1146 sequences from chr16\n",
      "\tExtracting 1661 sequences from chr17\n",
      "\tExtracting 330 sequences from chr18\n",
      "\tExtracting 1964 sequences from chr19\n",
      "\tExtracting 1641 sequences from chr2\n",
      "\tExtracting 668 sequences from chr20\n",
      "\tExtracting 382 sequences from chr21\n",
      "\tExtracting 617 sequences from chr22\n",
      "\tExtracting 1308 sequences from chr3\n",
      "\tExtracting 837 sequences from chr4\n",
      "\tExtracting 1072 sequences from chr5\n",
      "\tExtracting 1429 sequences from chr6\n",
      "\tExtracting 1201 sequences from chr7\n",
      "\tExtracting 889 sequences from chr8\n",
      "\tExtracting 925 sequences from chr9\n",
      "\tExtracting 605 sequences from chrX\n",
      "\tExtracting 2 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24974 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2701 sequences from chr1\n",
      "\tExtracting 917 sequences from chr10\n",
      "\tExtracting 1280 sequences from chr11\n",
      "\tExtracting 1275 sequences from chr12\n",
      "\tExtracting 439 sequences from chr13\n",
      "\tExtracting 893 sequences from chr14\n",
      "\tExtracting 673 sequences from chr15\n",
      "\tExtracting 1253 sequences from chr16\n",
      "\tExtracting 1655 sequences from chr17\n",
      "\tExtracting 302 sequences from chr18\n",
      "\tExtracting 2356 sequences from chr19\n",
      "\tExtracting 1517 sequences from chr2\n",
      "\tExtracting 608 sequences from chr20\n",
      "\tExtracting 669 sequences from chr21\n",
      "\tExtracting 591 sequences from chr22\n",
      "\tExtracting 1211 sequences from chr3\n",
      "\tExtracting 804 sequences from chr4\n",
      "\tExtracting 984 sequences from chr5\n",
      "\tExtracting 1266 sequences from chr6\n",
      "\tExtracting 1229 sequences from chr7\n",
      "\tExtracting 791 sequences from chr8\n",
      "\tExtracting 1030 sequences from chr9\n",
      "\tExtracting 443 sequences from chrX\n",
      "\tExtracting 2 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24889 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset1/wellington.tsv.gz tmp_static/Subset1/homer.tsv.gz tmp_static/Subset1/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset1/expression.tsv.gz tmp_static/Subset1/wellington.tsv.gz data/gene.bed tmp_static/Subset1/tssdist.tsv.gz\n",
      "python3 -m dictys  chromatin linking  tmp_static/Subset1/binding.tsv.gz tmp_static/Subset1/tssdist.tsv.gz tmp_static/Subset1/linking.tsv.gz\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset1/linking.tsv.gz tmp_static/Subset1/binlinking.tsv.gz 20\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset3/footprints.bed data/motifs.motif data/genome tmp_static/Subset3/expression.tsv.gz tmp_static/Subset3/motifs.bed tmp_static/Subset3/wellington.tsv.gz tmp_static/Subset3/homer.tsv.gz\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 24997\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 24997\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2477 sequences from chr1\n",
      "\tExtracting 1054 sequences from chr10\n",
      "\tExtracting 1318 sequences from chr11\n",
      "\tExtracting 1323 sequences from chr12\n",
      "\tExtracting 490 sequences from chr13\n",
      "\tExtracting 802 sequences from chr14\n",
      "\tExtracting 798 sequences from chr15\n",
      "\tExtracting 1092 sequences from chr16\n",
      "\tExtracting 1658 sequences from chr17\n",
      "\tExtracting 413 sequences from chr18\n",
      "\tExtracting 1658 sequences from chr19\n",
      "\tExtracting 1756 sequences from chr2\n",
      "\tExtracting 733 sequences from chr20\n",
      "\tExtracting 336 sequences from chr21\n",
      "\tExtracting 658 sequences from chr22\n",
      "\tExtracting 1419 sequences from chr3\n",
      "\tExtracting 823 sequences from chr4\n",
      "\tExtracting 1073 sequences from chr5\n",
      "\tExtracting 1372 sequences from chr6\n",
      "\tExtracting 1175 sequences from chr7\n",
      "\tExtracting 936 sequences from chr8\n",
      "\tExtracting 1020 sequences from chr9\n",
      "\tExtracting 572 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24957 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2545 sequences from chr1\n",
      "\tExtracting 1069 sequences from chr10\n",
      "\tExtracting 1235 sequences from chr11\n",
      "\tExtracting 1319 sequences from chr12\n",
      "\tExtracting 498 sequences from chr13\n",
      "\tExtracting 786 sequences from chr14\n",
      "\tExtracting 793 sequences from chr15\n",
      "\tExtracting 1129 sequences from chr16\n",
      "\tExtracting 1657 sequences from chr17\n",
      "\tExtracting 375 sequences from chr18\n",
      "\tExtracting 1775 sequences from chr19\n",
      "\tExtracting 1659 sequences from chr2\n",
      "\tExtracting 667 sequences from chr20\n",
      "\tExtracting 327 sequences from chr21\n",
      "\tExtracting 623 sequences from chr22\n",
      "\tExtracting 1441 sequences from chr3\n",
      "\tExtracting 845 sequences from chr4\n",
      "\tExtracting 1107 sequences from chr5\n",
      "\tExtracting 1391 sequences from chr6\n",
      "\tExtracting 1192 sequences from chr7\n",
      "\tExtracting 932 sequences from chr8\n",
      "\tExtracting 1055 sequences from chr9\n",
      "\tExtracting 556 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24976 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2593 sequences from chr1\n",
      "\tExtracting 1020 sequences from chr10\n",
      "\tExtracting 1332 sequences from chr11\n",
      "\tExtracting 1281 sequences from chr12\n",
      "\tExtracting 451 sequences from chr13\n",
      "\tExtracting 856 sequences from chr14\n",
      "\tExtracting 727 sequences from chr15\n",
      "\tExtracting 1143 sequences from chr16\n",
      "\tExtracting 1568 sequences from chr17\n",
      "\tExtracting 380 sequences from chr18\n",
      "\tExtracting 1916 sequences from chr19\n",
      "\tExtracting 1735 sequences from chr2\n",
      "\tExtracting 671 sequences from chr20\n",
      "\tExtracting 403 sequences from chr21\n",
      "\tExtracting 596 sequences from chr22\n",
      "\tExtracting 1358 sequences from chr3\n",
      "\tExtracting 839 sequences from chr4\n",
      "\tExtracting 1103 sequences from chr5\n",
      "\tExtracting 1381 sequences from chr6\n",
      "\tExtracting 1129 sequences from chr7\n",
      "\tExtracting 896 sequences from chr8\n",
      "\tExtracting 1070 sequences from chr9\n",
      "\tExtracting 505 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24954 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 20 (avg size of targets)\n",
      "\tBackground files for 20 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2639 sequences from chr1\n",
      "\tExtracting 1003 sequences from chr10\n",
      "\tExtracting 1306 sequences from chr11\n",
      "\tExtracting 1243 sequences from chr12\n",
      "\tExtracting 390 sequences from chr13\n",
      "\tExtracting 802 sequences from chr14\n",
      "\tExtracting 683 sequences from chr15\n",
      "\tExtracting 1191 sequences from chr16\n",
      "\tExtracting 1809 sequences from chr17\n",
      "\tExtracting 327 sequences from chr18\n",
      "\tExtracting 2310 sequences from chr19\n",
      "\tExtracting 1564 sequences from chr2\n",
      "\tExtracting 657 sequences from chr20\n",
      "\tExtracting 573 sequences from chr21\n",
      "\tExtracting 620 sequences from chr22\n",
      "\tExtracting 1322 sequences from chr3\n",
      "\tExtracting 815 sequences from chr4\n",
      "\tExtracting 992 sequences from chr5\n",
      "\tExtracting 1245 sequences from chr6\n",
      "\tExtracting 1214 sequences from chr7\n",
      "\tExtracting 811 sequences from chr8\n",
      "\tExtracting 1034 sequences from chr9\n",
      "\tExtracting 392 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24943 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset3/wellington.tsv.gz tmp_static/Subset3/homer.tsv.gz tmp_static/Subset3/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset3/expression.tsv.gz tmp_static/Subset3/wellington.tsv.gz data/gene.bed tmp_static/Subset3/tssdist.tsv.gz\n",
      "python3 -m dictys  chromatin linking  tmp_static/Subset3/binding.tsv.gz tmp_static/Subset3/tssdist.tsv.gz tmp_static/Subset3/linking.tsv.gz\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset3/linking.tsv.gz tmp_static/Subset3/binlinking.tsv.gz 20\n",
      "Reading BED File...\n",
      "Calculating footprints...\n",
      "Waiting for the last 30 jobs to finish...\n",
      "\n",
      "python3 -m dictys  chromatin homer --nth 4 tmp_static/Subset2/footprints.bed data/motifs.motif data/genome tmp_static/Subset2/expression.tsv.gz tmp_static/Subset2/motifs.bed tmp_static/Subset2/wellington.tsv.gz tmp_static/Subset2/homer.tsv.gz\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaad\n",
      "\tGenome = data/genome\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tOutput Directory = 15-motifscan/aaaad\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 24997\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 24997\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2573 sequences from chr1\n",
      "\tExtracting 1095 sequences from chr10\n",
      "\tExtracting 1369 sequences from chr11\n",
      "\tExtracting 1331 sequences from chr12\n",
      "\tExtracting 486 sequences from chr13\n",
      "\tExtracting 803 sequences from chr14\n",
      "\tExtracting 786 sequences from chr15\n",
      "\tExtracting 1075 sequences from chr16\n",
      "\tExtracting 1601 sequences from chr17\n",
      "\tExtracting 393 sequences from chr18\n",
      "\tExtracting 1750 sequences from chr19\n",
      "\tExtracting 1658 sequences from chr2\n",
      "\tExtracting 722 sequences from chr20\n",
      "\tExtracting 339 sequences from chr21\n",
      "\tExtracting 686 sequences from chr22\n",
      "\tExtracting 1410 sequences from chr3\n",
      "\tExtracting 833 sequences from chr4\n",
      "\tExtracting 1048 sequences from chr5\n",
      "\tExtracting 1416 sequences from chr6\n",
      "\tExtracting 1123 sequences from chr7\n",
      "\tExtracting 862 sequences from chr8\n",
      "\tExtracting 1059 sequences from chr9\n",
      "\tExtracting 534 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24953 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaab\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaab\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2606 sequences from chr1\n",
      "\tExtracting 1006 sequences from chr10\n",
      "\tExtracting 1329 sequences from chr11\n",
      "\tExtracting 1250 sequences from chr12\n",
      "\tExtracting 422 sequences from chr13\n",
      "\tExtracting 809 sequences from chr14\n",
      "\tExtracting 771 sequences from chr15\n",
      "\tExtracting 1161 sequences from chr16\n",
      "\tExtracting 1722 sequences from chr17\n",
      "\tExtracting 376 sequences from chr18\n",
      "\tExtracting 1959 sequences from chr19\n",
      "\tExtracting 1608 sequences from chr2\n",
      "\tExtracting 675 sequences from chr20\n",
      "\tExtracting 437 sequences from chr21\n",
      "\tExtracting 631 sequences from chr22\n",
      "\tExtracting 1366 sequences from chr3\n",
      "\tExtracting 818 sequences from chr4\n",
      "\tExtracting 1050 sequences from chr5\n",
      "\tExtracting 1293 sequences from chr6\n",
      "\tExtracting 1174 sequences from chr7\n",
      "\tExtracting 872 sequences from chr8\n",
      "\tExtracting 1081 sequences from chr9\n",
      "\tExtracting 521 sequences from chrX\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24937 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaac\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaac\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 19 (avg size of targets)\n",
      "\tBackground files for 19 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2503 sequences from chr1\n",
      "\tExtracting 1056 sequences from chr10\n",
      "\tExtracting 1346 sequences from chr11\n",
      "\tExtracting 1261 sequences from chr12\n",
      "\tExtracting 430 sequences from chr13\n",
      "\tExtracting 839 sequences from chr14\n",
      "\tExtracting 797 sequences from chr15\n",
      "\tExtracting 1123 sequences from chr16\n",
      "\tExtracting 1637 sequences from chr17\n",
      "\tExtracting 341 sequences from chr18\n",
      "\tExtracting 1800 sequences from chr19\n",
      "\tExtracting 1673 sequences from chr2\n",
      "\tExtracting 724 sequences from chr20\n",
      "\tExtracting 355 sequences from chr21\n",
      "\tExtracting 670 sequences from chr22\n",
      "\tExtracting 1432 sequences from chr3\n",
      "\tExtracting 817 sequences from chr4\n",
      "\tExtracting 1121 sequences from chr5\n",
      "\tExtracting 1353 sequences from chr6\n",
      "\tExtracting 1161 sequences from chr7\n",
      "\tExtracting 925 sequences from chr8\n",
      "\tExtracting 1056 sequences from chr9\n",
      "\tExtracting 550 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24971 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "\tPosition file = 14-reform-split/aaaaa\n",
      "\tGenome = data/genome\n",
      "\tOutput Directory = 15-motifscan/aaaaa\n",
      "\tUsing actual sizes of regions (-size given)\n",
      "\tFragment size set to given\n",
      "\tWill use repeat masked sequences\n",
      "\tWill find motif(s) in data/motifs.motif\n",
      "\tUsing Custom Genome\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 25001\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 25001\n",
      "\t\tRedundant Peak IDs: 0\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tPeak file looks good!\n",
      "\n",
      "\tBackground fragment size set to 20 (avg size of targets)\n",
      "\tBackground files for 20 bp fragments found.\n",
      "\tCustom genome sequence directory: data/genome\n",
      "\n",
      "\tExtracting sequences from file: data/genome/genome.fa\n",
      "\tLooking for peak sequences in a single file (data/genome/genome.fa)\n",
      "\tExtracting 2679 sequences from chr1\n",
      "\tExtracting 958 sequences from chr10\n",
      "\tExtracting 1280 sequences from chr11\n",
      "\tExtracting 1223 sequences from chr12\n",
      "\tExtracting 387 sequences from chr13\n",
      "\tExtracting 772 sequences from chr14\n",
      "\tExtracting 657 sequences from chr15\n",
      "\tExtracting 1229 sequences from chr16\n",
      "\tExtracting 1826 sequences from chr17\n",
      "\tExtracting 320 sequences from chr18\n",
      "\tExtracting 2415 sequences from chr19\n",
      "\tExtracting 1527 sequences from chr2\n",
      "\tExtracting 708 sequences from chr20\n",
      "\tExtracting 511 sequences from chr21\n",
      "\tExtracting 613 sequences from chr22\n",
      "\tExtracting 1333 sequences from chr3\n",
      "\tExtracting 811 sequences from chr4\n",
      "\tExtracting 971 sequences from chr5\n",
      "\tExtracting 1233 sequences from chr6\n",
      "\tExtracting 1195 sequences from chr7\n",
      "\tExtracting 827 sequences from chr8\n",
      "\tExtracting 1048 sequences from chr9\n",
      "\tExtracting 389 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\n",
      "\n",
      "\tReading input files...\n",
      "\t24913 total sequences read\n",
      "\t769 motifs loaded\n",
      "\tFinding instances of 769 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tCleaning up tmp files...\n",
      "\n",
      "\n",
      "python3 -m dictys  chromatin binding  tmp_static/Subset2/wellington.tsv.gz tmp_static/Subset2/homer.tsv.gz tmp_static/Subset2/binding.tsv.gz\n",
      "python3 -m dictys  chromatin tssdist  tmp_static/Subset2/expression.tsv.gz tmp_static/Subset2/wellington.tsv.gz data/gene.bed tmp_static/Subset2/tssdist.tsv.gz\n",
      "python3 -m dictys  chromatin linking  tmp_static/Subset2/binding.tsv.gz tmp_static/Subset2/tssdist.tsv.gz tmp_static/Subset2/linking.tsv.gz\n",
      "python3 -m dictys  chromatin binlinking  tmp_static/Subset2/linking.tsv.gz tmp_static/Subset2/binlinking.tsv.gz 20\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "make: Target `cpu' not remade because of errors.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "set -eo pipefail\n",
    "cd ..\n",
    "#Run CPU part of inference\n",
    "make -f makefiles/static.mk -j 32 -k cpu || true\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.2 CPU part 1, command description\n",
    "### preproc selects_rna\n",
    "This command subsets the RNA data separately for cells of each context/cell type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys preproc selects_rna [-h] fi_reads fi_names fo_reads\r\n",
      "\r\n",
      "Select samples/cells based on external table for RNA data.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  fi_reads    Path of input tsv file of full expression matrix\r\n",
      "  fi_names    Path of input text file of sample/cell names to select\r\n",
      "  fo_reads    Path of output tsv file of expression matrix of selected\r\n",
      "              samples/cells\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help  show this help message and exit\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys preproc selects_rna -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### preproc qc_reads\n",
    "This command removes low-read genes and cells for quality control."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys preproc qc_reads [-h]\r\n",
      "                               fi_reads fo_reads n_gene nc_gene ncp_gene\r\n",
      "                               n_cell nt_cell ntp_cell\r\n",
      "\r\n",
      "Quality control by bounding read counts. Quality control is perform separately\r\n",
      "on genes based on their cell statisics and on cells based on their gene\r\n",
      "statistics, iteratively until dataset remains unchanged. A gene or cell is\r\n",
      "removed if any of the QC criteria is violated at any time in the iteration.\r\n",
      "All QC parameters can be set to 0 to disable QC filtering for that criterion.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  fi_reads    Path of input tsv file of read count matrix. Rows are genes and\r\n",
      "              columns are cells.\r\n",
      "  fo_reads    Path of output tsv file of read count matrix after QC\r\n",
      "  n_gene      Lower bound on total read counts for gene QC\r\n",
      "  nc_gene     Lower bound on number of expressed cells for gene QC\r\n",
      "  ncp_gene    Lower bound on proportion of expressed cells for gene QC\r\n",
      "  n_cell      Lower bound on total read counts for cell QC\r\n",
      "  nt_cell     Lower bound on number of expressed genes for cell QC\r\n",
      "  ntp_cell    Lower bound on proportion of expressed genes for cell QC\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help  show this help message and exit\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys preproc qc_reads -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### preproc selects_atac\n",
    "This command combines per-cell bam files to a single bam file for each cell subset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys preproc selects_atac [-h] fi_exp fi_list fo_list\r\n",
      "\r\n",
      "Select chromatin accessibility samples/cells based on presence in expression\r\n",
      "matrix.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  fi_exp      Path of input tsv file of expression matrix. Column must be\r\n",
      "              sample/cell name.\r\n",
      "  fi_list     Path of input text file of selected cell names, one per line\r\n",
      "  fo_list     Path of output text file of selected cell names, one per line\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help  show this help message and exit\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys preproc selects_atac -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### chromatin macs2\n",
    "This command finds chromatin accessibility peaks with macs2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys chromatin macs2 [-h] [--qcut QCUT] [--nth NTH] [--nmax NMAX]\r\n",
      "                              fi_names fi_bam fo_bam fo_bai fo_bed genome_size\r\n",
      "\r\n",
      "Peak calling using macs2. Needs bam files for each cell in a given folder.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  fi_names     Path of input text file containing one sample/cell name per\r\n",
      "               line for macs2 peak calling\r\n",
      "  fi_bam       Path of input folder that contains each cell's bam file by name\r\n",
      "               in fi_names\r\n",
      "  fo_bam       Path of output bam file for select samples/cells\r\n",
      "  fo_bai       Path of output bai file for select samples/cells\r\n",
      "  fo_bed       Path of output bed file of peaks\r\n",
      "  genome_size  Genome size input of macs2. Use shortcuts hs or mm for human or\r\n",
      "               mouse.\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help   show this help message and exit\r\n",
      "  --qcut QCUT  Qvalue cutoff for macs2 (default: 0.05)\r\n",
      "  --nth NTH    Number of threads (default: 1)\r\n",
      "  --nmax NMAX  Maximum number of peaks to retain, ordered by macs2 score. Use\r\n",
      "               0 for no limit. (default: 500000)\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys chromatin macs2 -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### chromatin wellington\n",
    "This command finds transcription factor footprints with wellington/pyDNase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys chromatin wellington [-h] [--fi_blacklist FI_BLACKLIST]\r\n",
      "                                   [--cut CUT] [--nth NTH] [--nmax NMAX]\r\n",
      "                                   fi_bam fi_bai fi_bed fo_bed\r\n",
      "\r\n",
      "TF footprinting with wellington.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  fi_bam                Path of input bam file of all reads\r\n",
      "  fi_bai                Path of input bai file of all reads\r\n",
      "  fi_bed                Path of input bed file of peaks\r\n",
      "  fo_bed                Path of output bed file of footprints\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help            show this help message and exit\r\n",
      "  --fi_blacklist FI_BLACKLIST\r\n",
      "                        Path of input bed file of blacklisted genome regions\r\n",
      "                        to be removed (default: None)\r\n",
      "  --cut CUT             Cutoff for wellington score (default: 10)\r\n",
      "  --nth NTH             Number of threads (default: 1)\r\n",
      "  --nmax NMAX           Maximum number of footprints to retain, ordered by\r\n",
      "                        wellington score. Use 0 for no limit. (default:\r\n",
      "                        100000)\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys chromatin wellington -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### chromatin homer\n",
    "This command scans for motif occurrences within provided regions (here open chromatin peaks or footprints) with homer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys chromatin homer [-h] [--nth NTH]\r\n",
      "                              fi_bed fi_motif dirio_genome fi_exp fo_bed\r\n",
      "                              fo_wellington fo_homer\r\n",
      "\r\n",
      "Motif scan with homer.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  fi_bed         Path of input bed file of regions\r\n",
      "  fi_motif       Path of input motif PWM file in homer format. Motifs must be\r\n",
      "                 named in format 'gene...' where gene matches gene names in\r\n",
      "                 fi_exp. Should not contain duplicates.\r\n",
      "  dirio_genome   Path of input & output folder or file for reference genome\r\n",
      "                 for homer. A separate hard copy is recommended because homer\r\n",
      "                 may write into the folder to preparse genome.\r\n",
      "  fi_exp         Path of input expression matrix file in tsv format. Used for\r\n",
      "                 mapping motifs to genes.\r\n",
      "  fo_bed         Path of output bed file of detected motifs\r\n",
      "  fo_wellington  Path of output tsv file of wellington scores in shape\r\n",
      "                 (region,motif)\r\n",
      "  fo_homer       Path of output tsv file of homer scores in shape\r\n",
      "                 (region,motif)\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help     show this help message and exit\r\n",
      "  --nth NTH      Number of threads (default: 1)\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys chromatin homer -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### chromatin binding\n",
    "This command computes an integrative score for TF binding based on scores from footprint/peak discovery and from homer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys chromatin binding [-h] [--cuth CUTH] [--cutw CUTW] [--cut CUT]\r\n",
      "                                [--combine COMBINE] [--mode MODE]\r\n",
      "                                fi_wellington fi_homer fo_bind\r\n",
      "\r\n",
      "Finding TF binding events. Combines wellington and homer outputs to infer TF\r\n",
      "binding events by merging motifs to TFs.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  fi_wellington      Path of input tsv file of wellington output\r\n",
      "  fi_homer           Path of input tsv file of homer output\r\n",
      "  fo_bind            Path of output tsv file of binding events\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help         show this help message and exit\r\n",
      "  --cuth CUTH        Homer score cutoff (default: 0)\r\n",
      "  --cutw CUTW        Wellington score cutoff (default: 0)\r\n",
      "  --cut CUT          Final score (integrating homer & wellington) cutoff\r\n",
      "                     (default: None)\r\n",
      "  --combine COMBINE  Method to combine scores of motifs of the same TF.\r\n",
      "                     Accepts: max, mean, sum. (default: max)\r\n",
      "  --mode MODE        Mode to compute final score. Accepts binary flags: 1: Add\r\n",
      "                     log(wellington score) 2: Add log(homer score) 4: Subtract\r\n",
      "                     log(10)*(distance_to_tss)/1E6 (default: 3)\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys chromatin binding -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### chromatin tssdist\n",
    "This command computes the distance between each (open chromatin) region and each gene's transcription start site to prioritize nearby pairs that are more likely to have regulatory effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys chromatin tssdist [-h] [--cut CUT] [--nmin NMIN] [--nmax NMAX]\r\n",
      "                                fi_exp fi_wellington fi_tss fo_dist\r\n",
      "\r\n",
      "Annotating TF bond regions to target genes based on distance to TSS.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  fi_exp         Path of input expression matrix file in tsv format to obtain\r\n",
      "                 gene names\r\n",
      "  fi_wellington  Path of input tsv file of wellington scores to obtain DNA\r\n",
      "                 regions\r\n",
      "  fi_tss         Path of input bed file for gene region and strand\r\n",
      "  fo_dist        Path of output tsv file of distance from TF-bond regions to\r\n",
      "                 TSS\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help     show this help message and exit\r\n",
      "  --cut CUT      Distance cutoff between DNA region and target gene TSS\r\n",
      "                 (default: 500000)\r\n",
      "  --nmin NMIN    Minimal total number of links to recover (default: 1)\r\n",
      "  --nmax NMAX    Maximal total number of links to recover (default: 10000000)\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys chromatin tssdist -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### chromatin linking\n",
    "This command links TFs to its potential target genes via the relation: TF --- motif --- region --- gene.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys chromatin linking [-h] [--combine COMBINE] [--mode MODE]\r\n",
      "                                fi_binding fi_dist fo_linking\r\n",
      "\r\n",
      "Linking regulators and targets with scores.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  fi_binding         Path of input tsv file of binding events\r\n",
      "  fi_dist            Path of input tsv file of distance from TF-bond regions\r\n",
      "                     to TSS\r\n",
      "  fo_linking         Path of output matrix file of TF to potential target gene\r\n",
      "                     link scores\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help         show this help message and exit\r\n",
      "  --combine COMBINE  Method to combine scores of motifs of the same TF.\r\n",
      "                     Accepts: max, mean, sum. (default: max)\r\n",
      "  --mode MODE        Mode to compute final score. Accepts binary flags: 4:\r\n",
      "                     Subtract log(10)*(distance_to_tss)/1E6 (default: 4)\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys chromatin linking -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### chromatin binlinking\n",
    "This command selects the strongest TF-target gene pairs as a TF binding network that constrains the GRN to be inferred."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys chromatin binlinking [-h] [--axis AXIS] [--selfreg SELFREG]\r\n",
      "                                   [--inf INF]\r\n",
      "                                   fi_linking fo_binlinking n\r\n",
      "\r\n",
      "Converting regulator-target link score matrix to binary. Chooses the top\r\n",
      "regulator-target links, separately for each target gene by default.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  fi_linking         Path of input matrix file of TF to potential target gene\r\n",
      "                     link scores\r\n",
      "  fo_binlinking      Path of output matrix file of TF to potential target gene\r\n",
      "                     links\r\n",
      "  n                  Number of regulator-target links. n strongest links (with\r\n",
      "                     highest scores) are selected along axis axis. If greater\r\n",
      "                     than the maximum links available, all links will be\r\n",
      "                     selected subject to inf parameter constraint. Value -1\r\n",
      "                     selects all non-inf links.\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help         show this help message and exit\r\n",
      "  --axis AXIS        Axis to choose the top links. If None, n links in total\r\n",
      "                     are selected among all regulator-target links. Defaults\r\n",
      "                     to 1, indicating n strongest links/regulators are\r\n",
      "                     selected for each target. (default: 1)\r\n",
      "  --selfreg SELFREG  How to handle self regulation. Accepts: error: Raise\r\n",
      "                     ValueError if seen in fi_linking (default: error)\r\n",
      "  --inf INF          Whether to select links with -inf score. Accepts: never.\r\n",
      "                     (default: never)\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys chromatin binlinking -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1 GPU part, execution\n",
    "**This part should be skipped if you use CPU for pytorch.**\n",
    "\n",
    "This part performs GRN inference with scRNA-seq read counts on GPU."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset1/expression.tsv.gz tmp_static/Subset1/binlinking.tsv.gz tmp_static/Subset1/net_weight.tsv.gz tmp_static/Subset1/net_meanvar.tsv.gz tmp_static/Subset1/net_covfactor.tsv.gz tmp_static/Subset1/net_loss.tsv.gz tmp_static/Subset1/net_stats.tsv.gz\n",
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset10/expression.tsv.gz tmp_static/Subset10/binlinking.tsv.gz tmp_static/Subset10/net_weight.tsv.gz tmp_static/Subset10/net_meanvar.tsv.gz tmp_static/Subset10/net_covfactor.tsv.gz tmp_static/Subset10/net_loss.tsv.gz tmp_static/Subset10/net_stats.tsv.gz\n",
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset11/expression.tsv.gz tmp_static/Subset11/binlinking.tsv.gz tmp_static/Subset11/net_weight.tsv.gz tmp_static/Subset11/net_meanvar.tsv.gz tmp_static/Subset11/net_covfactor.tsv.gz tmp_static/Subset11/net_loss.tsv.gz tmp_static/Subset11/net_stats.tsv.gz\n",
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset12/expression.tsv.gz tmp_static/Subset12/binlinking.tsv.gz tmp_static/Subset12/net_weight.tsv.gz tmp_static/Subset12/net_meanvar.tsv.gz tmp_static/Subset12/net_covfactor.tsv.gz tmp_static/Subset12/net_loss.tsv.gz tmp_static/Subset12/net_stats.tsv.gz\n",
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset13/expression.tsv.gz tmp_static/Subset13/binlinking.tsv.gz tmp_static/Subset13/net_weight.tsv.gz tmp_static/Subset13/net_meanvar.tsv.gz tmp_static/Subset13/net_covfactor.tsv.gz tmp_static/Subset13/net_loss.tsv.gz tmp_static/Subset13/net_stats.tsv.gz\n",
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset14/expression.tsv.gz tmp_static/Subset14/binlinking.tsv.gz tmp_static/Subset14/net_weight.tsv.gz tmp_static/Subset14/net_meanvar.tsv.gz tmp_static/Subset14/net_covfactor.tsv.gz tmp_static/Subset14/net_loss.tsv.gz tmp_static/Subset14/net_stats.tsv.gz\n",
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset2/expression.tsv.gz tmp_static/Subset2/binlinking.tsv.gz tmp_static/Subset2/net_weight.tsv.gz tmp_static/Subset2/net_meanvar.tsv.gz tmp_static/Subset2/net_covfactor.tsv.gz tmp_static/Subset2/net_loss.tsv.gz tmp_static/Subset2/net_stats.tsv.gz\n",
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset3/expression.tsv.gz tmp_static/Subset3/binlinking.tsv.gz tmp_static/Subset3/net_weight.tsv.gz tmp_static/Subset3/net_meanvar.tsv.gz tmp_static/Subset3/net_covfactor.tsv.gz tmp_static/Subset3/net_loss.tsv.gz tmp_static/Subset3/net_stats.tsv.gz\n",
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset4/expression.tsv.gz tmp_static/Subset4/binlinking.tsv.gz tmp_static/Subset4/net_weight.tsv.gz tmp_static/Subset4/net_meanvar.tsv.gz tmp_static/Subset4/net_covfactor.tsv.gz tmp_static/Subset4/net_loss.tsv.gz tmp_static/Subset4/net_stats.tsv.gz\n",
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset5/expression.tsv.gz tmp_static/Subset5/binlinking.tsv.gz tmp_static/Subset5/net_weight.tsv.gz tmp_static/Subset5/net_meanvar.tsv.gz tmp_static/Subset5/net_covfactor.tsv.gz tmp_static/Subset5/net_loss.tsv.gz tmp_static/Subset5/net_stats.tsv.gz\n",
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset6/expression.tsv.gz tmp_static/Subset6/binlinking.tsv.gz tmp_static/Subset6/net_weight.tsv.gz tmp_static/Subset6/net_meanvar.tsv.gz tmp_static/Subset6/net_covfactor.tsv.gz tmp_static/Subset6/net_loss.tsv.gz tmp_static/Subset6/net_stats.tsv.gz\n",
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset7/expression.tsv.gz tmp_static/Subset7/binlinking.tsv.gz tmp_static/Subset7/net_weight.tsv.gz tmp_static/Subset7/net_meanvar.tsv.gz tmp_static/Subset7/net_covfactor.tsv.gz tmp_static/Subset7/net_loss.tsv.gz tmp_static/Subset7/net_stats.tsv.gz\n",
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset8/expression.tsv.gz tmp_static/Subset8/binlinking.tsv.gz tmp_static/Subset8/net_weight.tsv.gz tmp_static/Subset8/net_meanvar.tsv.gz tmp_static/Subset8/net_covfactor.tsv.gz tmp_static/Subset8/net_loss.tsv.gz tmp_static/Subset8/net_stats.tsv.gz\n",
      "python3 -m dictys  network reconstruct --device cuda:0 --nth 4 tmp_static/Subset9/expression.tsv.gz tmp_static/Subset9/binlinking.tsv.gz tmp_static/Subset9/net_weight.tsv.gz tmp_static/Subset9/net_meanvar.tsv.gz tmp_static/Subset9/net_covfactor.tsv.gz tmp_static/Subset9/net_loss.tsv.gz tmp_static/Subset9/net_stats.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "set -eo pipefail\n",
    "cd ..\n",
    "make -f makefiles/static.mk -j 2 -k gpu || true\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 GPU part, command description\n",
    "### network reconstruct\n",
    "This command uses pyro and pytorch to infer the GRN with stochastic process model under the TF binding network constraint."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys network reconstruct [-h] [--lr LR] [--lrd LRD] [--nstep NSTEP]\r\n",
      "                                  [--npc NPC] [--fi_cov FI_COV]\r\n",
      "                                  [--model MODEL]\r\n",
      "                                  [--nstep_report NSTEP_REPORT]\r\n",
      "                                  [--rseed RSEED] [--device DEVICE]\r\n",
      "                                  [--dtype DTYPE] [--loss LOSS] [--nth NTH]\r\n",
      "                                  [--varmean VARMEAN] [--varstd VARSTD]\r\n",
      "                                  [--fo_weightz FO_WEIGHTZ]\r\n",
      "                                  [--scale_lyapunov SCALE_LYAPUNOV]\r\n",
      "                                  fi_exp fi_mask fo_weight fo_meanvar\r\n",
      "                                  fo_covfactor fo_loss fo_stats\r\n",
      "\r\n",
      "Reconstruct network with any pyro model in net_pyro_models that is based on\r\n",
      "covariance_model and has binary masks.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  fi_exp                Path of input tsv file of expression matrix.\r\n",
      "  fi_mask               Path of input tsv file of mask matrix indicating which\r\n",
      "                        edges are allowed. Can be output of dictys chromatin\r\n",
      "                        binlinking.\r\n",
      "  fo_weight             Path of output tsv file of edge weight matrix\r\n",
      "  fo_meanvar            Path of output tsv file of mean and variance of each\r\n",
      "                        gene's relative log expression\r\n",
      "  fo_covfactor          Path of output tsv file of factors for the off-\r\n",
      "                        diagonal component of gene covariance matrix\r\n",
      "  fo_loss               Path of output tsv file of sublosses in each training\r\n",
      "                        step\r\n",
      "  fo_stats              Path of output tsv file of basic stats of each\r\n",
      "                        variable during training\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help            show this help message and exit\r\n",
      "  --lr LR               Initial iearning rate (default: 0.01)\r\n",
      "  --lrd LRD             Learning rate decay (default: 0.999)\r\n",
      "  --nstep NSTEP         Number of training steps (default: 4000)\r\n",
      "  --npc NPC             Number of unknown factors for covariance in\r\n",
      "                        multivariate distribution (default: 0)\r\n",
      "  --fi_cov FI_COV       Path of input tsv file of covariate matrix for each\r\n",
      "                        cell to be included. Should have covariate x cell\r\n",
      "                        shape. (default: None)\r\n",
      "  --model MODEL         Name of model to train (default: ou)\r\n",
      "  --nstep_report NSTEP_REPORT\r\n",
      "                        Number of steps to save each parameter distribution\r\n",
      "                        (default: 100)\r\n",
      "  --rseed RSEED         Initial random seed (default: 12345)\r\n",
      "  --device DEVICE       Device for pytorch and pyro. See TBA.... (default:\r\n",
      "                        cpu)\r\n",
      "  --dtype DTYPE         Data type for pytorch. See TBA... (default: float)\r\n",
      "  --loss LOSS           Loss function name (default: Trace_ELBO_site)\r\n",
      "  --nth NTH             Number of threads for CPU usage. When <1, use\r\n",
      "                        nth*(detected core count). (default: 1)\r\n",
      "  --varmean VARMEAN     Pyro parameter name for mean of effect size (default:\r\n",
      "                        N_0val)\r\n",
      "  --varstd VARSTD       Pyro parameter name for std of effect size to compute\r\n",
      "                        z scores. Use None if unavailable. (default: None)\r\n",
      "  --fo_weightz FO_WEIGHTZ\r\n",
      "                        Path of output tsv file of z score matrix of edge\r\n",
      "                        weights (default: None)\r\n",
      "  --scale_lyapunov SCALE_LYAPUNOV\r\n",
      "                        Scale of Lyapunov equation inequality loss (default:\r\n",
      "                        100000.0)\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys network reconstruct -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.1 CPU part 2, execution\n",
    "**This part should be skipped if you use CPU for pytorch.**\n",
    "\n",
    "This part performs network postprocessing to address variance estimation bias and indirect effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset1/net_weight.tsv.gz tmp_static/Subset1/net_meanvar.tsv.gz tmp_static/Subset1/net_covfactor.tsv.gz tmp_static/Subset1/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset1/net_meanvar.tsv.gz tmp_static/Subset1/net_weight.tsv.gz tmp_static/Subset1/net_covfactor.tsv.gz tmp_static/Subset1/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset10/net_weight.tsv.gz tmp_static/Subset10/net_meanvar.tsv.gz tmp_static/Subset10/net_covfactor.tsv.gz tmp_static/Subset10/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset10/net_meanvar.tsv.gz tmp_static/Subset10/net_weight.tsv.gz tmp_static/Subset10/net_covfactor.tsv.gz tmp_static/Subset10/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset11/net_weight.tsv.gz tmp_static/Subset11/net_meanvar.tsv.gz tmp_static/Subset11/net_covfactor.tsv.gz tmp_static/Subset11/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset11/net_meanvar.tsv.gz tmp_static/Subset11/net_weight.tsv.gz tmp_static/Subset11/net_covfactor.tsv.gz tmp_static/Subset11/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset12/net_weight.tsv.gz tmp_static/Subset12/net_meanvar.tsv.gz tmp_static/Subset12/net_covfactor.tsv.gz tmp_static/Subset12/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset12/net_meanvar.tsv.gz tmp_static/Subset12/net_weight.tsv.gz tmp_static/Subset12/net_covfactor.tsv.gz tmp_static/Subset12/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset13/net_weight.tsv.gz tmp_static/Subset13/net_meanvar.tsv.gz tmp_static/Subset13/net_covfactor.tsv.gz tmp_static/Subset13/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset13/net_meanvar.tsv.gz tmp_static/Subset13/net_weight.tsv.gz tmp_static/Subset13/net_covfactor.tsv.gz tmp_static/Subset13/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset14/net_weight.tsv.gz tmp_static/Subset14/net_meanvar.tsv.gz tmp_static/Subset14/net_covfactor.tsv.gz tmp_static/Subset14/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset14/net_meanvar.tsv.gz tmp_static/Subset14/net_weight.tsv.gz tmp_static/Subset14/net_covfactor.tsv.gz tmp_static/Subset14/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset2/net_weight.tsv.gz tmp_static/Subset2/net_meanvar.tsv.gz tmp_static/Subset2/net_covfactor.tsv.gz tmp_static/Subset2/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset2/net_meanvar.tsv.gz tmp_static/Subset2/net_weight.tsv.gz tmp_static/Subset2/net_covfactor.tsv.gz tmp_static/Subset2/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset3/net_weight.tsv.gz tmp_static/Subset3/net_meanvar.tsv.gz tmp_static/Subset3/net_covfactor.tsv.gz tmp_static/Subset3/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset3/net_meanvar.tsv.gz tmp_static/Subset3/net_weight.tsv.gz tmp_static/Subset3/net_covfactor.tsv.gz tmp_static/Subset3/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset4/net_weight.tsv.gz tmp_static/Subset4/net_meanvar.tsv.gz tmp_static/Subset4/net_covfactor.tsv.gz tmp_static/Subset4/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset4/net_meanvar.tsv.gz tmp_static/Subset4/net_weight.tsv.gz tmp_static/Subset4/net_covfactor.tsv.gz tmp_static/Subset4/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset5/net_weight.tsv.gz tmp_static/Subset5/net_meanvar.tsv.gz tmp_static/Subset5/net_covfactor.tsv.gz tmp_static/Subset5/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset5/net_meanvar.tsv.gz tmp_static/Subset5/net_weight.tsv.gz tmp_static/Subset5/net_covfactor.tsv.gz tmp_static/Subset5/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset6/net_weight.tsv.gz tmp_static/Subset6/net_meanvar.tsv.gz tmp_static/Subset6/net_covfactor.tsv.gz tmp_static/Subset6/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset6/net_meanvar.tsv.gz tmp_static/Subset6/net_weight.tsv.gz tmp_static/Subset6/net_covfactor.tsv.gz tmp_static/Subset6/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset7/net_weight.tsv.gz tmp_static/Subset7/net_meanvar.tsv.gz tmp_static/Subset7/net_covfactor.tsv.gz tmp_static/Subset7/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset7/net_meanvar.tsv.gz tmp_static/Subset7/net_weight.tsv.gz tmp_static/Subset7/net_covfactor.tsv.gz tmp_static/Subset7/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset8/net_weight.tsv.gz tmp_static/Subset8/net_meanvar.tsv.gz tmp_static/Subset8/net_covfactor.tsv.gz tmp_static/Subset8/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset8/net_meanvar.tsv.gz tmp_static/Subset8/net_weight.tsv.gz tmp_static/Subset8/net_covfactor.tsv.gz tmp_static/Subset8/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset9/net_weight.tsv.gz tmp_static/Subset9/net_meanvar.tsv.gz tmp_static/Subset9/net_covfactor.tsv.gz tmp_static/Subset9/net_nweight.tsv.gz\n",
      "python3 -m dictys  network indirect --nth 4 --fi_meanvar tmp_static/Subset9/net_meanvar.tsv.gz tmp_static/Subset9/net_weight.tsv.gz tmp_static/Subset9/net_covfactor.tsv.gz tmp_static/Subset9/net_iweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset13/net_iweight.tsv.gz tmp_static/Subset13/net_meanvar.tsv.gz tmp_static/Subset13/net_covfactor.tsv.gz tmp_static/Subset13/net_inweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset14/net_iweight.tsv.gz tmp_static/Subset14/net_meanvar.tsv.gz tmp_static/Subset14/net_covfactor.tsv.gz tmp_static/Subset14/net_inweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset11/net_iweight.tsv.gz tmp_static/Subset11/net_meanvar.tsv.gz tmp_static/Subset11/net_covfactor.tsv.gz tmp_static/Subset11/net_inweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset5/net_iweight.tsv.gz tmp_static/Subset5/net_meanvar.tsv.gz tmp_static/Subset5/net_covfactor.tsv.gz tmp_static/Subset5/net_inweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset12/net_iweight.tsv.gz tmp_static/Subset12/net_meanvar.tsv.gz tmp_static/Subset12/net_covfactor.tsv.gz tmp_static/Subset12/net_inweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset4/net_iweight.tsv.gz tmp_static/Subset4/net_meanvar.tsv.gz tmp_static/Subset4/net_covfactor.tsv.gz tmp_static/Subset4/net_inweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset8/net_iweight.tsv.gz tmp_static/Subset8/net_meanvar.tsv.gz tmp_static/Subset8/net_covfactor.tsv.gz tmp_static/Subset8/net_inweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset6/net_iweight.tsv.gz tmp_static/Subset6/net_meanvar.tsv.gz tmp_static/Subset6/net_covfactor.tsv.gz tmp_static/Subset6/net_inweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset9/net_iweight.tsv.gz tmp_static/Subset9/net_meanvar.tsv.gz tmp_static/Subset9/net_covfactor.tsv.gz tmp_static/Subset9/net_inweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset3/net_iweight.tsv.gz tmp_static/Subset3/net_meanvar.tsv.gz tmp_static/Subset3/net_covfactor.tsv.gz tmp_static/Subset3/net_inweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset2/net_iweight.tsv.gz tmp_static/Subset2/net_meanvar.tsv.gz tmp_static/Subset2/net_covfactor.tsv.gz tmp_static/Subset2/net_inweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset1/net_iweight.tsv.gz tmp_static/Subset1/net_meanvar.tsv.gz tmp_static/Subset1/net_covfactor.tsv.gz tmp_static/Subset1/net_inweight.tsv.gz\n",
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset10/net_iweight.tsv.gz tmp_static/Subset10/net_meanvar.tsv.gz tmp_static/Subset10/net_covfactor.tsv.gz tmp_static/Subset10/net_inweight.tsv.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python3 -m dictys  network normalize --nth 4 tmp_static/Subset7/net_iweight.tsv.gz tmp_static/Subset7/net_meanvar.tsv.gz tmp_static/Subset7/net_covfactor.tsv.gz tmp_static/Subset7/net_inweight.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "set -eo pipefail\n",
    "cd ..\n",
    "make -f makefiles/static.mk -j 32 -k cpu || true\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.2 CPU part 2, command description\n",
    "### network normalize\n",
    "This command normalizes network edges based on the standard deviation of regulator and target genes. This can overcome biases in the estimation of variance of gene expression due to single-cell sparsity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys network normalize [-h] [--norm NORM] [--nth NTH]\r\n",
      "                                fi_weight fi_meanvar fi_covfactor fo_nweight\r\n",
      "\r\n",
      "Normalize edge strength. So they are more resistant to estimation bias of true\r\n",
      "expression level variance.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  fi_weight     Path of input tsv file of edge weight matrix\r\n",
      "  fi_meanvar    Path of iput tsv file of mean and variance of each gene's\r\n",
      "                relative log expression\r\n",
      "  fi_covfactor  Path of iput tsv file of factors for the off-diagonal\r\n",
      "                component of gene covariance matrix\r\n",
      "  fo_nweight    Path of output tsv file of normalized edge weight matrix\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help    show this help message and exit\r\n",
      "  --norm NORM   Type of normalization as binary flag values. Accepts: 1:\r\n",
      "                Multiplying edge weight with stochastic noise std of TF 2:\r\n",
      "                Dividing edge weight with stochastic noise std of target\r\n",
      "                (default: 3)\r\n",
      "  --nth NTH     Number of threads for CPU usage. When <1, use nth*(detected\r\n",
      "                core count). (default: 1)\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys network normalize -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### network indirect\n",
    "This command computes the steady-state total effect (direct + indirect effects) networks from the kinetic direct effect network. The total effect network is **not yet integrated into network analysis functions**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys network indirect [-h] [--norm NORM] [--fi_meanvar FI_MEANVAR]\r\n",
      "                               [--eigmin EIGMIN] [--eigmax EIGMAX]\r\n",
      "                               [--multiplier MULTIPLIER] [--nth NTH]\r\n",
      "                               fi_weight fi_covfactor fo_iweight\r\n",
      "\r\n",
      "Computes steady-state indirect effect of gene perturbation from OU process.\r\n",
      "Performs extra regularization on network by bounding the eigenvalues of\r\n",
      "feedback loops with parameters eigmin and eigmax. Values away from 1 indicates\r\n",
      "stronger feedback loop effects. Set values closer to 1 to apply stronger\r\n",
      "regularization.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  fi_weight             Path of input tsv file of edge weight matrix\r\n",
      "  fi_covfactor          Path of iput tsv file of factors for the off-diagonal\r\n",
      "                        component of gene covariance matrix\r\n",
      "  fo_iweight            Path of output tsv file of steady-state indirect\r\n",
      "                        effect edge weight matrix\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help            show this help message and exit\r\n",
      "  --norm NORM           Normalizations of indirect effect matrix. Binary flag\r\n",
      "                        variable that accepts: 1: Divides the indirect effect\r\n",
      "                        size of each TF on other genes with that on itself to\r\n",
      "                        obtain logFC per unit logFC of TF. 2: Accounts for\r\n",
      "                        mRNA saturation/sampling effects. (default: 3)\r\n",
      "  --fi_meanvar FI_MEANVAR\r\n",
      "                        Path of iput tsv file of mean and variance of each\r\n",
      "                        gene's relative log expression. Needed if norm & 2.\r\n",
      "                        (default: None)\r\n",
      "  --eigmin EIGMIN       Lower bound of eigenvalue for finding the minimal\r\n",
      "                        required extra regularization. (default: 0.2)\r\n",
      "  --eigmax EIGMAX       Upper bound of eigenvalue for finding the minimal\r\n",
      "                        required extra regularization. (default: 1.8)\r\n",
      "  --multiplier MULTIPLIER\r\n",
      "                        Step size of strengthening regularization. Larger\r\n",
      "                        values provide faster but over-regularization. Must be\r\n",
      "                        greater than 1. (default: 1.1)\r\n",
      "  --nth NTH             Number of threads for CPU usage. When <1, use\r\n",
      "                        nth*(detected core count). (default: 1)\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys network indirect -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.1 Aggregating network part, execution\n",
    "This part aggregates all the inferred networks into a single h5 file as output. This single file can be shared and processed by network analysis/visualization functions in Dictys."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir -p output/\n",
      "python3 -m dictys  network tofile  data tmp_static data/subsets.txt output/static.h5\n",
      "rm -f tmp_static/Subset1/names_rna.txt tmp_static/Subset1/names_atac0.txt tmp_static/Subset1/names_atac.txt tmp_static/Subset1/expression0.tsv.gz tmp_static/Subset1/expression.tsv.gz tmp_static/Subset1/names_atac.txt tmp_static/Subset1/reads.bam tmp_static/Subset1/reads.bai tmp_static/Subset1/peaks.bed tmp_static/Subset1/footprints.bed tmp_static/Subset1/motifs.bed tmp_static/Subset1/homer.tsv.gz tmp_static/Subset1/wellington.tsv.gz tmp_static/Subset1/binding.tsv.gz tmp_static/Subset1/tssdist.tsv.gz tmp_static/Subset1/linking.tsv.gz tmp_static/Subset1/binlinking.tsv.gz tmp_static/Subset1/net_nweight.tsv.gz tmp_static/Subset1/net_iweight.tsv.gz tmp_static/Subset1/net_inweight.tsv.gz tmp_static/Subset10/names_rna.txt tmp_static/Subset10/names_atac0.txt tmp_static/Subset10/names_atac.txt tmp_static/Subset10/expression0.tsv.gz tmp_static/Subset10/expression.tsv.gz tmp_static/Subset10/names_atac.txt tmp_static/Subset10/reads.bam tmp_static/Subset10/reads.bai tmp_static/Subset10/peaks.bed tmp_static/Subset10/footprints.bed tmp_static/Subset10/motifs.bed tmp_static/Subset10/homer.tsv.gz tmp_static/Subset10/wellington.tsv.gz tmp_static/Subset10/binding.tsv.gz tmp_static/Subset10/tssdist.tsv.gz tmp_static/Subset10/linking.tsv.gz tmp_static/Subset10/binlinking.tsv.gz tmp_static/Subset10/net_nweight.tsv.gz tmp_static/Subset10/net_iweight.tsv.gz tmp_static/Subset10/net_inweight.tsv.gz tmp_static/Subset11/names_rna.txt tmp_static/Subset11/names_atac0.txt tmp_static/Subset11/names_atac.txt tmp_static/Subset11/expression0.tsv.gz tmp_static/Subset11/expression.tsv.gz tmp_static/Subset11/names_atac.txt tmp_static/Subset11/reads.bam tmp_static/Subset11/reads.bai tmp_static/Subset11/peaks.bed tmp_static/Subset11/footprints.bed tmp_static/Subset11/motifs.bed tmp_static/Subset11/homer.tsv.gz tmp_static/Subset11/wellington.tsv.gz tmp_static/Subset11/binding.tsv.gz tmp_static/Subset11/tssdist.tsv.gz tmp_static/Subset11/linking.tsv.gz tmp_static/Subset11/binlinking.tsv.gz tmp_static/Subset11/net_nweight.tsv.gz tmp_static/Subset11/net_iweight.tsv.gz tmp_static/Subset11/net_inweight.tsv.gz tmp_static/Subset12/names_rna.txt tmp_static/Subset12/names_atac0.txt tmp_static/Subset12/names_atac.txt tmp_static/Subset12/expression0.tsv.gz tmp_static/Subset12/expression.tsv.gz tmp_static/Subset12/names_atac.txt tmp_static/Subset12/reads.bam tmp_static/Subset12/reads.bai tmp_static/Subset12/peaks.bed tmp_static/Subset12/footprints.bed tmp_static/Subset12/motifs.bed tmp_static/Subset12/homer.tsv.gz tmp_static/Subset12/wellington.tsv.gz tmp_static/Subset12/binding.tsv.gz tmp_static/Subset12/tssdist.tsv.gz tmp_static/Subset12/linking.tsv.gz tmp_static/Subset12/binlinking.tsv.gz tmp_static/Subset12/net_nweight.tsv.gz tmp_static/Subset12/net_iweight.tsv.gz tmp_static/Subset12/net_inweight.tsv.gz tmp_static/Subset13/names_rna.txt tmp_static/Subset13/names_atac0.txt tmp_static/Subset13/names_atac.txt tmp_static/Subset13/expression0.tsv.gz tmp_static/Subset13/expression.tsv.gz tmp_static/Subset13/names_atac.txt tmp_static/Subset13/reads.bam tmp_static/Subset13/reads.bai tmp_static/Subset13/peaks.bed tmp_static/Subset13/footprints.bed tmp_static/Subset13/motifs.bed tmp_static/Subset13/homer.tsv.gz tmp_static/Subset13/wellington.tsv.gz tmp_static/Subset13/binding.tsv.gz tmp_static/Subset13/tssdist.tsv.gz tmp_static/Subset13/linking.tsv.gz tmp_static/Subset13/binlinking.tsv.gz tmp_static/Subset13/net_nweight.tsv.gz tmp_static/Subset13/net_iweight.tsv.gz tmp_static/Subset13/net_inweight.tsv.gz tmp_static/Subset14/names_rna.txt tmp_static/Subset14/names_atac0.txt tmp_static/Subset14/names_atac.txt tmp_static/Subset14/expression0.tsv.gz tmp_static/Subset14/expression.tsv.gz tmp_static/Subset14/names_atac.txt tmp_static/Subset14/reads.bam tmp_static/Subset14/reads.bai tmp_static/Subset14/peaks.bed tmp_static/Subset14/footprints.bed tmp_static/Subset14/motifs.bed tmp_static/Subset14/homer.tsv.gz tmp_static/Subset14/wellington.tsv.gz tmp_static/Subset14/binding.tsv.gz tmp_static/Subset14/tssdist.tsv.gz tmp_static/Subset14/linking.tsv.gz tmp_static/Subset14/binlinking.tsv.gz tmp_static/Subset14/net_nweight.tsv.gz tmp_static/Subset14/net_iweight.tsv.gz tmp_static/Subset14/net_inweight.tsv.gz tmp_static/Subset2/names_rna.txt tmp_static/Subset2/names_atac0.txt tmp_static/Subset2/names_atac.txt tmp_static/Subset2/expression0.tsv.gz tmp_static/Subset2/expression.tsv.gz tmp_static/Subset2/names_atac.txt tmp_static/Subset2/reads.bam tmp_static/Subset2/reads.bai tmp_static/Subset2/peaks.bed tmp_static/Subset2/footprints.bed tmp_static/Subset2/motifs.bed tmp_static/Subset2/homer.tsv.gz tmp_static/Subset2/wellington.tsv.gz tmp_static/Subset2/binding.tsv.gz tmp_static/Subset2/tssdist.tsv.gz tmp_static/Subset2/linking.tsv.gz tmp_static/Subset2/binlinking.tsv.gz tmp_static/Subset2/net_nweight.tsv.gz tmp_static/Subset2/net_iweight.tsv.gz tmp_static/Subset2/net_inweight.tsv.gz tmp_static/Subset3/names_rna.txt tmp_static/Subset3/names_atac0.txt tmp_static/Subset3/names_atac.txt tmp_static/Subset3/expression0.tsv.gz tmp_static/Subset3/expression.tsv.gz tmp_static/Subset3/names_atac.txt tmp_static/Subset3/reads.bam tmp_static/Subset3/reads.bai tmp_static/Subset3/peaks.bed tmp_static/Subset3/footprints.bed tmp_static/Subset3/motifs.bed tmp_static/Subset3/homer.tsv.gz tmp_static/Subset3/wellington.tsv.gz tmp_static/Subset3/binding.tsv.gz tmp_static/Subset3/tssdist.tsv.gz tmp_static/Subset3/linking.tsv.gz tmp_static/Subset3/binlinking.tsv.gz tmp_static/Subset3/net_nweight.tsv.gz tmp_static/Subset3/net_iweight.tsv.gz tmp_static/Subset3/net_inweight.tsv.gz tmp_static/Subset4/names_rna.txt tmp_static/Subset4/names_atac0.txt tmp_static/Subset4/names_atac.txt tmp_static/Subset4/expression0.tsv.gz tmp_static/Subset4/expression.tsv.gz tmp_static/Subset4/names_atac.txt tmp_static/Subset4/reads.bam tmp_static/Subset4/reads.bai tmp_static/Subset4/peaks.bed tmp_static/Subset4/footprints.bed tmp_static/Subset4/motifs.bed tmp_static/Subset4/homer.tsv.gz tmp_static/Subset4/wellington.tsv.gz tmp_static/Subset4/binding.tsv.gz tmp_static/Subset4/tssdist.tsv.gz tmp_static/Subset4/linking.tsv.gz tmp_static/Subset4/binlinking.tsv.gz tmp_static/Subset4/net_nweight.tsv.gz tmp_static/Subset4/net_iweight.tsv.gz tmp_static/Subset4/net_inweight.tsv.gz tmp_static/Subset5/names_rna.txt tmp_static/Subset5/names_atac0.txt tmp_static/Subset5/names_atac.txt tmp_static/Subset5/expression0.tsv.gz tmp_static/Subset5/expression.tsv.gz tmp_static/Subset5/names_atac.txt tmp_static/Subset5/reads.bam tmp_static/Subset5/reads.bai tmp_static/Subset5/peaks.bed tmp_static/Subset5/footprints.bed tmp_static/Subset5/motifs.bed tmp_static/Subset5/homer.tsv.gz tmp_static/Subset5/wellington.tsv.gz tmp_static/Subset5/binding.tsv.gz tmp_static/Subset5/tssdist.tsv.gz tmp_static/Subset5/linking.tsv.gz tmp_static/Subset5/binlinking.tsv.gz tmp_static/Subset5/net_nweight.tsv.gz tmp_static/Subset5/net_iweight.tsv.gz tmp_static/Subset5/net_inweight.tsv.gz tmp_static/Subset6/names_rna.txt tmp_static/Subset6/names_atac0.txt tmp_static/Subset6/names_atac.txt tmp_static/Subset6/expression0.tsv.gz tmp_static/Subset6/expression.tsv.gz tmp_static/Subset6/names_atac.txt tmp_static/Subset6/reads.bam tmp_static/Subset6/reads.bai tmp_static/Subset6/peaks.bed tmp_static/Subset6/footprints.bed tmp_static/Subset6/motifs.bed tmp_static/Subset6/homer.tsv.gz tmp_static/Subset6/wellington.tsv.gz tmp_static/Subset6/binding.tsv.gz tmp_static/Subset6/tssdist.tsv.gz tmp_static/Subset6/linking.tsv.gz tmp_static/Subset6/binlinking.tsv.gz tmp_static/Subset6/net_nweight.tsv.gz tmp_static/Subset6/net_iweight.tsv.gz tmp_static/Subset6/net_inweight.tsv.gz tmp_static/Subset7/names_rna.txt tmp_static/Subset7/names_atac0.txt tmp_static/Subset7/names_atac.txt tmp_static/Subset7/expression0.tsv.gz tmp_static/Subset7/expression.tsv.gz tmp_static/Subset7/names_atac.txt tmp_static/Subset7/reads.bam tmp_static/Subset7/reads.bai tmp_static/Subset7/peaks.bed tmp_static/Subset7/footprints.bed tmp_static/Subset7/motifs.bed tmp_static/Subset7/homer.tsv.gz tmp_static/Subset7/wellington.tsv.gz tmp_static/Subset7/binding.tsv.gz tmp_static/Subset7/tssdist.tsv.gz tmp_static/Subset7/linking.tsv.gz tmp_static/Subset7/binlinking.tsv.gz tmp_static/Subset7/net_nweight.tsv.gz tmp_static/Subset7/net_iweight.tsv.gz tmp_static/Subset7/net_inweight.tsv.gz tmp_static/Subset8/names_rna.txt tmp_static/Subset8/names_atac0.txt tmp_static/Subset8/names_atac.txt tmp_static/Subset8/expression0.tsv.gz tmp_static/Subset8/expression.tsv.gz tmp_static/Subset8/names_atac.txt tmp_static/Subset8/reads.bam tmp_static/Subset8/reads.bai tmp_static/Subset8/peaks.bed tmp_static/Subset8/footprints.bed tmp_static/Subset8/motifs.bed tmp_static/Subset8/homer.tsv.gz tmp_static/Subset8/wellington.tsv.gz tmp_static/Subset8/binding.tsv.gz tmp_static/Subset8/tssdist.tsv.gz tmp_static/Subset8/linking.tsv.gz tmp_static/Subset8/binlinking.tsv.gz tmp_static/Subset8/net_nweight.tsv.gz tmp_static/Subset8/net_iweight.tsv.gz tmp_static/Subset8/net_inweight.tsv.gz tmp_static/Subset9/names_rna.txt tmp_static/Subset9/names_atac0.txt tmp_static/Subset9/names_atac.txt tmp_static/Subset9/expression0.tsv.gz tmp_static/Subset9/expression.tsv.gz tmp_static/Subset9/names_atac.txt tmp_static/Subset9/reads.bam tmp_static/Subset9/reads.bai tmp_static/Subset9/peaks.bed tmp_static/Subset9/footprints.bed tmp_static/Subset9/motifs.bed tmp_static/Subset9/homer.tsv.gz tmp_static/Subset9/wellington.tsv.gz tmp_static/Subset9/binding.tsv.gz tmp_static/Subset9/tssdist.tsv.gz tmp_static/Subset9/linking.tsv.gz tmp_static/Subset9/binlinking.tsv.gz tmp_static/Subset9/net_nweight.tsv.gz tmp_static/Subset9/net_iweight.tsv.gz tmp_static/Subset9/net_inweight.tsv.gz tmp_static/Subset1/net_weight.tsv.gz tmp_static/Subset1/net_meanvar.tsv.gz tmp_static/Subset1/net_covfactor.tsv.gz tmp_static/Subset1/net_loss.tsv.gz tmp_static/Subset1/net_stats.tsv.gz tmp_static/Subset10/net_weight.tsv.gz tmp_static/Subset10/net_meanvar.tsv.gz tmp_static/Subset10/net_covfactor.tsv.gz tmp_static/Subset10/net_loss.tsv.gz tmp_static/Subset10/net_stats.tsv.gz tmp_static/Subset11/net_weight.tsv.gz tmp_static/Subset11/net_meanvar.tsv.gz tmp_static/Subset11/net_covfactor.tsv.gz tmp_static/Subset11/net_loss.tsv.gz tmp_static/Subset11/net_stats.tsv.gz tmp_static/Subset12/net_weight.tsv.gz tmp_static/Subset12/net_meanvar.tsv.gz tmp_static/Subset12/net_covfactor.tsv.gz tmp_static/Subset12/net_loss.tsv.gz tmp_static/Subset12/net_stats.tsv.gz tmp_static/Subset13/net_weight.tsv.gz tmp_static/Subset13/net_meanvar.tsv.gz tmp_static/Subset13/net_covfactor.tsv.gz tmp_static/Subset13/net_loss.tsv.gz tmp_static/Subset13/net_stats.tsv.gz tmp_static/Subset14/net_weight.tsv.gz tmp_static/Subset14/net_meanvar.tsv.gz tmp_static/Subset14/net_covfactor.tsv.gz tmp_static/Subset14/net_loss.tsv.gz tmp_static/Subset14/net_stats.tsv.gz tmp_static/Subset2/net_weight.tsv.gz tmp_static/Subset2/net_meanvar.tsv.gz tmp_static/Subset2/net_covfactor.tsv.gz tmp_static/Subset2/net_loss.tsv.gz tmp_static/Subset2/net_stats.tsv.gz tmp_static/Subset3/net_weight.tsv.gz tmp_static/Subset3/net_meanvar.tsv.gz tmp_static/Subset3/net_covfactor.tsv.gz tmp_static/Subset3/net_loss.tsv.gz tmp_static/Subset3/net_stats.tsv.gz tmp_static/Subset4/net_weight.tsv.gz tmp_static/Subset4/net_meanvar.tsv.gz tmp_static/Subset4/net_covfactor.tsv.gz tmp_static/Subset4/net_loss.tsv.gz tmp_static/Subset4/net_stats.tsv.gz tmp_static/Subset5/net_weight.tsv.gz tmp_static/Subset5/net_meanvar.tsv.gz tmp_static/Subset5/net_covfactor.tsv.gz tmp_static/Subset5/net_loss.tsv.gz tmp_static/Subset5/net_stats.tsv.gz tmp_static/Subset6/net_weight.tsv.gz tmp_static/Subset6/net_meanvar.tsv.gz tmp_static/Subset6/net_covfactor.tsv.gz tmp_static/Subset6/net_loss.tsv.gz tmp_static/Subset6/net_stats.tsv.gz tmp_static/Subset7/net_weight.tsv.gz tmp_static/Subset7/net_meanvar.tsv.gz tmp_static/Subset7/net_covfactor.tsv.gz tmp_static/Subset7/net_loss.tsv.gz tmp_static/Subset7/net_stats.tsv.gz tmp_static/Subset8/net_weight.tsv.gz tmp_static/Subset8/net_meanvar.tsv.gz tmp_static/Subset8/net_covfactor.tsv.gz tmp_static/Subset8/net_loss.tsv.gz tmp_static/Subset8/net_stats.tsv.gz tmp_static/Subset9/net_weight.tsv.gz tmp_static/Subset9/net_meanvar.tsv.gz tmp_static/Subset9/net_covfactor.tsv.gz tmp_static/Subset9/net_loss.tsv.gz tmp_static/Subset9/net_stats.tsv.gz               \n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "set -eo pipefail\n",
    "cd ..\n",
    "#Combine inferred networks to single h5 file\n",
    "make -f makefiles/static.mk combine\n",
    "#Optional step: Cleanup intermediate files\n",
    "make -f makefiles/static.mk clean\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.2 Aggregating network part, command description\n",
    "### network tofile\n",
    "This command aggregates all inferred networks to a single output file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: dictys network tofile [-h] [--dynamic] [--nettype NETTYPE]\r\n",
      "                             [--optional OPTIONAL] [--fi_c FI_C]\r\n",
      "                             diri_data diri_work fi_subsets fo_networks\r\n",
      "\r\n",
      "Saving networks to a single file.\r\n",
      "\r\n",
      "positional arguments:\r\n",
      "  diri_data            Path of input data folder to load from\r\n",
      "  diri_work            Path of input working folder to load from\r\n",
      "  fi_subsets           Path of input txt file for cell subset names\r\n",
      "  fo_networks          Path of output h5 file for all networks\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help           show this help message and exit\r\n",
      "  --dynamic            Whether to load a dynamic network instead of a set of\r\n",
      "                       static networks (default: False)\r\n",
      "  --nettype NETTYPE    Type of network. Accepts: '': Unnormalized direct\r\n",
      "                       network 'n': Normalized direct network 'i':\r\n",
      "                       Unnormalized steady-state network 'in': Normalized\r\n",
      "                       steady-state network (default: n)\r\n",
      "  --optional OPTIONAL  Optional data to include. Accepts: readcount: RNA read\r\n",
      "                       count for each cell (default: readcount)\r\n",
      "  --fi_c FI_C          Path of input tsv file for extra property columns for\r\n",
      "                       each cell (default: None)\r\n"
     ]
    }
   ],
   "source": [
    "!python3 -m dictys network tofile -h"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
